---
layout: introduction_slides
logo: "GTN"

title: "Introduction to Galaxy"
type: "introduction"
contributors:
  - bagnacan
  - bebatut
  - shiltemann
  - pajanne
  - nsoranzo
---

# What is Galaxy?

---

[![Galaxy](../../../shared/images/galaxy_logo_25percent_transparent.png)](https://galaxyproject.org/)

- **Web-based** platform for computational biomedical research (analysis and data integration)
  - Developed at Penn State, Johns Hopkins, OHSU and Cleveland Clinic with substantial outside contributions
  - **Open source** under [Academic Free License](https://opensource.org/licenses/AFL-3.0)
- More than 5,300 [citations](https://www.zotero.org/groups/1732893/galaxy)

![Galaxy](../images/servers.png)

???

- The Galaxy Team is composed by bioinformaticians and software engineers
- OHSU = Oregon Health & Science University

- More than 125+ [Galaxy servers](https://galaxyproject.org/use/)

  

  - Many more non-public
  - Both general-purpose and domain-specific

---

### Core values

- **Accessibility**
  - Users without programming experience can easily upload/retrieve data, run complex tools and workflows, and visualize data
- **Reproducibility**
  - Galaxy captures information so that any user can understand and repeat a complete computational analysis
- **Transparency**
  - Users can share or publish their analyses (histories, workflows, visualizations)
  - Pages: online Methods for your paper

???

**accessible** **reproducible** **transparent** research means <i>sharing everything</i>.

If the Galaxy framework makes everything as simple as possible, researchers are able to:
- share their analyses
- track all used tools and versions
- check all parameters
- justify each step in the analysis
- publish the findings with all aforementioned information

Pages: interactive, web-based documents that describe a complete analysis.

---

# User Interface

---

### Main Galaxy interface

![Galaxy user interface](../images/galaxy_interface.png)

Home page divided into 3 panels

---

### Top menu

![Top menu](../images/galaxy_interface_menu.png)

Link | Usage
--- | ---
*Analyze Data* | go back to the homepage
*Workflow* | access existing workflows or create new one using the editable diagrammatic pipeline
*Shared data* | access data libraries, histories, workflows, visualizations and pages shared with you
*Visualization* | create new visualisations and launch Interactive Environments
*Help* | links to Galaxy Help Forum (Q&A), Galaxy Community Hub (Wiki), and Interactive Tours
*User* | your preferences and saved histories, datasets, pages and visualizations

---

### Tool interface

![Tool interface](../images/101_11.png)

---

### Tools

- A tool = a text file describing:
  - Input datasets, parameters, commands, and outputs
  - Help, tests, citations, dependency requirements
- Free tool store: [Galaxy Tool Shed](https://toolshed.g2.bx.psu.edu/)
  - Thousands of tools already available
  - Every software can be embedded
     
      If a tool is not available, ask the Galaxy community for help!

  - Only a Galaxy admin can install tools
- New versions can be installed without removing old ones to ensure reproducibility

???

A tool describes:
- the input datasets and their datatypes
- the tool parameters (numerical, text, boolean, selections, colour)
- how to generate a command to execute the tool with the specified inputs and parameters
- the output datasets the tool should produce and their datatypes

Tools can be viewed as tiny LEGO pieces: each one solves a specific problem, and they can be combined together to build complex analysis pipelines.

---

### History

- Location of all analyses <img style="float: right;" alt="History" src="../images/history.png" />
  - collects all datasets produced by tools
  - collects all operations performed on the data

- For each dataset (the heart of Galaxyâ€™s reproducibility), the history tracks 
  - name, format, size, creation time, datatype-specific metadata
  - tool id, version, inputs, parameters
  - standard output (`stdout`) and error (`stderr`)
  - state (<span style="background-color: grey">waiting</span>, <span style="background-color: yellow">running</span>, <span style="background-color: green">success</span>, <span style="background-color: red">failed</span>)
  - hidden, deleted, purged

- Multiple histories
  - each history should correspond to a **different analysis**
  - need a meaningful **name**


???

We say <i>datasets</i> to refer to files as well as databases

Purged means permanently deleted
---

### History

History behavior is controlled by the {% icon galaxy-gear %} *History options* <img style="float: right;" alt="History menu" src="../images/galaxy_interface_history_long_menu.png" />

- *Create New* history will **not** make your current history disappear
- *Saved Histories* to list all your histories
- *Copy Datasets* from one history to another and save disk space for your quota

---

### Importing data

- Copy/paste from a file
- Upload data from a local computer
- Upload data from internet using URL
- Upload data from online databases: UCSC, BioMart, ENCODE, modENCODE, Flymine etc.
- Import from Shared Data (libraries, histories, pages)
- Upload data from FTP (>2GB)

See [Getting data into Galaxy](../../galaxy-data-manipulation/tutorials/get-data/slides.html)

---

### Datatypes

- Uploaded dataset
  - Datatype can be automatically detected
  - Datatype assigned by user
- Dataset produced by a tool: datatype assigned by the tool
- Tools only accept input datasets with the appropriate datatypes
- To change the datatype of a dataset: 
  - {% icon galaxy-pencil %} *Edit Attributes* and *Datatype*
  - {% icon galaxy-pencil %} *Edit Attributes* and *Convert Formats*

???

- Edit Attributes -> Datatype is used to fix a wrongly assigned datatype
- Edit Attributes -> Convert Formats creates a new dataset using a tool that converts the original dataset in the new format
- New datatypes can be added to the Galaxy code base, if missing

---

### Reference genomes

- Genome build specifies which genome assembly a dataset is associated with
  - e.g. mm10, hg38...
- Genome build can be automatically detected or assigned by user
- User can define their own custom genome build
- New genome assemblies can be added by the site Galaxy admin

---

### Workflows

![Workflow interface](../images/RNA-Seq_workflow_editing.png)

- **Extracted** from a history
- **Built manually** by adding and configuring tools using the canvas
- **Imported** using an existing shared workflow

???

- Boxes are workflow steps
  - 2 types: *input* and *tool* steps
- Steps are connected by arrows representing the flow of datasets
- Tool panel on the left with Inputs on top (to add input datasets and collections)
- Small tool form on the right

- **Extracted** from a history
- Allow to easily convert an existing history into an analysis workflow

---

### Why would you want to create workflows?

- **Re-run** the same analysis on different input data sets
- **Change parameters** before re-running a similar analysis
- Make use of the workflow job **scheduling**
  - jobs are submitted as soon as their inputs are ready
- Create **sub**-workflows: a workflow inside another workflow
- **Share** workflows for publication and with the community

---

### Visualizations

.pull-left[
![Visualization](../images/visualizations.png)
]

.pull-right[
- Visualization
  - Charts 
  - {% icon galaxy-barchart %} Datatype associated visualizations
  - Trackster: Track browser to visualize genomic data in a tightly integrated way
- Interactive environments: Jupyter, RStudio, etc
]

---

### Shared data

- 3 ways to share Galaxy items - histories, workflows, visualizations, and pages
  - Directly using a Galaxy account's email addresses on the same instance
  - Using a web link, with anyone who knows the link
  - Using a web link and publishing it to make it accessible to everyone from the *Shared Data* menu

- Tools are shared using the free tool store: [Galaxy Tool Shed](https://toolshed.g2.bx.psu.edu/)

---

### Community

Be part of an active and friendly community

- Support and Questions: [Galaxy Help](https://help.galaxyproject.org/)

  ![Galaxy Help](../images/galaxy_help.png)

- Community curated documentation: [Galaxy Community Hub](https://www.galaxyproject.org/)
- Training around Galaxy for scientists, developers, admins, instructors: [Galaxy Training Community]({{site.url}}/{{site.baseurl}})