<!doctype html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="x-ua-compatible" content="ie=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        <title>Galaxy Training!</title>
        <link rel="stylesheet" href="/training-material/assets/css/bootstrap.min.css?v=3">
        <link rel="stylesheet" href="/training-material/assets/css/main.css?v=2">
        <link rel="stylesheet" href="/training-material/assets/css/font-awesome.css">
        <link rel="stylesheet" href="/training-material/assets/css/academicons.css">
        <link rel="stylesheet" href="/training-material/assets/css/syntax_highlighting.css">
        <link rel="shortcut icon" href="/training-material/favicon.ico" type="image/x-icon" />
    </head>
    <body>
        




<header>
    <nav class="navbar navbar-expand-lg navbar-dark">
        <div class="container">
            <a class="navbar-brand" href="/training-material/">
                <img src="/training-material/assets/images/GTN-60px.png" height="30" alt="Galaxy Training Network logo">
                Galaxy Training!
            </a>

            <button class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#top-navbar" aria-controls="top-navbar" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>
            <div class="collapse navbar-collapse" id="top-navbar">
                <ul class="navbar-nav">
                    <li class="nav-item">
                        <a class="nav-link" href="/training-material/topics/proteomics" title="Go back to list of tutorials">
                            <i class="fa fa-folder-o" aria-hidden="true"></i><span class="visually-hidden">topic</span> Proteomics
                        </a>
                    </li>

                    
                    
                    

                    
                        
                    

                    
                        <li class="nav-item dropdown">
                            <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Where to run the tutorial">
                                <i class="fa fa-cog" aria-hidden="true"></i><span class="visually-hidden">instances</span> Galaxy Instances
                            </a>
                            <div class="dropdown-menu">
                                
                                    <a class="dropdown-item" href="https://github.com/galaxyproject/training-material/tree/master/topics/proteomics/docker" title="Docker image for this tutorial">
                                        <i class="fa fa-ship" aria-hidden="true"></i><span class="visually-hidden">docker_image</span> Docker image
                                    </a>
                                
                                
                                    <div class="dropdown-divider"></div>
                                
                                
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                        <a class="dropdown-item" href="https://usegalaxy.eu" title="Galaxy Europe">
                                            Galaxy Europe
                                        </a>
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                        <a class="dropdown-item" href="http://mississippi.fr/" title="MISSISSIPPI">
                                            MISSISSIPPI
                                        </a>
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                        <a class="dropdown-item" href="http://galaxy.workflow4metabolomics.org" title="workflow4metabolomics">
                                            workflow4metabolomics
                                        </a>
                                        
                                    
                                
                            </div>
                        </li>
                    

                    

                    

                    
                    <li class="nav-item">
                        <a class="nav-link" href="/training-material/topics/proteomics#references" title="References">
                            <i class="fa fa-book" aria-hidden="true"></i><span class="visually-hidden">references</span> Literature
                        </a>
                    </li>
                    

                    <li class="nav-item dropdown">
    <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Help">
        <i class="fa fa-life-ring" aria-hidden="true"></i><span class="visually-hidden">help</span> Help
    </a>
    <div class="dropdown-menu">
        <a class="dropdown-item" href="/training-material/faq" title="Check our FAQ">
            FAQ
        </a>
        <a class="dropdown-item" href="https://gitter.im/Galaxy-Training-Network/Lobby" title="Chat on Gitter">
            Gitter
        </a>
        <a class="dropdown-item" href="https://biostar.usegalaxy.org/" title="Ask on Biostars">
            Biostars
        </a>
    </div>
</li>


                    <li class="nav-item">
                        <a class="nav-link" href="https://github.com/galaxyproject/training-material/edit/master/topics/proteomics/tutorials/database-handling/tutorial.md">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> Edit
                        </a>
                    </li>
                </ul>
            </div>
        </div>
    </nav>
</header>

<div class="container main-content">
    <section class="tutorial">
        <h1>Protein FASTA Database Handling</h1>

        <blockquote class="overview">
            <h3>Overview</h3>

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>How to download protein FASTA databases of a certain organism?</li>
            
            <li>How to download a contaminant database?</li>
            
            <li>How to create a decoy database?</li>
            
            <li>How to combine databases?</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>Creation of a protein FASTA database ready for use with database search algorithms.</li>
            
            </ul>

            
            <strong><i class="fa fa-check-circle" aria-hidden="true"></i><span class="visually-hidden">requirements</span> Requirements</strong>
            <ul>
            
                <li>
                    
                    <a href="/training-material/topics/introduction/">Galaxy introduction</a>
                    
                </li>
            
            
            </ul>
            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 30min</p>
            
        </blockquote>

        <h1 class="no_toc" id="introduction">Introduction</h1>

<p>In mass spectrometry based proteomics experiments, peptides are assigned to experimentally acquired tandem mass spectra (MS2) by a method called as peptide-spectral matching. Peptide spectral matching is commonly achieved by using search algorithms to match the MS2 spectra to theoretical spectra. The theoretical spectra are generated from theoritical digestion of proteins in the FASTA database. Ideally, the protein FASTA databases will contain all proteins of the organism under investigation.</p>

<blockquote class="agenda">
  <h3 id="agenda">Agenda</h3>

  <p>In this tutorial, we will deal with:</p>

<ol id="markdown-toc">
  <li><a href="#uploading-a-protein-search-database" id="markdown-toc-uploading-a-protein-search-database">Uploading a protein search Database</a></li>
  <li><a href="#contaminant-databases" id="markdown-toc-contaminant-databases">Contaminant databases</a></li>
  <li><a href="#merging-databases" id="markdown-toc-merging-databases">Merging databases</a></li>
  <li><a href="#creating-a-decoy-database" id="markdown-toc-creating-a-decoy-database">Creating a Decoy Database</a></li>
</ol>

</blockquote>

<h1 id="uploading-a-protein-search-database">Uploading a protein search Database</h1>

<p>There are a many ways how you can upload your protein search database (FASTA file with protein sequences). Three of these ways are:</p>

<ul>
  <li>Using <strong>Protein Database Downloader</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> .</li>
  <li>Using a direct weblink to the database.</li>
  <li>Uploading a database from the data library.</li>
</ul>

<p>In this tutorial, we will explore <strong>Protein Database Downloader</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> for generating a protein search database. For this we will download the proteome of an organism of interest. In this tutorial, we will use a database of the human proteome.</p>

<blockquote class="hands_on">
  <h3 id="hands_on-hands-on-uploading-a-protein-search-database">
<i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Uploading a protein search database</h3>

  <ol>
    <li>Create a new history for this Database Handling exercise.</li>
    <li>Open <strong>Protein Database Downloader</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>
</li>
    <li>Select in the drop-down menues <code class="highlighter-rouge">Taxonomy</code>: “Homo sapiens (Human)” and <code class="highlighter-rouge">reviewed</code>: “UniprotKB/Swiss-Prot (reviewed only)”.</li>
    <li>Click on <code class="highlighter-rouge">Execute</code>. There will be a new dataset named <code class="highlighter-rouge">Protein database</code> in your history, now.</li>
    <li>Rename the <code class="highlighter-rouge">Protein database</code> to <code class="highlighter-rouge">Main database</code>.</li>
  </ol>

  <blockquote class="tip">
    <h3 id="tip-tip-types-of-uniprot-databases">
<i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Types of uniprot databases</h3>
    <p>Uniprot offers several types of databases. You may choose to download only reviewed (UniProtKB/Swissprot) databases, only unreviewed (UniProtKB/TREMBL) or both (UniProtKB). In model organisms which are well-researched, e.g. <em>Homo sapiens</em> or <em>D. melanogaster</em>, reviewed (Swissprot) databases contain curated proteins and may lead to smaller databases and cleaner search results. However, if the researcher is interested in identifying proteins that are unreviewed, it might be wiser to include the TrEMBL database.</p>

    <p>You may also include protein isoforms by setting the tick box <code class="highlighter-rouge">Include isoform data</code> to <code class="highlighter-rouge">Yes</code>.</p>
  </blockquote>

  <blockquote class="question">
    <h3 id="question-question">
<i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>
    <p>What is the difference between a “reference proteome set” and a “complete proteome set”?</p>

    <blockquote class="solution">
      <h3 id="solution-solution">
<i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>
      <ol>
        <li>A UniProt complete proteome consists of the set of proteins thought to be expressed by an organism whose genome has been completely sequenced. A reference proteome is the complete proteome of a representative, well-studied model organism or an organism of interest for biomedical and biotechnological research. Reference proteomes constitute a representative cross-section of the taxonomic diversity to be found within UniProtKB. Species of particular importance may be represented by numerous reference proteomes for specific ecotypes or strains of interest. <a href="http://www.uniprot.org/help/reference_proteome">Link to source</a>
</li>
      </ol>
    </blockquote>
  </blockquote>
</blockquote>

<h1 id="contaminant-databases">Contaminant databases</h1>

<p>In proteomic samples, some protein contaminants are commonly present. These protein contamianants are introduced into sample during sample preparation. The source of these contamiants could be an experimenter or contaminated cell culture. In order to avoid misidentification of spectra derived from contaminants, protein sequences of common laboratory contaminants are added to the database. This has two benefits:</p>
<ol>
  <li>Contamination can be observed, heavily contaminated samples can be excluded from analysis.</li>
  <li>Contaminant peptides cannot be misassigned to similar peptides in the database reducing the risk of identifying false positives.</li>
</ol>

<p>A widely used database for common contaminants is the <strong>c</strong>ommon <strong>R</strong>epository of <strong>A</strong>dventitious <strong>P</strong>roteins (cRAP). When using samples generated in cell cultures, it is furthermore recommended to include <em>Mycoplasma</em> proteomes in the search database. <em>Mycoplasma</em> infections are very common in cell culture and often go unnoticed (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3463982/">Drexler and Uphoff, Cytotechnology, 2002</a>).</p>

<blockquote class="hands_on">
  <h3 id="hands_on-hands-on-contaminant-databases">
<i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Contaminant databases</h3>
  <ol>
    <li>Open <strong>Protein Database Downloader</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>.</li>
    <li>Select <code class="highlighter-rouge">Download from</code>: “cRAP (contaminants)” and execute.</li>
    <li>Rename the new database to “crap database”.</li>
  </ol>

  <p>To be able to distinguish contaminants from proteins of interest, you should add a tag to each contaminant protein.</p>

  <ol>
    <li>Run <strong>FASTA-to-Tabular</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> on your crap database.</li>
    <li>Run <strong>Add column</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> on the new output. In the field <code class="highlighter-rouge">Add this value</code> enter “CONTAMINANT” and execute.</li>
    <li>Run <strong>Tabular-to-FASTA</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>. Use column 1 and column 3 as Title columns and column 2 as sequence column.</li>
    <li>Rename the <strong>Tabular-to-FASTA</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> output to “Tagged cRAP database”.</li>
  </ol>

  <blockquote class="question">
    <h3 id="question-question-1">
<i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>
    <ol>
      <li>The cRAP database contains some human proteins. What does it mean if you identify those typical contaminants in a human sample?</li>
      <li>What does it mean in a non-human sample?</li>
    </ol>

    <blockquote class="solution">
      <h3 id="solution-solution-1">
<i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>
      <ol>
        <li>In samples stemming from a human source, identified human contaminants do not necessarily mean a contaminated sample. The proteins may as well originate from the research study sample. Users are advised to use discretion when interpreting the data.</li>
        <li>In samples from non-human sources, identified human contaminants do mean contamination by the experimenter.</li>
      </ol>
    </blockquote>
  </blockquote>
</blockquote>

<blockquote class="hands_on">
  <h3 id="hands_on-optional-hands-on-mycoplasma-databases">
<i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Optional Hands-On: <em>Mycoplasma</em> databases</h3>
  <p>90 - 95 % of mycoplasma infection in cell culture originates from the following species: <em>M. orale, M. hyorhinis, M. arginini, M. fermentans, M. hominis</em> and <em>A. laidlawii</em> (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3463982/">Drexler and Uphoff, Cytotechnology, 2002</a>).</p>

  <ol>
    <li>Use <strong>Protein Database Downloader</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> to download the six mycoplasma databases. We will merge them to the main database in the next part of the tutorial.</li>
    <li>Run <strong>FASTA Merge Files and Filter Unique Sequences</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> to combine all mycoplasma databases into a single one.</li>
    <li>Tag each entry in the combined database with the string “MYCOPLASMA_CONTAMINANT” by using <strong>FASTA-to-Tabular</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>, <strong>Add column</strong> <img class="emoji" title=":wrench:" alt=":wrench:" src="https://assets-cdn.github.com/images/icons/emoji/unicode/1f527.png" height="20" width="20"> and <strong>Tabular-to-FASTA</strong> <img class="emoji" title=":wrench:" alt=":wrench:" src="https://assets-cdn.github.com/images/icons/emoji/unicode/1f527.png" height="20" width="20">, as explained <a href="#contaminant-databases">above</a>.</li>
    <li>Rename the <strong>Tabular-to-FASTA</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> output to “Tagged Mycoplasma database”.</li>
  </ol>

  <blockquote class="comment">
    <h3 id="comment-comment">
<i class="fa fa-commenting-o" aria-hidden="true"></i><span class="visually-hidden">comment</span> Comment</h3>
    <p>The reviewed mycoplasma databases do not contain all known proteins. It is better to also include the TREMBL database. <em>Mycoplasma</em> proteomes are relatively small, so even downloading TrEMBL sequences will not incraese the size of your main database by much.</p>
  </blockquote>
</blockquote>

<h1 id="merging-databases">Merging databases</h1>

<p>Depending on the search algorithm in use,  you might need to merge all FASTA entries (i.e. proteins of interest and contaminants) in a single database. Make sure to merge the tagged versions of your contaminant databases.</p>

<blockquote class="hands_on">
  <h3 id="hands_on-hands-on-merging-databases">
<i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Merging databases</h3>

  <ol>
    <li>
      <p>Run <strong>FASTA Merge Files and Filter Unique Sequences</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> on the main database and the tagged cRAP database.</p>
    </li>
    <li>
      <p><strong>Optional</strong>: Merging mycoplasma databases</p>

      <p>At this step you may also merge the mycoplasma protein databases that you downloaded earlier on. Simply enter them as additional inputs in <strong>FASTA Merge Files and Filter Unique Sequences</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>. You can enter any number of protein databases when you click on <code class="highlighter-rouge">Insert Input FASTA file(s)</code>.</p>
    </li>
  </ol>
</blockquote>

<h1 id="creating-a-decoy-database">Creating a Decoy Database</h1>

<p>The most common method of peptide and protein False Discovery Rate (FDR) calculation is by adding protein sequences that are not expected to be present in the sample. These are also called decoy protein sequences. This can be done by generating reverse sequences of the target protein entries and appending these protein entries to the protein database. Some search algoritmms use premade target-decoy protein sequences while others can generate a target-decoy protein sequence database from a target protein sequence database before using them for peptide spectral matching.</p>

<blockquote class="hands_on">
  <h3 id="hands_on-hands-on-creating-a-decoy-database">
<i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Creating a Decoy Database</h3>
  <ol>
    <li>Run <strong>DecoyDatabase</strong>  <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> on the merged database.</li>
    <li>Set the flag (-append) to <code class="highlighter-rouge">Yes</code> and execute.</li>
  </ol>

  <blockquote class="tip">
    <h3 id="tip-tip-decoy-tags">
<i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Decoy tags</h3>
    <p>The string you enter as a decoy tag will be added as a prefix or suffix (your choice) to the description of each decoy protein entry. Thus you can see from which entry in the target database the decoy was computed.</p>
  </blockquote>

  <blockquote class="comment">
    <h3 id="comment-comment-1">
<i class="fa fa-commenting-o" aria-hidden="true"></i><span class="visually-hidden">comment</span> Comment</h3>
    <p><strong>DecoyDatabase</strong>  <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> may also take several databases as input which are then automatically merged into one database.</p>
  </blockquote>
</blockquote>

<h1 class="no_toc" id="concluding-remarks">Concluding remarks</h1>

<p>In order to keep your proteins databases up-to-date, it is recommended to create a workflow out of the hands-on sections (to learn about workflows see <a href="/training-material/topics/introduction/tutorials/galaxy-intro-101/tutorial.html">this tutorial</a>). You might also want to combine the mycoplasma databases to a single file, which you then easily can add to each of your main databases.</p>

<p>Often you may not want to use the most recent database for reasons of reproducibility. If so, you can transfer the final database of this tutorial into other histories to work with it.</p>

<p>Further reading about construction of the optimal database: (<a href="https://www.ncbi.nlm.nih.gov/pubmed/27975281">Kumar et al., Methods in molecular biology, 2017</a>).</p>

<p>This tutorial is based upon parts of the GalaxyP-101 tutorial (https://usegalaxyp.readthedocs.io/en/latest/sections/galaxyp_101.html).</p>


        
        <blockquote class="key_points">
            <h3>
<i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

            <ul>
                
                <li>There are several types of Uniprot databases.</li>
                
                <li>Search databases should always include possible contaminants.</li>
                
                <li>For analyzing cell culture or organic samples, search databases should include mycoplasma databases.</li>
                
                <li>Some peptide search engines depend on decoys to calculate the FDR.</li>
                
            </ul>
        </blockquote>
        

        
        <h1>Useful literature</h1>
        <p>Further information, including links to documentation and original publications, regarding the tools, analysis techniques and the interpretation of results described in this tutorial can be found <a href="/training-material/topics/proteomics#references">here</a>.</p>
        

        

        <h3>
<i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>
        <hr>
        <br>

        <iframe class="google-form" src="https://docs.google.com/forms/d/e/1FAIpQLSd4VZptFTQ03kHkMz0JyW9b6_S8geU5KjNE_tLM0dixT3ZQmA/viewform?embedded=true&amp;entry.1235803833=Protein%20FASTA%20Database%20Handling%20(Proteomics)">Loading...
        </iframe>

        <blockquote class="feedback">
            <h3>
<i class="fa fa-comments-o" aria-hidden="true"></i><span class="visually-hidden">feedback</span> Give us even more feedback on this content!</h3>
            <p>To give us more detailed feedback about these materials, please take a moment to fill in the extended <a href="https://docs.google.com/forms/d/e/1FAIpQLSc_GLZRpxZGAL9tN6DA1bE6WkNhmQdXT7B16TpOb-X4YwKUsQ/viewform?entry.1548889262=Protein%20FASTA%20Database%20Handling%20(Proteomics)">Feedback Form</a>.</p>
        </blockquote>
    </section>
</div>


<footer>
    <div class="container">
        <p>
            This material is the result of a collaborative work. Thanks to the
            <a href="https://wiki.galaxyproject.org/Teach/GTN">Galaxy Training Network</a>
            and all the <a href="/training-material/hall-of-fame">contributors</a> (Björn Grüning, Florian Christoph Sigloch)!
        </p>
        <p>
            Found a typo? Something is wrong in this tutorial? Edit it on
            <a href="https://github.com/galaxyproject/training-material/tree/master/topics/proteomics/tutorials/database-handling/tutorial.md">GitHub</a>.
        </p>
    </div>
</footer>

    </body>
    <script type="text/javascript" src="/training-material/assets/js/jquery.slim.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/popper.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/bootstrap.min.js?v=3"></script>
    <script type="text/javascript" src="/training-material/assets/js/details-element-polyfill.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="/training-material/assets/js/main.js"></script>
</html>
