---
layout: tutorial_slides
logo: "GTN"

title: "Introduction to metatranscriptomics"
contributors:
  - subinamehta
  - pratikdjagtap
  - shiltemann
---

## Why study the microbiome?

.pull-left[

- Health care research
 - Humans are full of microorganisms
 - Skin, gut, oral cavity, nasal cavity, eyes, ..
 - Affects health, drug efficacy, etc

]
.pull-right[

.image-100[ ![](../../images/human_microbiome.png) ]

]

- Sometimes referred to as your **second genome**
 - ~10 times more cells than you
 - ~100 times more genes than you
 - ~1000s different species


---

## Why study the microbiome?

- Environmental studies
  - Microbes in the soil affect plants and animals
  - Improve agriculture

.image-75[ ![](../../images/environmentalstudies.png) ]


---

## Meta- Omics

![](../../images/metatranscriptomics/metaomics.png)


---

## Meta- Omics

![](../../images/metatranscriptomics/metaomics2.png)

---

## ASaiM pipeline

![ASiaiM diagram](../../images/asaim-wf.png)

.footnote[Batut et al Gigascience. 2018 7(6) doi: 10.1093/gigascience/giy057]

---

## This tutorial


???

For this short tutorial, while the workflow is running, these slides can be useful
to explain the tools that are being run in that section. After explaining the tools,
the workflows should be far enough along to start showing the results


---

## Cellulose Degradation in a Biogas Reactor

![](../../images/metatranscriptomics/experimental_setup.png)



???

A 100 µl aliquot of an enriched community from a biogas reactor was transferred to 27 anaerobic bottles containing a rich medium and 10g/L of cellulose as sole carbon source and incubated at 65 °C.

Three bottles were collected at 9 different time points (0, 8, 13, 18, 23, 28, 33, 38 and 43 h) and processed in triplicates. Metatranscriptomic analysis was performed on all time points. Metaproteomics analysis on 4 data points.

---

## Input: FastQ Files

```
@Identifier1 (comment)
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+
QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ
@Identifier2 (comment)
XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+
QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ
```

???

- Four lines per read
- @ + identifier on first line, just like fasta
- sequence
- +
- quality score characters

Segue: so what do the quality chars mean?

---

### Quality score

Measure of the quality of the identification of the nucleobases <br>generated by automated DNA sequencing

<small>

Phred Quality Score | Probability of incorrect base call | Base call accuracy
--- | --- | ---
10 | 1 in 10 | 90%
20 | 1 in 100 | 99%
30 | 1 in 1000 | 99.9%
40 | 1 in 10,000 | 99.99%
50 | 1 in 100,000 | 99.999%
60 | 1 in 1,000,000 | 99.9999%

</small>

???

- Logarithmic scale

---

### Quality score

![](../../../sequence-analysis/images/quality_score.png)

???

- Solexa uses slightly different measure

---

### Quality score encoding

![](../../../sequence-analysis/images/quality_score_encoding.png)

???

- Earlier saw sequence + quality were same length of text
- Each score encoded as a single character
- Uses ASCII (common computer encoding, maps numbers to characters)
- Each technology has a different mapping
- Mostly see illumina 1.8 these days, others only if historical



---

## Preprocessing

In this tutorial we start with some preprocessing steps

| Step | Tools |
|:------|-------|
|Quality Control reports &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; | FastQC {% icon tool %} and MultiQC {% icon tool %} |
|Trimming and Filtering | Cutadapt {% icon tool %} |
|Filter ribosomal RNA | SortMeRNA {% icon tool %} |
|Interlace FastQ files | FastQ interlacer {% icon tool %} |

---

### FastQC

- Generate a web report with quality metrics of your FastQ file

![](../../../sequence-analysis/images/fastqc-report.png)

.footnote[see also our [dedicated QC tutorial]({% link topics/sequence-analysis/tutorials/quality-control/tutorial.md %})  ]

---

### FastQC

- Many different QC plots
- Example: Per-base sequence Quality plot

![](../../../sequence-analysis/images/per_base_sequence_quality_good.png)


.footnote[explanation of different plots: [dedicated QC tutorial]({% link topics/sequence-analysis/tutorials/quality-control/tutorial.md %})  ]

???
- in the per-base sequence quality plot, a boxplot of the base quality (y-axis) per position in the read (x-axis) is drawn
- often you might observe a drop in quality towards the end of the reads, and may consider trimming ends
- this example is very good

---

### FastQC

- Many more plots
- See [QC tutorial]({% link topics/sequence-analysis/tutorials/quality-control/tutorial.md %}) for more information

![](../../../sequence-analysis/images/quality-control/all_plots.png)


---

### MultiQC

.pull-left[
- Combine multiple FastQC reports into one report

- Also for outputs of other tools

- Great when sequencing large numbers of samples
]

.pull-right[
![](../../images/metatranscriptomics/multiqc.png)
]

---

### Read Trimming and Filtering

- Trim low-quality bases from reads
- Filter reads based on length, mean quality score, ..
- Remove adapters/primers
  .image-75[![adapter trimming](../../../sequence-analysis/images/quality-control/trimming.png)]

- Many tools: CutAdapt {% icon tool %}, TrimGalore {% icon tool %}, Trimmomatic {% icon tool%} ..

???
These are some examples of ways to trim and filter data, but many more are possible and depend on your experiment what is necessary

---

### SortMeRNA

.pull-left[

- Most RNA sequences will be ribosomal RNA (rRNA)

- Great for taxonomical assignment (who is there?)

- Not informative for functional analysis (What are they doing?)

- Filter out rRNA before doing functional analysis
]

.pull-right[
.image-90[![SortMeRNA](../../images/metatranscriptomics/sortmerna.png)]
]

---

### FastQ interlace

- Paired-end data often comes in two separate FastQ files
- One file with *forward* reads, one with *reverse* reads

<br>

![paired end deinterlaced file](../../../sequence-analysis/images/mapping/pairedend_fastq.png)

---

### FastQ interlace

- Some tools require a single **interlaced** FastQ file
- Galaxy has tools for **interlacing** and **deinterlacing** FastQ files

<br>

![paired end interlaced file](../../../sequence-analysis/images/mapping/pairedend_interlaced.png)


???
forward and reverse files are 'zipped' together into a single file

---

## Community Profile

- We want to identify which organisms are present in our sample, and their relative abundances

![](../../images/microbial-community.png)

- **MetaPhlan2** {% icon tool %} for identification
- **Krona** {% icon tool %} and **Graphlan** {% icon tool %} for visualisation

---

## MetaPhlan2 {% icon tool %}

- Estimates the presence and relative abundance of microbial cells
- Maps reads against a set of marker sequences
- Caveat: this tool is designed for DNA-seq
  - Be careful interpreting abundances when using this tool with transcriptomics data

.footnote[Nat Methods. 2012 Jun 10;9(8):811-4. doi: 10.1038/nmeth.2066.]

???

About the caveat: The theoretical problem is that we quantify species abundance by averaging the coverage of marker genes. Marker genes are supposed to be at the same coverage as they are single copy genes from the same genome, but this is not true for their transcripts. So MetaPhlAn2 on metatranscriptomics gives an idea about the average transcriptional rate of a given species. So it _can_ be used with caution...


---

## Krona {% icon tool %}

- Visualization of community composition, interactive plot

<iframe id="krona" src="krona.html" frameBorder="0" width="80%" height="600px"> ![Krona at bacteria level](../../images/metatranscriptomics/krona_bacteria.png) </iframe>

---

## Graphlan {% icon tool %}

- *Cladogram* visualisation

![](../../images/metatranscriptomics/graphlan_general.png)

---

## Genus Abundance

- Tutorial: one timepoint
- Over multiple timepoints:

![](../../images/metatranscriptomics/genus_abundance.png)

---

## Functional Analysis

<!-- TODO: Subina -->

