<!doctype html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="x-ua-compatible" content="ie=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        <title>RNA-seq counts to genes and pathways</title>
        
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
            ga('create', 'UA-45719423-18' , 'auto');
            ga('send', 'pageview');
        </script>
        
        <link rel="stylesheet" href="/training-material/assets/css/bootstrap.min.css?v=3">
        <link rel="stylesheet" href="/training-material/assets/css/main.css?v=2">
        <link rel="stylesheet" href="/training-material/assets/css/font-awesome.css">
        <link rel="stylesheet" href="/training-material/assets/css/academicons.css">
        <link rel="stylesheet" href="/training-material/assets/css/syntax_highlighting.css">
        <link rel="shortcut icon" href="/training-material/favicon.ico" type="image/x-icon" />

        
        
        
        
        
        <meta name="description" content="Training material for all kinds of transcriptomics analysis." />
        <meta property="og:title" content="Galaxy Training: RNA-seq counts to genes and pathways" />
        <meta property="og:description" content="Training material for all kinds of transcriptomics analysis." />
        <meta property="og:image" content="/training-material/assets/images/GTNLogo1000.png" />
    </head>
    <body>
        





<header>
    <nav class="navbar navbar-expand-lg navbar-dark">
        <div class="container">
            <a class="navbar-brand" href="/training-material/">
                <img src="/training-material/assets/images/GTN-60px.png" height="30" alt="Galaxy Training Network logo">
                Galaxy Training!
            </a>

            <button class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#top-navbar" aria-controls="top-navbar" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>
            <div class="collapse navbar-collapse" id="top-navbar">
                <ul class="navbar-nav">
                    <li class="nav-item">
                        <a class="nav-link" href="/training-material/topics/transcriptomics" title="Go back to list of tutorials">
                            <i class="fa fa-folder-o" aria-hidden="true"></i><span class="visually-hidden">topic</span> Transcriptomics
                        </a>
                    </li>

                    
                    
                    

                    
                        
                            <li class="nav-item dropdown">
                                <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Introduction slides">
                                    <i class="fa fa-slideshare" aria-hidden="true"></i><span class="visually-hidden">slides</span> Introduction slides
                                </a>
                                <div class="dropdown-menu">
                                    
                                        
                                            
                                                <a class="dropdown-item" href="/training-material/topics/transcriptomics/slides/introduction.html">
                                                    Introduction to Transcriptomics
                                                </a>
                                            
                                        
                                    
                                        
                                            
                                        
                                    
                                        
                                            
                                        
                                    
                                        
                                            
                                        
                                    
                                        
                                            
                                        
                                    
                                        
                                    
                                        
                                    
                                        
                                            
                                        
                                    
                                        
                                            
                                        
                                    
                                        
                                            
                                        
                                    
                                </div>
                            </li>
                        
                    

                    

                    

                    
                    <li class="nav-item">
                        <a class="nav-link" href="https://figshare.com/s/f5d63d8c265a05618137" title="Links to data">
                            <i class="fa fa-files-o" aria-hidden="true"></i><span class="visually-hidden">zenodo_link</span> Input Dataset
                        </a>
                    </li>
                    

                    
                    <li class="nav-item">
                        <a class="nav-link" href="/training-material/topics/transcriptomics#references" title="References">
                            <i class="fa fa-bookmark" aria-hidden="true"></i><span class="visually-hidden">references</span> Literature
                        </a>
                    </li>
                    

                    <li class="nav-item dropdown">
                        <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="EN">
                            EN
                        </a>
                        <div class="dropdown-menu">
                            
                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=fr&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Ftranscriptomics%2Ftutorials%2Flimma-voom%2Ftutorial.html&edit-text=&act=url" title="">
                                FR
                            </a>
                            
                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=ja&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Ftranscriptomics%2Ftutorials%2Flimma-voom%2Ftutorial.html&edit-text=&act=url" title="">
                                JA
                            </a>
                            
                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=es&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Ftranscriptomics%2Ftutorials%2Flimma-voom%2Ftutorial.html&edit-text=&act=url" title="">
                                ES
                            </a>
                            
                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=pt&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Ftranscriptomics%2Ftutorials%2Flimma-voom%2Ftutorial.html&edit-text=&act=url" title="">
                                PT
                            </a>
                            
                        </div>
                    </li>

                    <li class="nav-item dropdown">
    <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Help">
        <i class="fa fa-life-ring" aria-hidden="true"></i><span class="visually-hidden">help</span> Help
    </a>
    <div class="dropdown-menu">
        <a class="dropdown-item" href="/training-material/faq" title="Check our FAQ">
            FAQ
        </a>
        <a class="dropdown-item" href="https://gitter.im/Galaxy-Training-Network/Lobby" title="Chat on Gitter">
            Chat on Gitter
        </a>
        <a class="dropdown-item" href="https://biostar.usegalaxy.org/" title="Ask on Biostars">
            Ask questions on Biostars
        </a>
    </div>
</li>


                    <li class="nav-item">
                        <a class="nav-link" href="https://github.com/galaxyproject/training-material/edit/master/topics/transcriptomics/tutorials/limma-voom/tutorial.md">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> Edit
                        </a>
                    </li>
                </ul>
            </div>
        </div>
    </nav>
</header>

<div class="container main-content">
    <script type="application/ld+json">
   
{
  "@context": "http://schema.org",
  "@type": "CreativeWork",
  "name": "RNA-seq counts to genes and pathways",
  "accessModeSufficient": "visual",
  "accessibilityControl": "fullMouseControl",
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "https://github.com/galaxyproject/training-material/blob/master/LICENSE.md",
  "learningResourceType": "",
  "url": "/topics/transcriptomics/tutorials/limma-voom/tutorial.html",
  "mentions": [
    {
      "@type": "Thing",
      "url": "https://figshare.com/s/f5d63d8c265a05618137",
      "name": "Training data for RNA-seq counts to genes and pathways tutorial"
    }
  ],
  "timeRequired": "PT3H",
  "contributor": [
    {
      "@type": "Person",
      "name": "mblue9"
    },
    {
      "@type": "Person",
      "name": "bphipson"
    },
    {
      "@type": "Person",
      "name": "annatrigos"
    },
    {
      "@type": "Person",
      "name": "mritchie"
    },
    {
      "@type": "Person",
      "name": "hdashnow"
    },
    {
      "@type": "Person",
      "name": "charitylaw"
    }
  ],
  "description": "The questions this  addresses are:\n - What are the differentially expressed genes and pathways in the mammary gland of pregnant versus lactating mice?\n - How to analyze RNA count data using limma-voom?\n - How to perform quality control of RNA-seq count data?\n\n\\nThe objectives are:\n - Analysis of RNA-seq count data using limma-voom\n - Quality control of count data\n - Visualisation and interactive exploration of count data\n - Identification of differentially expressed genes and pathways\n\n"
}

    </script>
    <section class="tutorial">
        <h1>RNA-seq counts to genes and pathways</h1>

        <blockquote class="overview">
            <h3>Overview</h3>

            <strong><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li>What are the differentially expressed genes and pathways in the mammary gland of pregnant versus lactating mice?</li>
            
            <li>How to analyze RNA count data using limma-voom?</li>
            
            <li>How to perform quality control of RNA-seq count data?</li>
            
            </ul>

            <strong><i class="fa fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li>Analysis of RNA-seq count data using limma-voom</li>
            
            <li>Quality control of count data</li>
            
            <li>Visualisation and interactive exploration of count data</li>
            
            <li>Identification of differentially expressed genes and pathways</li>
            
            </ul>

            
            <strong><i class="fa fa-check-circle" aria-hidden="true"></i><span class="visually-hidden">requirements</span> Requirements</strong>
            <ul>
            
                <li>
                    
                    <a href="/training-material/topics/introduction/">Galaxy introduction</a>
                    
                </li>
            
                <li>
                    
                    <a href="/training-material/topics/sequence-analysis/">Quality control</a>
                    
                </li>
            
            
            </ul>
            

            
            <p><strong><i class="fa fa-hourglass-end" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 3 hours</p>
            
        </blockquote>

        <h1 class="no_toc" id="introduction">Introduction</h1>

<p>Measuring gene expression on a genome-wide scale has become common practice over the last two decades or so, with microarrays predominantly used pre-2008. With the advent of next generation sequencing technology in 2008, an increasing number of scientists use this technology to measure and understand changes in gene expression in often complex systems. As sequencing costs have decreased, using RNA-Seq to simultaneously measure the expression of tens of thousands of genes for multiple samples has never been easier. The cost of these experiments has now moved from generating the data to storing and analysing it.</p>

<p>There are many steps involved in analysing an RNA-Seq experiment. Analysing an RNA-seq experiment begins with sequencing reads. These are aligned to a reference genome, then the number of reads mapped to each gene can be counted. This results in a table of counts, which is what we perform statistical analyses on to determine differentially expressed genes and pathways. The purpose of this tutorial is to demonstrate how to perform differential expression on count data with <strong>limma-voom</strong>. How to generate counts from reads (FASTQs) is covered in the accompanying tutorial <a href="/training-material/topics/transcriptomics/tutorials/limma-voom_fastqs_to_counts/tutorial.html">RNA-seq reads to counts</a>.</p>

<h2 id="mouse-mammary-gland-dataset">Mouse mammary gland dataset</h2>

<p>The data for this tutorial comes from a Nature Cell Biology paper, <a href="https://www.ncbi.nlm.nih.gov/pubmed/25730472">EGF-mediated induction of Mcl-1 at the switch to lactation is essential for alveolar cell survival</a>), Fu et al. 2015. Both the raw data (sequence reads) and processed data (counts) can be downloaded from Gene Expression Omnibus database (GEO) under accession number <a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE60450">GSE60450</a>.</p>

<p>This study examined the expression profiles of basal stem-cell enriched cells (B) and committed luminal cells (L) in the mammary gland of virgin, pregnant and lactating mice. Six groups are present, with one for each combination of cell type and mouse status. Note that two biological replicates are used here, two independent sorts of cells from the mammary glands of virgin, pregnant or lactating mice, however three replicates is usually recommended as a minimum requirement for RNA-seq. In this tutorial we will use the GEO counts file as a starting point for our analysis. Alternatively, you could create a count matrix from the raw sequence reads, as demonstrated in the <a href="/training-material/topics/transcriptomics/tutorials/limma-voom_fastqs_to_counts/tutorial.html">RNA-seq reads to counts tutorial</a>. The GEO count file was generated from aligning the reads to the mouse <code class="highlighter-rouge">mm10</code> genome with the <a href="https://www.biorxiv.org/content/early/2018/08/15/377762">Rsubread</a> aligner, followed by counting reads mapped to RefSeq genes with <a href="https://academic.oup.com/bioinformatics/article/30/7/923/232889">featureCounts</a> (Liao, Smyth, and Shi 2014), see the <a href="https://www.nature.com/articles/ncb3117">Fu paper</a> for details.</p>

<p>We will use <strong>limma-voom</strong> for identifying differentially expressed genes here. Other popular alternatives are edgeR and DESeq2. Limma-voom has been shown to be perform well in terms of precision, accuracy and sensitivity <a href="https://www.ncbi.nlm.nih.gov/pubmed/29267363">Costa-Silva, Domingues and Lopes 2017</a> and, due to its speed, it’s particularly recommended for large-scale datasets with 100s of samples <a href="https://f1000research.com/articles/5-1438/v2">Chen, Lun, Smyth 2016</a>.</p>

<p>This is a Galaxy tutorial based on material from the <a href="http://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html">COMBINE R RNAseq workshop</a>, first taught <a href="http://combine-australia.github.io/2016-05-11-RNAseq/">here</a>. Some of the gene set testing material is inspired by the Cancer Research UK workshop <a href="https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2018/RNASeq2018/html/06_Gene_set_testing.nb.html">here</a></p>

<figure id="figure-1"><img src="../../images/limma-voom_f2c/mouse_exp.png" alt="Tutorial Dataset" title="Tutorial Dataset" /><figcaption><span class="figcaption-prefix">Figure 1:</span> Tutorial Dataset</figcaption></figure>

<blockquote class="agenda">
  <h3 id="agenda">Agenda</h3>

  <p>In this tutorial, we will deal with:</p>

<ol id="markdown-toc">
  <li><a href="#mouse-mammary-gland-dataset" id="markdown-toc-mouse-mammary-gland-dataset">Mouse mammary gland dataset</a></li>
  <li><a href="#preparing-the-inputs" id="markdown-toc-preparing-the-inputs">Preparing the inputs</a>    <ol>
      <li><a href="#import-data" id="markdown-toc-import-data">Import data</a></li>
      <li><a href="#format-the-data" id="markdown-toc-format-the-data">Format the data</a></li>
      <li><a href="#get-gene-annotations" id="markdown-toc-get-gene-annotations">Get gene annotations</a></li>
    </ol>
  </li>
  <li><a href="#differential-expression-with-limma-voom" id="markdown-toc-differential-expression-with-limma-voom">Differential expression with limma-voom</a>    <ol>
      <li><a href="#filtering-to-remove-lowly-expressed-genes" id="markdown-toc-filtering-to-remove-lowly-expressed-genes">Filtering to remove lowly expressed genes</a></li>
      <li><a href="#normalization-for-composition-bias" id="markdown-toc-normalization-for-composition-bias">Normalization for composition bias</a></li>
      <li><a href="#specify-contrasts-of-interest" id="markdown-toc-specify-contrasts-of-interest">Specify Contrast(s) of interest</a></li>
    </ol>
  </li>
  <li><a href="#quality-control-of-count-data" id="markdown-toc-quality-control-of-count-data">Quality Control of count data</a>    <ol>
      <li><a href="#multidimensional-scaling-plot" id="markdown-toc-multidimensional-scaling-plot">Multidimensional scaling plot</a></li>
      <li><a href="#density-plots" id="markdown-toc-density-plots">Density plots</a></li>
      <li><a href="#box-plots" id="markdown-toc-box-plots">Box plots</a></li>
      <li><a href="#md-plots-for-samples" id="markdown-toc-md-plots-for-samples">MD plots for samples</a></li>
      <li><a href="#voom-variance-plot" id="markdown-toc-voom-variance-plot">Voom variance plot</a></li>
      <li><a href="#md-and-volcano-plots-for-de-results" id="markdown-toc-md-and-volcano-plots-for-de-results">MD and Volcano plots for DE results</a></li>
    </ol>
  </li>
  <li><a href="#testing-relative-to-a-threshold-treat" id="markdown-toc-testing-relative-to-a-threshold-treat">Testing relative to a threshold (TREAT)</a></li>
  <li><a href="#visualising-results" id="markdown-toc-visualising-results">Visualising results</a>    <ol>
      <li><a href="#heatmap-of-top-genes" id="markdown-toc-heatmap-of-top-genes">Heatmap of top genes</a></li>
      <li><a href="#stripcharts-of-top-genes" id="markdown-toc-stripcharts-of-top-genes">Stripcharts of top genes</a></li>
      <li><a href="#interactive-plots-glimma" id="markdown-toc-interactive-plots-glimma">Interactive plots (Glimma)</a></li>
      <li><a href="#create-heatmap-of-custom-genes" id="markdown-toc-create-heatmap-of-custom-genes">Create heatmap of custom genes</a></li>
    </ol>
  </li>
  <li><a href="#gene-set-testing" id="markdown-toc-gene-set-testing">Gene Set Testing</a>    <ol>
      <li><a href="#gene-ontology-testing-with-goseq" id="markdown-toc-gene-ontology-testing-with-goseq">Gene Ontology testing with <strong>goseq</strong></a></li>
      <li><a href="#gene-set-enrichment-analysis-with-fgsea" id="markdown-toc-gene-set-enrichment-analysis-with-fgsea">Gene Set Enrichment Analysis with <strong>fgsea</strong></a></li>
      <li><a href="#ensemble-gene-set-enrichment-analyses-with-egsea" id="markdown-toc-ensemble-gene-set-enrichment-analyses-with-egsea">Ensemble gene set enrichment analyses with <strong>EGSEA</strong></a></li>
    </ol>
  </li>
</ol>

</blockquote>

<h1 id="preparing-the-inputs">Preparing the inputs</h1>

<p>We will use three files for this analysis:</p>

<ul>
  <li><strong>Count matrix</strong> (genes in rows, samples in columns)</li>
  <li><strong>Sample information</strong> file (sample id, group)</li>
  <li><strong>Gene annotation</strong> file (gene id, symbol, description)</li>
</ul>

<h2 id="import-data">Import data</h2>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-data-upload"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Data upload</h3>

  <ol>
    <li>Create a new history for this RNA-seq exercise e.g. <code class="highlighter-rouge">RNA-seq with limma-voom</code></li>
    <li>
      <p>Import the mammary gland counts table and the associated sample information file.</p>

      <p>To import the files, there are two options:</p>
      <ul>
        <li>Option 1: From a shared data library if available (ask your instructor)</li>
        <li>
          <p>Option 2: From <a href="https://figshare.com/s/1d788fd384d33e913a2a">Figshare</a></p>

          <blockquote class="tip">
            <h3 id="tip-tip-importing-data-via-links"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Importing data via links</h3>

            <ul>
              <li>Copy the link location</li>
              <li>Open the Galaxy Upload Manager</li>
              <li>Select <strong>Paste/Fetch Data</strong></li>
              <li>Paste the link into the text field</li>
              <li>Press <strong>Start</strong></li>
            </ul>
          </blockquote>

          <ul>
            <li>
              <p>You can paste both links below into the <strong>Paste/Fetch</strong> box:</p>

              <div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>https://ndownloader.figshare.com/files/5057929?private_link=1d788fd384d33e913a2a
https://ndownloader.figshare.com/files/5999829?private_link=1d788fd384d33e913a2a
</code></pre></div>              </div>
            </li>
            <li>
              <p>Select <em>“Genome”</em>: <code class="highlighter-rouge">mm10</code></p>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li>Rename the counts dataset as <code class="highlighter-rouge">seqdata</code> and the sample information dataset as <code class="highlighter-rouge">sampleinfo</code> using the <i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">galaxy-pencil</span> (pencil) icon.</li>
    <li>
      <p>Check that the datatype is <code class="highlighter-rouge">tabular</code>.
If the datatype is not <code class="highlighter-rouge">tabular</code>, please change the file type to <code class="highlighter-rouge">tabular</code>.</p>

      <blockquote class="tip">
        <h3 id="tip-tip-changing-the-datatype"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Changing the datatype</h3>
        <ul>
          <li>Click on the <i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">galaxy-pencil</span> (pencil) icon displayed in your dataset in the history</li>
          <li>Choose <strong>Datatype</strong> on the top</li>
          <li>Select <code class="highlighter-rouge">tabular</code></li>
          <li>Press <strong>Save</strong></li>
        </ul>
      </blockquote>
    </li>
  </ol>
</blockquote>

<p>Let’s take a look at the data. The <code class="highlighter-rouge">seqdata</code> file contains information about genes (one gene per row), the first column has the Entrez gene id, the second has the gene length and the remaining columns contain information about the number of reads aligning to the gene in each experimental sample. There are two replicates for each cell type and time point (detailed sample info can be found in file “GSE60450_series_matrix.txt” from the GEO website). The first few rows and columns of the seqdata file are shown below.</p>

<figure id="figure-2" width="50%"><img src="../../images/limma-voom/seqdata.png" alt="seqdata file" title="Count file (before formatting)" /><figcaption><span class="figcaption-prefix">Figure 2:</span> Count file (before formatting)</figcaption></figure>

<p>The <code class="highlighter-rouge">sampleinfo</code> file contains basic information about the samples that we will need for the analysis. See below.</p>

<figure id="figure-3" width="50%"><img src="../../images/limma-voom/sampleinfo.png" alt="sampleinfo file" title="Sample information file (before formatting)" /><figcaption><span class="figcaption-prefix">Figure 3:</span> Sample information file (before formatting)</figcaption></figure>

<h2 id="format-the-data">Format the data</h2>

<p>Let’s create a new file, <code class="highlighter-rouge">countdata</code>, that contains only the counts for the 12 samples i.e. we’ll remove the gene length column with the <strong>Cut columns from a table (cut)</strong> tool. The sample names are also pretty long so we’ll use the <strong>Replace Text in entire line</strong> tool to shorten these to contain only the relevant information about each sample.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-format-the-counts-data"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Format the counts data</h3>

  <ol>
    <li><strong>Cut</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>: Run <strong>Cut columns from a table (cut)</strong> with the following settings:
      <ul>
        <li><em>“File to cut”</em>: <code class="highlighter-rouge">seqdata</code></li>
        <li><em>“Operation”</em>: <code class="highlighter-rouge">Discard</code></li>
        <li><em>“List of fields”</em>: Select <code class="highlighter-rouge">Column:2</code></li>
      </ul>
    </li>
    <li><strong>Replace Text</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>: Run <strong>Replace Text in entire line</strong> with the following settings:
      <ul>
        <li><em>“File to process”</em>: output of <strong>Cut</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span></li>
        <li><em>“Find pattern”</em>: <code class="highlighter-rouge">_B[A-Z0-9_]+</code></li>
      </ul>
    </li>
    <li>Rename file as <code class="highlighter-rouge">countdata</code> using the <i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">galaxy-pencil</span> (pencil) icon. The file should look like below.
      <figure id="figure-4"><img src="../../images/limma-voom/countdata.png" alt="countdata file" title="Count file (after formatting)" /><figcaption><span class="figcaption-prefix">Figure 4:</span> Count file (after formatting)</figcaption></figure>
    </li>
  </ol>
</blockquote>

<p>Next, let’s create a new file, <code class="highlighter-rouge">factordata</code>, that contains the groups information that we need for the limma-voom tool. We’ll combine the cell type and mouse status to make 6 groups e.g. we’ll combine the CellType <code class="highlighter-rouge">basal</code> with the Status <code class="highlighter-rouge">pregnant</code> for the group <code class="highlighter-rouge">basalpregnant</code>. We’ll use the <strong>Merge Columns</strong> tool to combine the cell type and mouse status columns in the sample information file, making a column with the 6 group names.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-format-the-sample-information-file"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Format the sample information file</h3>

  <ol>
    <li><strong>Merge Columns</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>: Run <strong>Merge Columns together</strong> with the following settings:
      <ul>
        <li><em>“Select data”</em>: <code class="highlighter-rouge">sampleinfo</code></li>
        <li><em>“Merge column”</em>: <code class="highlighter-rouge">Column: 3</code></li>
        <li><em>“with column”</em>: <code class="highlighter-rouge">Column: 4</code></li>
      </ul>
    </li>
    <li><strong>Cut</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>: Run <strong>Cut columns from a table (cut)</strong> with the following settings:
      <ul>
        <li><em>“File to cut”</em>: output of <strong>Merge Columns</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span></li>
        <li><em>“Operation”</em>: <code class="highlighter-rouge">Keep</code></li>
        <li><em>“List of fields”</em>: Select <code class="highlighter-rouge">Column:2</code> and <code class="highlighter-rouge">Column:5</code></li>
      </ul>
    </li>
    <li>Rename file as <code class="highlighter-rouge">factordata</code> using the <i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">galaxy-pencil</span> (pencil) icon. The file should look like below.
      <figure id="figure-5"><img src="../../images/limma-voom/factordata.png" alt="factordata file" title="Sample information file (after formatting)" /><figcaption><span class="figcaption-prefix">Figure 5:</span> Sample information file (after formatting)</figcaption></figure>
    </li>
  </ol>
</blockquote>

<h2 id="get-gene-annotations">Get gene annotations</h2>

<p>Optionally, gene annotations can be provided to the limma-voom tool and if provided the annotation will be available in the output files. We’ll get gene symbols and descriptions for these genes using the Galaxy <strong>annotateMyIDs</strong> tool, which provides annotations for human, mouse, fruitfly and zebrafish.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-get-gene-annotations"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Get gene annotations</h3>

  <ol>
    <li><strong>annotateMyIDs</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>: Run <strong>annotateMyIDs</strong> with the following settings:
      <ul>
        <li><em>“File with IDs”</em>: <code class="highlighter-rouge">countdata</code></li>
        <li><em>“File has header”</em>: <code class="highlighter-rouge">Yes</code></li>
        <li><em>“Organism”</em>: <code class="highlighter-rouge">Mouse</code></li>
        <li><em>“ID Type”</em>: <code class="highlighter-rouge">Entrez</code></li>
        <li>“*Output columns”: tick
          <ul>
            <li><code class="highlighter-rouge">ENTREZID</code></li>
            <li><code class="highlighter-rouge">SYMBOL</code></li>
            <li><code class="highlighter-rouge">GENENAME</code></li>
          </ul>
        </li>
      </ul>
    </li>
    <li>Rename file as <code class="highlighter-rouge">annodata</code> using the <i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">galaxy-pencil</span> (pencil) icon. The file should look like below.
      <figure id="figure-6" width="50%"><img src="../../images/limma-voom/annodata.png" alt="annodata file" title="Gene annotation file" /><figcaption><span class="figcaption-prefix">Figure 6:</span> Gene annotation file</figcaption></figure>
    </li>
  </ol>
</blockquote>

<h1 id="differential-expression-with-limma-voom">Differential expression with limma-voom</h1>

<h2 id="filtering-to-remove-lowly-expressed-genes">Filtering to remove lowly expressed genes</h2>

<p>It is recommended to filter for lowly expressed genes when running the limma-voom tool. Genes with very low counts across all samples provide little evidence for differential expression and they interfere with some of the statistical approximations that are used later in the pipeline. They also add to the multiple testing burden when estimating false discovery rates, reducing power to detect differentially expressed genes. These genes should be filtered out prior to further analysis.</p>

<p>There are a few ways to filter out lowly expressed genes. When there are biological replicates in each group, in this case we have a sample size of 2 in each group, we favour filtering on a minimum counts per million threshold present in at least 2 samples. Two represents the smallest sample size for each group in our experiment. In this dataset, we choose to retain genes if they are expressed at a counts-per-million (CPM) above 0.5 in at least two samples. The CPM threshold selected can be compared to the raw count with the CpmPlots (see below).</p>

<blockquote class="details">
  <h3 id="details-more-details-on-filtering"><i class="fa fa-info-circle" aria-hidden="true"></i><span class="visually-hidden">details</span> More details on filtering</h3>

  <p>The limma tool uses the <code class="highlighter-rouge">cpm</code> function from the edgeR package <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2796818/">Robinson, McCarthy, and Smyth 2010</a> to generate the CPM values which can then be filtered. Note that by converting to CPMs we are normalizing for the different sequencing depths for each sample. A CPM of 0.5 is used as it corresponds to a count of 10-15 for the library sizes in this data set. If the count is any smaller, it is considered to be very low, indicating that the associated gene is not expressed in that sample. A requirement for expression in two or more libraries is used as each group contains two replicates. This ensures that a gene will be retained if it is only expressed in one group. Smaller CPM thresholds are usually appropriate for larger libraries. As a general rule, a good threshold can be chosen by identifying the CPM that corresponds to a count of 10, which in this case is about 0.5. You should filter with CPMs rather than filtering on the counts directly, as the latter does not account for differences in library sizes between samples.</p>
</blockquote>

<h2 id="normalization-for-composition-bias">Normalization for composition bias</h2>

<p>In RNA-seq the counts are normalized for different sequencing depths between samples. Normalizing to eliminate composition biases between samples is also typically performed. Composition biases can occur, for example, if there are a few highly expressed genes dominating in some samples, leading to less reads from other genes. By default, TMM normalization <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2864565/">(Robinson and Oshlack 2010)</a> is performed by the limma tool using the edgeR <code class="highlighter-rouge">calcNormFactors</code> function (this can be changed under <strong>Advanced Options</strong>). TMM stands for Trimmed Mean of M values, where a weighted trimmed mean of the log expression ratios is used to scale the counts for the samples. See the figure from the TMM paper below. Note the plot (Figure 1c) that shows how a few highly expressed genes in the liver sample (where the arrow is) results in the majority of other genes in the sample having the appearance of being expressed lower in liver. The mid-line through the points is offset from the expected zero and the TMM normalization factor (red line) scales the counts to adjust for this.</p>

<figure id="figure-7" width="50%"><img src="../../images/limma-voom/TMM.png" alt="TMM normalization" title="TMM normalization (Robinson and Oshlack 2010)" /><figcaption><span class="figcaption-prefix">Figure 7:</span> TMM normalization (Robinson and Oshlack 2010)</figcaption></figure>

<h2 id="specify-contrasts-of-interest">Specify Contrast(s) of interest</h2>

<p>Since we are interested in differences between groups, we need to specify which comparisons we want to test. For example, if we are interested in knowing which genes are differentially expressed between the pregnant and lactating group in the basal cells we specify <code class="highlighter-rouge">basalpregnant-basallactate</code> for the <em>Contrast of Interest</em>. Note that the group names in the contrast must exactly match the names of the groups in the <code class="highlighter-rouge">factordata</code> file. More than one contrast can be specified using the <code class="highlighter-rouge">Insert Contrast</code> button, so we could look at more comparisons of the groups here, but first we’ll take a look at <code class="highlighter-rouge">basalpregnant-basallactate</code>.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-differential-expression-with-limma-voom"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Differential expression with limma-voom</h3>

  <ol>
    <li><strong>limma</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>: Run <strong>limma</strong> with the following settings:
      <ul>
        <li><em>“Differential Expression Method”</em>: <code class="highlighter-rouge">limma-voom</code></li>
        <li><em>“Count Files or Matrix?</em>”: <code class="highlighter-rouge">Single Count Matrix</code>
          <ul>
            <li><em>“Count Matrix”</em>: Select <code class="highlighter-rouge">countdata</code></li>
          </ul>
        </li>
        <li><em>“Input factor information from file?”</em>: <code class="highlighter-rouge">Yes</code>
          <ul>
            <li><em>“Factor File”</em>: Select <code class="highlighter-rouge">factordata</code></li>
          </ul>
        </li>
        <li><em>“Use Gene Annotations?”</em>: <code class="highlighter-rouge">Yes</code>
          <ul>
            <li><em>“Factor File”</em>: Select <code class="highlighter-rouge">annodata</code></li>
          </ul>
        </li>
        <li><em>“Contrast of Interest”</em>: <code class="highlighter-rouge">basalpregnant-basallactate</code></li>
        <li><em>“Filter lowly expressed genes?”</em>: <code class="highlighter-rouge">Yes</code>
          <ul>
            <li><em>“Filter on CPM or Count values?”</em>: <code class="highlighter-rouge">CPM</code></li>
            <li><em>“Minimum CPM”</em>: <code class="highlighter-rouge">0.5</code></li>
            <li><em>“Minimum Samples”</em>: <code class="highlighter-rouge">2</code></li>
          </ul>
        </li>
      </ul>
    </li>
    <li>Inspect the <code class="highlighter-rouge">Report</code> produced by clicking on the <i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">galaxy-eye</span> (eye) icon</li>
  </ol>
</blockquote>

<h1 id="quality-control-of-count-data">Quality Control of count data</h1>

<p>Before we check out the differentially expressed genes, we can look at the <code class="highlighter-rouge">Report</code> information to check that the data is good quality and that the samples are as we would expect.</p>

<h2 id="multidimensional-scaling-plot">Multidimensional scaling plot</h2>

<p>By far, one of the most important plots we make when we analyse RNA-Seq data are MDS plots. An MDS plot is a visualisation of a principal components analysis, which determines the greatest sources of variation in the data. A principal components analysis is an example of an unsupervised analysis, where we don’t need to specify the groups. If your experiment is well controlled and has worked well, what we hope to see is that the greatest sources of variation in the data are the treatments/groups we are interested in. It is also an incredibly useful tool for quality control and checking for outliers. This Galaxy limma tool outputs an MDS plot by default in the <code class="highlighter-rouge">Report</code> and a link is also provided to a PDF version (<code class="highlighter-rouge">MDSPlot_CellTypeStatus.pdf</code>). a scree plot is also produced that shows how much variation is attributed to each dimension. If there was a batch effect for example, you may see high values for additional dimensions. The limma tool plots the first two dimensions by default (1 vs 2), however you can also plot additional dimensions 2 vs 3 and 3 vs 4 using under <strong>Output Options</strong> Additional Plots <code class="highlighter-rouge">MDS Extra</code> These are displayed in the <code class="highlighter-rouge">Report</code> along with a link to a PDF version (<code class="highlighter-rouge">MDSPlot_extra.pdf</code>).</p>

<figure id="figure-8"><img src="../../images/limma-voom/mdsscree.png" alt="MDS Plot" title="MDS Plot" /><figcaption><span class="figcaption-prefix">Figure 8:</span> MDS Plot</figcaption></figure>

<p>Take a look at the MDS plot coloured by group.</p>

<blockquote class="question">
  <h3 id="question-question"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <p>Do you notice anything about the samples in this plot?</p>

  <blockquote class="solution">
    <h3 id="solution-solution"><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>

    <p>Two samples don’t appear to be in the right place.</p>

  </blockquote>
</blockquote>

<p>It turns out that there has been a mix-up with two samples, they have been mislabelled in the sample information file. This shows how the MDS plot can also be useful to help identify if sample mix-ups may have occurred. We need to redo the limma-voom analysis with the correct sample information.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-use-the-rerun-button-to-redo-steps"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Use the Rerun button to redo steps</h3>

  <ol>
    <li>Import the correct sample information file from <code class="highlighter-rouge">https://ndownloader.figshare.com/files/5999832?private_link=1d788fd384d33e913a2a</code></li>
    <li>Use the Rerun button in the History to redo the <strong>Merge Columns</strong> and <strong>Cut</strong> steps on the correct sample information file.</li>
    <li>Delete the incorrect sample information datasets to avoid any confusion.</li>
    <li>Rerun <strong>limma</strong> as before with the correct <code class="highlighter-rouge">sampleinfo</code> file and adding the following settings:
      <ul>
        <li><strong>Output Options</strong>
          <ul>
            <li><em>“Additional Plots”</em> tick:
              <ul>
                <li><code class="highlighter-rouge">Density Plots (if filtering)</code></li>
                <li><code class="highlighter-rouge">CpmsVsCounts Plots (if filtering on cpms)</code></li>
                <li><code class="highlighter-rouge">Box Plots (if normalising)</code></li>
                <li><code class="highlighter-rouge">MDS Extra (Dims 2vs3 and 3vs4)</code></li>
                <li><code class="highlighter-rouge">MD Plots for individual samples</code></li>
                <li><code class="highlighter-rouge">Heatmaps (top DE genes)</code></li>
                <li><code class="highlighter-rouge">Stripcharts (top DE genes)</code></li>
              </ul>
            </li>
            <li><em>“Output Library information file?”</em>: <code class="highlighter-rouge">Yes</code></li>
          </ul>
        </li>
      </ul>
    </li>
  </ol>
</blockquote>

<p>In the <code class="highlighter-rouge">Report</code> you should then see the correct MDS plot as below.</p>

<figure id="figure-9"><img src="../../images/limma-voom/mdsscree_corr.png" alt="MDS Plot" title="MDS Plot (correct samples)" /><figcaption><span class="figcaption-prefix">Figure 9:</span> MDS Plot (correct samples)</figcaption></figure>

<blockquote class="details">
  <h3 id="details-more-details-on-mds-plots"><i class="fa fa-info-circle" aria-hidden="true"></i><span class="visually-hidden">details</span> More details on MDS plots</h3>

  <p>The distance between each pair of samples in the MDS plot is calculated as the leading fold change, defined as the root-mean-square of the largest 500 log2-fold changes between that pair of samples. Replicate samples from the same group cluster together in the plot, while samples from different groups form separate clusters. This indicates that the differences between groups are larger than those within groups, i.e., differential expression is greater than the variance and can be detected. In the MDS plot, the distance between basal samples on the left and luminal cells on the right is about 6 units, corresponding to a leading fold change of about 64-fold (2^6 = 64) between basal and luminal. The expression differences between virgin, pregnant and lactating are greater for luminal cells than for basal.</p>

  <p>Clustering in the MDS plot can be used to motivate changes to the analysis in light of potential batch effects. For example, imagine that the first replicate of each group was prepared at a separate time from the second replicate. If the MDS plot showed separation of samples by time, it might be worthwhile including time in the downstream analysis to account for the time-based effect.</p>
</blockquote>

<blockquote class="question">
  <h3 id="question-question-1"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <p>What is the greatest source of variation in the data (i.e. what does dimension 1 represent)?
What is the second greatest source of variation in the data?</p>

  <blockquote class="solution">
    <h3 id="solution-solution-1"><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>

    <p>Dimension 1 represents the variation due to cell type, basal vs luminal. Dimension 2 represents the variation due to the stages, virgin, pregnant or lactating.</p>

  </blockquote>
</blockquote>

<p>Next, scroll down the <code class="highlighter-rouge">Report</code> to take a look at the <strong>Additional information</strong> and <strong>Summary of experimental data</strong> sections near the bottom. It should look similar to below. Here you can check that the correct samples have been assigned to the correct groups, what settings were used (e.g. filters, normalization method) and also how many genes were filtered out due to low expression.</p>

<p><img src="../../images/limma-voom/report_summary.png" alt="Report summary" width="600px" /></p>

<blockquote class="question">
  <h3 id="question-question-2"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <p>How many genes have been filtered out for low expression?</p>

  <blockquote class="solution">
    <h3 id="solution-solution-2"><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>

    <p>11375 genes were filtered out as insignificant as they were without more than 0.5 CPM in at least 2 samples.</p>

  </blockquote>
</blockquote>

<blockquote>
  <blockquote class="tip">
    <h3 id="tip-tip"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip</h3>

    <ul>
      <li>A threshold of 1 CPM in at least minimum group sample size is a good rule of thumb for samples with about 10 million reads. For larger library sizes increase the CPM theshold and for smaller library sizes decrease it. Check the CpmsVsCounts plots to see if the selected threshold looks appropriate for the samples (equivalent to ~10 reads).</li>
    </ul>

  </blockquote>

</blockquote>

<h2 id="density-plots">Density plots</h2>

<p>Density plots can be output in the <code class="highlighter-rouge">Report</code> if <em>Filter lowly expressed genes</em> is applied. A link is also provided in the <code class="highlighter-rouge">Report</code> to a PDF version (<code class="highlighter-rouge">DensityPlots.pdf</code>). These plots allow comparison of the counts distributions before and after filtering. The samples are coloured by the groups. Count data is not normally distributed, so if we want to examine the distributions of the raw counts we need to log the counts. We typically check the distribution of the read counts on the log2 scale. A CPM value of 1 is equivalent to a log-CPM value of 0 and the CPM we used of 0.5 is equivalent to a log-CPM of -1. It can be seen in the before filtering plot below that a large proportion of genes within each sample are not expressed or lowly-expressed and our filter of CPM of 0.5 (in at least 2 samples) removes a lot of these uninformative genes.</p>

<figure id="figure-10"><img src="../../images/limma-voom/densityplots.png" alt="Density Plots" title="Density Plots" /><figcaption><span class="figcaption-prefix">Figure 10:</span> Density Plots</figcaption></figure>

<p>We can also have a look more closely to see whether our threshold of 0.5 CPM does indeed correspond to a count of about 10-15 reads in each sample with the plots of CPM versus raw counts.</p>

<p>The <code class="highlighter-rouge">Report</code> provides links to PDFs of all plots shown in the <code class="highlighter-rouge">Report</code> and also to the rest of the additional plots selected to be output.</p>

<figure id="figure-11"><img src="../../images/limma-voom/report_plots.png" alt="Report Outputs" title="Report outputs" /><figcaption><span class="figcaption-prefix">Figure 11:</span> Report outputs</figcaption></figure>

<p>Click on the <code class="highlighter-rouge">CpmPlots.pdf</code> link in the <code class="highlighter-rouge">Report</code>. You should see 12 plots, one for each sample. Two of the plots are shown below. From these plots we can see that 0.5 CPM is equivalent to ~10 counts in each of the 12 samples, so 0.5 seems to be an appropriate threshold for this dataset (these samples all have sequencing depth of 20-30 million (see the <code class="highlighter-rouge">Library information</code> file below) so a CPM value of 0.5 would be ~10 counts).</p>

<figure id="figure-12"><img src="../../images/limma-voom/cpmsvscounts.png" alt="CPM threshold Plots" title="CPM vs Raw Counts Plots" /><figcaption><span class="figcaption-prefix">Figure 12:</span> CPM vs Raw Counts Plots</figcaption></figure>

<blockquote class="tip">
  <h3 id="tip-tip-1"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip</h3>

  <ul>
    <li>A threshold of 1 CPM in at least minimum group sample size is a good rule of thumb for samples with about 10 million reads. For larger library sizes increase the CPM theshold and for smaller library sizes decrease it. Check the CpmPlots to see if the selected threshold looks appropriate for the samples (equivalent to ~10 reads).</li>
  </ul>

</blockquote>

<h2 id="box-plots">Box plots</h2>

<p>We can also use box plots to check the distributions of counts in the samples. Box plots can be selected to be output by the Galaxy limma-voom tool if normalization is applied (TMM is applied by default). The plots are output in the <code class="highlighter-rouge">Report</code> and a link is also provided to a PDF version (<code class="highlighter-rouge">BoxPlots.pdf</code>). With the box plots for these samples we can see that overall the distributions are not identical but still not very different. If a sample is really far above or below the blue horizontal line we may need to investigate that sample further.</p>

<figure id="figure-13"><img src="../../images/limma-voom/boxplots.png" alt="Box Plots" title="Box Plots" /><figcaption><span class="figcaption-prefix">Figure 13:</span> Box Plots</figcaption></figure>

<blockquote class="question">
  <h3 id="question-question-3"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <p>Compare the box plots before and after TMM normalisation. Can you see any differences?</p>

  <blockquote class="solution">
    <h3 id="solution-solution-3"><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>

    <p>After the normalization more of the samples are closer to the median horizontal line.</p>

  </blockquote>
</blockquote>

<p>The TMM normalization generates normalization factors, where the product of these factors and the library sizes defines the effective library size. TMM normalization (and most scaling normalization methods) scale relative to one sample. The normalization factors multiply to unity across all libraries. A normalization factor below one indicates that the library size will be scaled down, as there is more suppression (i.e., composition bias) in that library relative to the other libraries. This is also equivalent to scaling the counts upwards in that sample. Conversely, a factor above one scales up the library size and is equivalent to downscaling the counts. We can see the normalization factors for these samples in the <code class="highlighter-rouge">Library information</code> file that we selected to output. Click on the <i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">galaxy-eye</span> (eye) icon to view.</p>

<figure id="figure-14"><img src="../../images/limma-voom/libinfo.png" alt="Library Info file" title="Library information file" /><figcaption><span class="figcaption-prefix">Figure 14:</span> Library information file</figcaption></figure>

<blockquote class="question">
  <h3 id="question-question-4"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <p>Which sample has the largest normalization factor? Which sample has the smallest?</p>

  <blockquote class="solution">
    <h3 id="solution-solution-4"><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>

    <p>MCL1.LA has the largest normalization factor and MCL1.LE the smallest.</p>

  </blockquote>
</blockquote>

<h2 id="md-plots-for-samples">MD plots for samples</h2>

<p>It is considered good practice to make mean-difference (MD) plots for all the samples as a quality check, as described in this <a href="https://f1000research.com/articles/5-1438/v2">edgeR workflow article</a>. These plots allow expression profiles of individual samples to be explored more closely. An MD plot shows the log-fold change between a sample against the average expression across all the other samples. This visualisation can help you see if there are genes highly upregulated or downregulated in a sample. If we look at mean difference plots for these samples, we should be able to see the composition bias problem. The mean-difference plots show average expression (mean: x-axis) against log-fold-changes (difference: y-axis).</p>

<p>Click on the <code class="highlighter-rouge">MDPlots_Samples.pdf</code> link in the <code class="highlighter-rouge">Report</code>. You should see 12 MD plots, one for each sample. Let’s take a look at the plots for the two samples MCL1.LA and MCL1.LE that had the largest and smallest normalization factors. The MD plots on the left below show the counts normalized for library size and the plots on the right show the counts after the TMM normalization has been applied. MCL1.LA had the largest normalization factor and was above the median line in the unnormalized by TMM box plots. MCL1.LE had the smallest normalization factor and was below the median line in the box plots. These MD plots help show the composition bias problem has been addressed.</p>
<figure id="figure-15"><img src="../../images/limma-voom/mdsampleLA.png" alt="MD Plot LA" title="MD Plots for MCL1.LA before and after TMM normalization" /><figcaption><span class="figcaption-prefix">Figure 15:</span> MD Plots for MCL1.LA before and after TMM normalization</figcaption></figure>
<figure id="figure-16"><img src="../../images/limma-voom/mdsampleLE.png" alt="MD Plot LE" title="MD Plots for MCL1.LE before and after TMM normalization" /><figcaption><span class="figcaption-prefix">Figure 16:</span> MD Plots for MCL1.LE before and after TMM normalization</figcaption></figure>

<h2 id="voom-variance-plot">Voom variance plot</h2>

<p>This plot is generated by the voom method and displayed in the <code class="highlighter-rouge">Report</code> along with a link to a PDF version (<code class="highlighter-rouge">VoomPlot.pdf</code>). It shows the mean-variance relationship of the genes in the dataset. It can help show if low counts have been filtered adequately and if there is a lot of variation in our data.</p>

<figure id="figure-17"><img src="../../images/limma-voom/voomplot.png" alt="Voom Plot" title="Voom Plot" /><figcaption><span class="figcaption-prefix">Figure 17:</span> Voom Plot</figcaption></figure>

<blockquote class="details">
  <h3 id="details-more-details-on-voom-variance-plots"><i class="fa fa-info-circle" aria-hidden="true"></i><span class="visually-hidden">details</span> More details on Voom variance plots</h3>

  <p>If we didn’t filter this dataset for the lowly expressed genes the variance plot would look like below.</p>

  <figure id="figure-18"><img src="../../images/limma-voom/voomplot_nofilt.png" alt="Voom Nofilter Plot" title="Voom Plot unfiltered counts" /><figcaption><span class="figcaption-prefix">Figure 18:</span> Voom Plot unfiltered counts</figcaption></figure>

  <p>If we look at the plot generated with the two samples mixed up we can see there’s more variation.</p>

  <figure id="figure-19"><img src="../../images/limma-voom/voomplot_mixed.png" alt="Voom Mixedup Plot" title="Voom Plot mixed-up samples" /><figcaption><span class="figcaption-prefix">Figure 19:</span> Voom Plot mixed-up samples</figcaption></figure>

  <p>More examples of the variation this plot can show can be seen in Figure 1 from the <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29">limma-voom</a> paper, shown below.</p>

  <figure id="figure-20"><img src="../../images/limma-voom/voom_variance_examples.jpg" alt="Voom Plot Examples" title="Voom Plot more examples" /><figcaption><span class="figcaption-prefix">Figure 20:</span> Voom Plot more examples</figcaption></figure>

  <p><em>Figure 1: Mean-variance relationships. Gene-wise means and variances of RNA-seq data are represented by black points with a LOWESS trend. Plots are ordered by increasing levels of biological variation in datasets. (a) voom trend for HBRR and UHRR genes for Samples A, B, C and D of the SEQC project; technical variation only. (b) C57BL/6J and DBA mouse experiment; low-level biological variation. (c) Simulation study in the presence of 100 upregulating genes and 100 downregulating genes; moderate-level biological variation. (d) Nigerian lymphoblastoid cell lines; high-level biological variation. (e) Drosophila melanogaster embryonic developmental stages; very high biological variation due to systematic differences between samples. (f) LOWESS voom trends for datasets (a)–(e). HBRR, Ambion’s Human Brain Reference RNA; LOWESS, locally weighted regression; UHRR, Stratagene’s Universal Human Reference RNA</em>.</p>
</blockquote>

<h2 id="md-and-volcano-plots-for-de-results">MD and Volcano plots for DE results</h2>

<p>Genome-wide plots that are useful for checking differentially expressed (DE) results are MD plots (or MA plots) and Volcano plots. There are functions in limma for generating these plots and they are used by this tool. These plots are output by default and shown in the <code class="highlighter-rouge">Report</code> along with a link to PDF versions (<code class="highlighter-rouge">MDPlot_basalpregnant-basallactate.pdf</code> and <code class="highlighter-rouge">VolcanoPlot_basalpregnant-basallactate.pdf</code>). In the volcano plot the top genes (by adjusted p-value) are highlighted. The number of top genes is 10 by default and the user can specify the number of top genes to view (up to 100) under <strong>Advanced Options</strong>.</p>

<figure id="figure-21"><img src="../../images/limma-voom/mdvolplot_basalpregnant-basallactate.png" alt="MDVol Plot" title="MD Plot and Volcano Plot" /><figcaption><span class="figcaption-prefix">Figure 21:</span> MD Plot and Volcano Plot</figcaption></figure>

<p>The MD Plot highlighted genes are significant at an adjusted p-value (adj.P) threshold of 0.05 and exhibit log2-fold-change (lfc) of at least 0. These thresholds can be changed under <strong>Advanced Options</strong>.</p>

<blockquote class="question">
  <h3 id="question-question-5"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <p>How many genes are differentially expressed at the default thresholds of adj.P=0.05 and lfc=0?</p>

  <blockquote class="solution">
    <h3 id="solution-solution-5"><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>

    <p>The number of DE genes at these adj.P and lfc thresholds is shown in the table in the <code class="highlighter-rouge">Report</code> as below.</p>

    <p><img src="../../images/limma-voom/decounts.png" alt="DE counts" width="500px" /></p>

  </blockquote>
</blockquote>

<blockquote class="comment">
  <h3 id="comment-comment-on-adjusted-p-value"><i class="fa fa-commenting-o" aria-hidden="true"></i><span class="visually-hidden">comment</span> Comment on adjusted P value</h3>

  <p><strong>A note about deciding how many genes are significant</strong>: In order to decide which genes are differentially expressed, we usually take a cut-off (e.g. 0.05 or 0.01) on the adjusted p-value, NOT the raw p-value. This is because we are testing many genes (more than 15000 genes here), and the chances of finding differentially expressed genes is very high when you do that many tests. Hence we need to control the false discovery rate, which is the adjusted p-value column in the results table. What this means is that, if we choose an adjusted p-value cut-off of 0.05, and if 100 genes are significant at a 5% false discovery rate, we are willing to accept that 5 will be false positives.</p>
</blockquote>

<h1 id="testing-relative-to-a-threshold-treat">Testing relative to a threshold (TREAT)</h1>

<p>When there is a lot of differential expression, sometimes we may want to cut-off on a fold change threshold as well as a p-value threshold so that we follow up on the most biologically significant genes. However, it is not recommended to simply rank by p-value and then discard genes with small logFC’s, as this has been shown to increase the false discovery rate. In other words, you are not controlling the false discovery rate at 5% any more. There is a function called <code class="highlighter-rouge">treat</code> in limma that performs this style of analysis correctly <a href="https://www.ncbi.nlm.nih.gov/pubmed/19176553">(McCarthy and Smyth 2009)</a>. TREAT will simply take a user-specified log fold change cut-off and recalculate the moderated t-statistics and p-values with the new information about logFC. There are thousands of genes differentially expressed in this <code class="highlighter-rouge">basalpregnant-basallactate</code> comparison, so let’s rerun the analysis applying TREAT and similar thresholds to what was used in the Fu paper: an adjusted P value of 0.01 (1% false discovery rate) and a log-fold-change cutoff of 0.58 (equivalent to a fold change of 1.5).</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-testing-relative-to-a-threshold-treat"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Testing relative to a threshold (TREAT)</h3>

  <ol>
    <li><strong>limma</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>: Rerun <strong>limma</strong> with the following settings:
      <ul>
        <li><em>“<strong>Output Options</strong>”</em>
          <ul>
            <li>“Output Library information file?”: <code class="highlighter-rouge">No</code></li>
          </ul>
        </li>
        <li><em>“<strong>Advanced Options</strong>”</em>
          <ul>
            <li><em>“Minimum Log2 Fold Change”</em>: <code class="highlighter-rouge">0.58</code></li>
            <li><em>“P-Value Adjusted Threshold”</em>: <code class="highlighter-rouge">0.01</code></li>
            <li><em>“Test significance relative to a fold-change threshold (TREAT)”</em>: <code class="highlighter-rouge">Yes</code></li>
          </ul>
        </li>
      </ul>
    </li>
    <li>Inspect the <code class="highlighter-rouge">Report</code></li>
  </ol>
</blockquote>

<p>We can see that much fewer genes are now highlighted in the MD plot and identified as differentially expressed.</p>

<figure id="figure-22"><img src="../../images/limma-voom/TREAT_mdvolplot_basalpregnant-basallactate.png" alt="TREAT MDVol Plot" title="TREAT MD and Volcano plots" /><figcaption><span class="figcaption-prefix">Figure 22:</span> TREAT MD and Volcano plots</figcaption></figure>

<figure id="figure-23" width="400px"><img src="../../images/limma-voom/TREATdecounts.png" alt="TREAT DE counts" title="TREAT differentially expressed genes" /><figcaption><span class="figcaption-prefix">Figure 23:</span> TREAT differentially expressed genes</figcaption></figure>

<h1 id="visualising-results">Visualising results</h1>

<p>Before following up on the DE genes with further lab work, it is recommended to have a look at the expression levels of the individual samples for the genes of interest. The Galaxy limma tool can auto-generate heatmaps of the top genes to show the expression levels across the <em>samples</em>. It can also generate stripcharts to view the expression of the top genes across the <em>groups</em>. This enables a quick view of the expression of the top differentially expressed genes. These types of plots can also help show if expression is consistent between replicates in the groups.</p>

<h2 id="heatmap-of-top-genes">Heatmap of top genes</h2>

<p>Click on the <code class="highlighter-rouge">Heatmap_basalpregnant-basallactate.pdf</code> link in the <code class="highlighter-rouge">Report</code>. You should see a plot like below.</p>

<figure id="figure-24"><img src="../../images/limma-voom/heatmap.png" alt="Heatmap" title="Heatmap of top genes" /><figcaption><span class="figcaption-prefix">Figure 24:</span> Heatmap of top genes</figcaption></figure>

<h2 id="stripcharts-of-top-genes">Stripcharts of top genes</h2>

<p>Click on the <code class="highlighter-rouge">Stripcharts_basalpregnant-basallactate.pdf</code> link in the <code class="highlighter-rouge">Report</code>. You should see 10 plots, one for each top gene. Four are shown below.</p>

<figure id="figure-25" width="950px"><img src="../../images/limma-voom/stripcharts.png" alt="Stripchart Plot" title="Stripcharts of top genes" /><figcaption><span class="figcaption-prefix">Figure 25:</span> Stripcharts of top genes</figcaption></figure>

<h2 id="interactive-plots-glimma">Interactive plots (Glimma)</h2>

<p>An interactive version of the mean-difference plots is output via the <a href="https://github.com/Shians/Glimma">Glimma</a> package if a gene annotation file is provided to the Galaxy limma tool. A link to a html page is generated in the <code class="highlighter-rouge">Report</code> that allows the user to search for their favourite gene.</p>

<p>Click on the <code class="highlighter-rouge">Glimma_MDPlot_basalpregnant-basallactate.html</code> link in the <code class="highlighter-rouge">Report</code>. You should see a two-panel interactive MD plot like below. The left plot shows the log-fold-change vs average expression. The right plot shows the expression levels of a particular gene of each sample by groups (similar to the stripcharts). Hovering over points on left plot will plot expression level for corresponding gene, clicking on points will fix the expression plot to gene. Clicking on rows on the table has the same effect as clicking on the corresponding gene in the plot.</p>

<iframe src="../../images/limma-voom/glimma/MD-Plot.html" width="100%" height="1000"></iframe>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-search-for-a-gene-of-interest"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Search for a gene of interest</h3>

  <p><code class="highlighter-rouge">Egf</code> was a gene identifed as very highly expressed in the Fu paper and confirmed with qRT-PCR, see Fig. 6c from the paper below.</p>
  <figure id="figure-26"><img src="../../images/limma-voom/fu_egf.png" alt="Fu EGF gene" title="Fu et al, Nat Cell Biol 2015" /><figcaption><span class="figcaption-prefix">Figure 26:</span> Fu et al, Nat Cell Biol 2015</figcaption></figure>

  <p>Search for <code class="highlighter-rouge">Egf</code> in the Glimma interactive table. You should see something similar to below.</p>

  <figure id="figure-27"><img src="../../images/limma-voom/glimma_egf.png" alt="Glimma EGF gene" title="Glimma EGF gene" /><figcaption><span class="figcaption-prefix">Figure 27:</span> Glimma EGF gene</figcaption></figure>

  <p><code class="highlighter-rouge">Egf</code> is not identified as differentially expressed in the basal cells (see the Adj.PValue) but in the plot showing the Expression levels of all samples, notice that we can see it is highly expressed in the luminallactate group.</p>
</blockquote>

<p>Multiple contrasts can be run with the limma tool. For example, we can compare the pregnant and lactating conditions for both the basal and luminal cells. So let’s rerun the limma-voom TREAT analysis (adj.P &lt;0.01 and lfc=0.58) and this time use the <code class="highlighter-rouge">Insert Contrast</code> button to include the additional contrast <code class="highlighter-rouge">luminalpregnant - luminallactate</code>. We can then see how the number of differentially expressed genes in the luminal cells compares to the basal cells.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-run-multiple-contrasts"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Run multiple contrasts</h3>

  <ol>
    <li><strong>limma</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>: Rerun <strong>limma</strong> adding the following settings <em>(i.e. run with 2 contrasts)</em>:
      <ul>
        <li><em>“Contrast of Interest”</em>: <code class="highlighter-rouge">basalpregnant-basallactate</code></li>
        <li><em>“Contrast of Interest”</em>: <code class="highlighter-rouge">luminalpregnant-luminallactate</code></li>
        <li><em>“<strong>Advanced Options</strong>”</em>
          <ul>
            <li><em>“Minimum Log2 Fold Change”</em>: <code class="highlighter-rouge">0.58</code></li>
            <li><em>“P-Value Adjusted Threshold”</em>: <code class="highlighter-rouge">0.01</code></li>
            <li><em>“Test significance relative to a fold-change threshold (TREAT)”</em>: <code class="highlighter-rouge">Yes</code></li>
          </ul>
        </li>
      </ul>
    </li>
    <li>Inspect the <code class="highlighter-rouge">Report</code></li>
  </ol>

  <p>You should find that there are more genes differentially expressed for the luminal cells than the basal. There are ~274 genes DE in basal cells versus ~ 1610 in the luminal cells.</p>
  <figure id="figure-28" width="50%"><img src="../../images/limma-voom/basal_luminal_decounts.png" alt="Basal Luminal DE counts" title="Basal vs Luminal DE counts" /><figcaption><span class="figcaption-prefix">Figure 28:</span> Basal vs Luminal DE counts</figcaption></figure>

  <p>This is similar to what Fu et al found, many more genes differentially expressed in the luminal cells on lactation, compared to the basal cells.</p>
  <figure id="figure-29" width="25%"><img src="../../images/limma-voom/fu_degenes.png" alt="Fu DE genes" title="Fu et al, Nat Cell Biol 2015" /><figcaption><span class="figcaption-prefix">Figure 29:</span> Fu et al, Nat Cell Biol 2015</figcaption></figure>
</blockquote>

<p>Note that the tables of differentially expressed genes are output as links in the <code class="highlighter-rouge">Report</code> (<code class="highlighter-rouge">limma-voom_basalpregnant-basallactate.tsv</code> and <code class="highlighter-rouge">limma-voom_luminalpregnant-luminallactate.tsv</code>), see below, and also in the history (<code class="highlighter-rouge">DE tables</code>). With multiple contrasts, a plot for each contrast is generated for relevant plots, as shown below.</p>

<figure id="figure-30"><img src="../../images/limma-voom/multiple_contrasts.png" alt="Multiple contrasts" title="Multiple contrasts output" /><figcaption><span class="figcaption-prefix">Figure 30:</span> Multiple contrasts output</figcaption></figure>

<blockquote class="tip">
  <h3 id="tip-tip-2"><i class="fa fa-lightbulb-o" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip</h3>

  <p>The <code class="highlighter-rouge">Report</code> with all the plots and tables can be downloaded by clicking on the floppy disk icon on the dataset in the history as shown below.</p>
  <figure id="figure-31" width="20%"><img src="../../images/limma-voom/download_report.png" alt="Report download" title="Download limma report" /><figcaption><span class="figcaption-prefix">Figure 31:</span> Download limma report</figcaption></figure>

</blockquote>

<h2 id="create-heatmap-of-custom-genes">Create heatmap of custom genes</h2>

<p>You may want to create a heatmap for a set of genes of interest, such as the 31 genes from the original paper using this dataset, Fig. 6b below. These 31 genes include the authors’ main gene of interest in the paper, Mcl1, and a set of cytokines/growth factors, identified as differentially expressed in the luminal pregnant vs lactating cells by the authors. We will recreate the heatmap to show the steps. It also serves as a sanity check, to check that our results look similar to what was shown in the paper. For this we need the normalized counts which we can get from the limma tool.</p>

<figure id="figure-32" width="50%"><img src="../../images/limma-voom/fu_heatmap.png" alt="Fu heatmap" title="Fu et al, Nat Cell Biol 2015" /><figcaption><span class="figcaption-prefix">Figure 32:</span> Fu et al, Nat Cell Biol 2015</figcaption></figure>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>GeneID
Mcl1
Hbegf
Tgfb2
Cxcl16
Csf1
Pdgfb
Edn1
Lif
Kitl
Bmp1
Pdgfa
Cmtm3
Cx3cl1
Ctgf
Wnt5a
Ptn
Spp1
Bmp3
Cmtm8
Gmfg
Cxcl2
Cxcl3
Il15
Egf
Cmtm7
Il34
Pdgfd
Nov
Cmtm6
Ccl28
Cxcl1
</code></pre></div></div>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-extract-the-normalized-counts-for-the-genes-of-interest"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Extract the normalized counts for the genes of interest</h3>
  <ol>
    <li>Rerun <strong>limma</strong> selecting <em>“Output Normalized Counts Table?”</em>: <code class="highlighter-rouge">Yes</code></li>
    <li>Create a file of the gene symbols of interest
      <ul>
        <li>Paste the information above (the 31 gene symbols and header) into the Galaxy Data Uploader Paste/Fetch box</li>
        <li>Set File Type to <code class="highlighter-rouge">tabular</code></li>
        <li>Use the <i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">galaxy-pencil</span> (pencil) icon to rename the file to <code class="highlighter-rouge">heatmap genes</code></li>
      </ul>
    </li>
    <li><strong>Join two Datasets</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with the following parameters:
      <ul>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“Join”</em>: the <code class="highlighter-rouge">heatmap genes</code> file</li>
        <li><em>“using column”</em>: <code class="highlighter-rouge">Column: 1</code></li>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“with”</em>: <code class="highlighter-rouge">Normalised counts</code> file (output of <strong>limma</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>)</li>
        <li><em>“and column”</em>: <code class="highlighter-rouge">Column: 2</code></li>
        <li><em>“Keep lines of first input that do not join with second input”</em>: <code class="highlighter-rouge">No</code></li>
        <li><em>“Keep the header lines”</em>: <code class="highlighter-rouge">Yes</code></li>
      </ul>

      <p>The generated file has more columns than we need for the heatmap. In addition to the columns with normalized counts (in log2), there is the <script type="math/tex">log_{2} FC</script> and other information. We need to remove the extra columns.</p>
    </li>
    <li><strong>Cut columns from a table</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> to extract the columns with the gene ids and normalized counts
      <ul>
        <li><em>“Cut columns”</em>: <code class="highlighter-rouge">c1,c5-c16</code></li>
        <li><em>“Delimited by”</em>: <code class="highlighter-rouge">Tab</code></li>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“From”</em>: the joined dataset (output of <strong>Join two Datasets</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>)</li>
      </ul>

      <p>The genes are in rows and the samples in columns, we could leave the genes in rows but we will transpose to have genes in columns and samples in rows as in the Figure in the paper.</p>
    </li>
    <li><strong>Transpose</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> to have samples in rows and genes in columns
      <ul>
        <li><em>“Input tabular dataset”</em>:
          <ul>
            <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“From”</em>: the <code class="highlighter-rouge">Cut</code> dataset (output of <strong>Cut</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>)</li>
          </ul>
        </li>
      </ul>
    </li>
  </ol>
</blockquote>

<p>We now have a table with the 31 genes in columns and the 12 samples in rows.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-plot-the-heatmap-of-custom-genes"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Plot the heatmap of custom genes</h3>

  <ol>
    <li><strong>heatmap2</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> to plot the heatmap
      <ul>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“Input should have column headers”</em>: the generated table (output of <strong>Transpose</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>)</li>
        <li><em>“Data transformation”</em>: <code class="highlighter-rouge">Plot the data as it is</code></li>
        <li><em>“Enable data clustering”</em>: <code class="highlighter-rouge">No</code></li>
        <li><em>“Labeling columns and rows”</em>: <code class="highlighter-rouge">Label my columns and rows</code></li>
        <li><em>“Coloring groups”</em>: <code class="highlighter-rouge">Blue to white to red</code></li>
        <li><em>“Data scaling”</em>: <code class="highlighter-rouge">Scale my data by column</code> (scale genes)</li>
      </ul>
    </li>
  </ol>
</blockquote>

<p>You should see a heatmap like below.</p>

<figure id="figure-33" width="30%"><img src="../../images/limma-voom/fu_heatmap_regenerated.png" alt="Fu heatmap regenerated" title="Fu heatmap regenerated" /><figcaption><span class="figcaption-prefix">Figure 33:</span> Fu heatmap regenerated</figcaption></figure>

<blockquote class="question">
  <h3 id="question-question-6"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Question</h3>

  <p>How does the heatmap compare to the one from the Fu paper Fig 6 (above)?</p>

  <blockquote class="solution">
    <h3 id="solution-solution-6"><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>

    <p>The heatmap looks similar to the heatmap in the paper, which is reassuring.</p>

  </blockquote>
</blockquote>

<h1 id="gene-set-testing">Gene Set Testing</h1>

<p>We have identified genes that are differentially expressed. However, sometimes there is quite a long list of differentially expressed genes to interpret after a differential expression analysis, and it is usually infeasible to go through the list one gene at a time trying to understand it’s biological function. A common downstream procedure is gene set testing, which aims to understand which pathways/gene networks the differentially expressed genes are implicated in. There are many different gene set testing methods that can be applied and it can be useful to try several.</p>

<h2 id="gene-ontology-testing-with-goseq">Gene Ontology testing with <strong>goseq</strong></h2>

<p>We would like to know if there are biological categories that are enriched among the differentially expressed genes. To do this we will perform a Gene Ontology analysis, similar to the <a href="/training-material/topics/transcriptomics/tutorials/ref-based/tutorial.html">RNA-seq ref-based tutorial</a>.</p>

<p><a href="http://www.geneontology.org/">Gene Ontology (GO)</a> analysis is widely used to reduce complexity and highlight biological processes in genome-wide expression studies. However, standard methods give biased results on RNA-seq data due to over-detection of differential expression for long and highly-expressed transcripts.</p>

<p><a href="https://bioconductor.org/packages/release/bioc/vignettes/goseq/inst/doc/goseq.pdf">goseq tool</a> provides methods for performing GO analysis of RNA-seq data, taking length bias into account. The methods and software used by goseq are equally applicable to other category based tests of RNA-seq data, such as KEGG pathway analysis.</p>

<p>goseq needs 2 files as inputs:</p>
<ul>
  <li>a <strong>differentially expressed genes</strong> file. Information for all genes tested for differential expression (all genes after filtering lowly expressed). This file should have 2 columns:
    <ul>
      <li>the Gene IDs (unique within the file)</li>
      <li>True (differentially expressed) or False (not differentially expressed)</li>
    </ul>
  </li>
  <li>a <strong>gene lengths</strong> file. Information to correct for potential length bias in differentially expressed genes. This file should have 2 columns:
    <ul>
      <li>the Gene IDs (unique within the file)</li>
      <li>the gene lengths</li>
    </ul>
  </li>
</ul>

<p>We will use the tables of differentially expressed results output from the limma-voom tool, for both the basal and luminal constrasts, and call genes differentially expressed if they have an adjusted P value below 0.01 and a fold change of 1.5 (equivalent to a <script type="math/tex">log_{2} FC</script> of 0.58), as in the Fu paper. We can use the gene lengths from the original table we imported from GEO (<code class="highlighter-rouge">seqdata</code>). But if we didn’t have that we could use a tool like <strong>featureCounts</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> to output a gene lengths file. The original file with gene lengths contains all &gt;20k genes, but we only want the ~15k we have in our differentially expressed genes file after filtering low counts, and in the same order. So we will join the lengths file with the differentially expressed genes file, keeping only the lengths information for genes present in the differentially expressed genes file. We can then cut out the columns we need for the two inputs (gene id, length) (gene id, DE status) and as a bonus they will both be sorted in the same order, what we need for goseq.</p>

<p>To generate the two input files we will use:</p>
<ul>
  <li><strong>Compute</strong> to add a column to the limma-voom table that gives genes meeting our adj.P and lfc thresholds the value “True” and all other genes the value “False”. We want genes that have a lfc &lt; -0.58 (downregulated) and lfc &gt; 0.58 (upregulated). We could use separate filters (e.g. <code class="highlighter-rouge">bool(c4&lt;-0.58) or bool(c4&gt;0.58)</code>) or more simply, we can use the absolute (abs) value 0.58, where the minus sign is ignored.</li>
  <li><strong>Join two Datasets</strong> to add the gene lengths information to the differentially expressed genes, matching on gene ids</li>
  <li><strong>Cut</strong> to extract the two columns for the differentially expressed genes information</li>
  <li><strong>Cut</strong> to extract the two columns for the gene lengths information</li>
</ul>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-prepare-the-two-inputs-for-goseq"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Prepare the two inputs for GOSeq</h3>

  <ol>
    <li><strong>Compute</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with
      <ul>
        <li><em>“Add expression”</em>: <code class="highlighter-rouge">bool(c8&lt;0.01) and bool(abs(c4)&gt;0.58)</code> (adj.P &lt; 0.01 and lfc of 0.58)</li>
        <li><i class="fa fa-folder-o" aria-hidden="true"></i><span class="visually-hidden">param-collection</span> <em>“as a new column to”</em>: the <code class="highlighter-rouge">DE tables</code> output of <strong>limma</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> (containing both the basal and luminal contrasts)</li>
      </ul>
    </li>
    <li><strong>Join two Datasets</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with
      <ul>
        <li><em>“Join”</em>: output of <strong>Compute</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span></li>
        <li><em>“using column”</em>: <code class="highlighter-rouge">Column: 1</code></li>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“with”</em> the original GEO counts file <code class="highlighter-rouge">seqdata</code></li>
        <li><em>“and column”</em>: <code class="highlighter-rouge">Column: 1</code></li>
        <li><em>“Keep lines of first input that do not join with second input”</em>: <code class="highlighter-rouge">No</code></li>
        <li><em>“Keep lines of first input that are incomplete”</em>: <code class="highlighter-rouge">No</code></li>
        <li><em>“Fill empty columns”</em>: <code class="highlighter-rouge">No</code></li>
        <li><em>“Keep the header lines”</em>: <code class="highlighter-rouge">No</code></li>
      </ul>
    </li>
    <li><strong>Cut columns from a table</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with
      <ul>
        <li><em>“Cut columns”</em>: <code class="highlighter-rouge">c1,c9</code> (the gene ids and DE status)</li>
        <li><em>“Delimited by”</em>: <code class="highlighter-rouge">Tab</code></li>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“From”</em>: the output of <strong>Join</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span></li>
        <li>Rename to <code class="highlighter-rouge">goseq DE status</code></li>
      </ul>
    </li>
    <li><strong>Cut columns from a table</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with
      <ul>
        <li><em>“Cut columns”</em>: <code class="highlighter-rouge">c1,c11</code> (the gene ids and lengths)</li>
        <li><em>“Delimited by”</em>: <code class="highlighter-rouge">Tab</code></li>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“From”</em>: the output of <strong>Join</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span></li>
        <li>Rename to <code class="highlighter-rouge">goseq gene lengths</code></li>
      </ul>
    </li>
  </ol>
</blockquote>

<p>We now have the two required input files for goseq for both our basal and luminal contrasts.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-perform-go-analysis"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Perform GO analysis</h3>

  <ol>
    <li><strong>goseq</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with
      <ul>
        <li><i class="fa fa-folder-o" aria-hidden="true"></i><span class="visually-hidden">param-collection</span> <em>“Differentially expressed genes file”</em>: <code class="highlighter-rouge">goseq DE status</code></li>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“Gene lengths file”</em>: <code class="highlighter-rouge">goseq gene lengths</code></li>
        <li><em>“Gene categories”</em>:  <code class="highlighter-rouge">Get categories</code>
          <ul>
            <li><em>“Select a genome to use”</em>:  <code class="highlighter-rouge">Mouse(mm10)</code></li>
            <li><em>“Select Gene ID format”</em>:  <code class="highlighter-rouge">Entrez Gene ID</code></li>
            <li><em>“Select one or more categories”</em>: <code class="highlighter-rouge">GO: Biological Process</code></li>
          </ul>
        </li>
        <li><em>“Output Options”</em>
          <ul>
            <li><em>“Output Top GO terms plot?”</em> <code class="highlighter-rouge">Yes</code></li>
          </ul>
        </li>
      </ul>
    </li>
  </ol>
</blockquote>

<p>goseq generates a big table with the following columns for each GO term:</p>
<ol>
  <li><code class="highlighter-rouge">category</code>: GO category</li>
  <li><code class="highlighter-rouge">over_rep_pval</code>: <em>p</em>-value for over representation of the term in the differentially expressed genes</li>
  <li><code class="highlighter-rouge">under_rep_pval</code>: <em>p</em>-value for under representation of the term in the differentially expressed genes</li>
  <li><code class="highlighter-rouge">numDEInCat</code>: number of differentially expressed genes in this category</li>
  <li><code class="highlighter-rouge">numInCat</code>: number of genes in this category</li>
  <li><code class="highlighter-rouge">term</code>: detail of the term</li>
  <li><code class="highlighter-rouge">ontology</code>: MF (Molecular Function - molecular activities of gene products), CC (Cellular Component - where gene products are active), BP (Biological Process - pathways and larger processes made up of the activities of multiple gene products)</li>
  <li><code class="highlighter-rouge">p.adjust.over_represented</code>: <em>p</em>-value for over representation of the term in the differentially expressed genes, adjusted for multiple testing with the Benjamini-Hochberg procedure</li>
  <li><code class="highlighter-rouge">p.adjust.under_represented</code>: <em>p</em>-value for over representation of the term in the differentially expressed genes, adjusted for multiple testing with the Benjamini-Hochberg procedure</li>
</ol>

<p>To identify categories significantly enriched/unenriched below some p-value cutoff, it is necessary to use the adjusted <em>p</em>-value.</p>

<p>A plot of the top 10 over-represented GO terms (by adjusted <em>p</em>-value) can be output from the goseq tool to help visualise results. Note that the top 10 are selected by adjusted p-value so if there are multiple terms with the same value there will be more than 10 terms in the plot. Click on the <code class="highlighter-rouge">Top over-represented GO terms plot</code> in the history. There should be 2 PDFs, one for each contrast, that look similar to below.</p>

<figure id="figure-34"><img src="../../images/limma-voom/basal_top_GO.png" alt="Basal Plot" title="Basal pregnant vs lactating top 10 GO terms" /><figcaption><span class="figcaption-prefix">Figure 34:</span> Basal pregnant vs lactating top 10 GO terms</figcaption></figure>

<figure id="figure-35"><img src="../../images/limma-voom/luminal_top_GO.png" alt="Luminal Plot" title="Luminal pregnant vs lactating top 10 GO terms" /><figcaption><span class="figcaption-prefix">Figure 35:</span> Luminal pregnant vs lactating top 10 GO terms</figcaption></figure>

<p>The Fu paper also used goseq and found enrichment for cell contractility genes in the basal cells and enrichment in the luminal cells for general metabolic processes, lipid biosynthesis and transport proteins, and .</p>

<blockquote class="question">
  <h3 id="question-questions"><i class="fa fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</h3>

  <p>Take a look at the top 10 GO plots for the luminal and basal contrast. How do you think they compare to what the authors found?</p>

  <blockquote class="solution">
    <h3 id="solution-solution-7"><i class="fa fa-eye" aria-hidden="true"></i><span class="visually-hidden">solution</span> Solution</h3>

    <p>The top 10 GO terms seem to describe similar processes to what the authors found.</p>

  </blockquote>
</blockquote>

<h2 id="gene-set-enrichment-analysis-with-fgsea">Gene Set Enrichment Analysis with <strong>fgsea</strong></h2>

<p>Gene Set Enrichment Analysis (GSEA) <a href="https://www.ncbi.nlm.nih.gov/pubmed/16199517">(Subramanian et al., 2005)</a> is a widely used method that determines whether a set of genes is enriched in a list of differentially expressed genes. Unlike the previous method with goseq, no threshold is applied for what is considered “differentially expressed”, all genes are used. If a gene set falls at either the top (over-expressed) or bottom (under-expressed) of the list it is said to be enriched. <a href="https://www.biorxiv.org/content/early/2016/06/20/060012">fgsea</a> is a faster implementation of the GSEA method. fgsea requires a ranked list of genes and some gene sets to test.</p>

<p>The <a href="http://software.broadinstitute.org/gsea/msigdb/index.jsp">Molecular Signatures Database (MSigDB)</a> contains curated collections of gene sets that are commonly used in a GSEA analysis. They can be downloaded from the <a href="http://software.broadinstitute.org/gsea/downloads.jsp">Broad website</a>. But these collections are only of human gene sets. If working with another species you would need to first map the genes to their human orthologues. However, MSigDB versions for mouse are provided <a href="http://bioinf.wehi.edu.au/software/MSigDB/index.html">by the Smyth lab here</a> so we’ll use those. There are several MSigDB collections, we’ll use the <a href="https://www.cell.com/cell-systems/fulltext/S2405-4712(15)00218-5">Hallmark collection</a>, which contains 50 gene sets. According to MSigDB, “each gene set in the hallmark collection consists of a “refined” gene set, derived from multiple “founder” sets, that conveys a specific biological state or process and displays coherent expression. The hallmarks effectively summarize most of the relevant information of the original founder sets and, by reducing both variation and redundancy, provide more refined and concise inputs for gene set enrichment analysis”.</p>

<p>There are several ways we could choose to rank our genes, we could rank by log-fold change (most upregulated to most downregulated) but that doesn’t take into account any error in the log fold change value. Another way is to use the “signed fold change” which is to rank by the sign of the fold change multiplied by the P value (as described <a href="http://genomespot.blogspot.com/2014/09/data-analysis-step-8-pathway-analysis.html">here</a>. We could also use the t statistic that’s output from limma, as that takes into account the log-fold change and it’s standard error, see <a href="https://support.bioconductor.org/p/6124/">here</a> for more explanation on the t statistic. We’ll use the t statistic to rank here.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-perform-gene-set-enrichment-with-fgsea"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Perform gene set enrichment with fgsea</h3>

  <ol>
    <li>Import the mouse Hallmark collection of gene sets from <code class="highlighter-rouge">http://bioinf.wehi.edu.au/software/MSigDB/mouse_H_v5p2.rdata</code> using the Paste/Fetch upload box
      <ul>
        <li>Set the file <strong>Type</strong> to <code class="highlighter-rouge">rdata</code></li>
        <li>Rename file as <code class="highlighter-rouge">mouse_hallmark_sets</code></li>
      </ul>
    </li>
    <li><strong>Cut columns from a table</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with
      <ul>
        <li><em>“Cut columns”</em>: <code class="highlighter-rouge">c1,c6</code> (the Entrez gene ids and t-statistic)</li>
        <li><em>“Delimited by”</em>: <code class="highlighter-rouge">Tab</code></li>
        <li><i class="fa fa-folder-o" aria-hidden="true"></i><span class="visually-hidden">param-collection</span> <em>“From”</em>: the <code class="highlighter-rouge">DE tables</code> output of <strong>limma</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span></li>
      </ul>
    </li>
    <li><strong>Sort data in ascending or descending order</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with
      <ul>
        <li><i class="fa fa-folder-o" aria-hidden="true"></i><span class="visually-hidden">param-collection</span> <em>“Sort Query”</em>: the output of <strong>Cut</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span></li>
        <li><em>“Number of header lines”</em>: <code class="highlighter-rouge">1</code></li>
        <li><em>“Column selections”</em>:
          <ul>
            <li><em>“on column”</em>: <code class="highlighter-rouge">Column: 2</code></li>
            <li><em>“in”</em>: <code class="highlighter-rouge">Descending order</code></li>
            <li><em>“Flavor”</em>: <code class="highlighter-rouge">Fast numeric sort (-n)</code></li>
          </ul>
        </li>
      </ul>
    </li>
    <li><strong>fgsea</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with
      <ul>
        <li><i class="fa fa-folder-o" aria-hidden="true"></i><span class="visually-hidden">param-collection</span> <em>“Ranked Genes”</em>: the output of <strong>Sort</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span></li>
        <li><em>“File has header?”</em>: <code class="highlighter-rouge">Yes</code></li>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“Gene Sets”</em>: <code class="highlighter-rouge">mouse_hallmark_sets</code></li>
        <li><em>“Minimum Size of Gene Set”</em>: <code class="highlighter-rouge">15</code></li>
        <li><em>“Output plots”</em>: <code class="highlighter-rouge">Yes</code></li>
      </ul>
    </li>
  </ol>
</blockquote>

<p>fgsea outputs a table of results containing a list of pathways with P values and enrichment scores. It can also output a summary table plot of the top pathways like the one shown below for the <code class="highlighter-rouge">basallpregnant-basallactate</code> contrast.</p>

<figure id="figure-36" width="50%"><img src="../../images/limma-voom/fgsea_table.png" alt="fgsea Table" title="fgsea Summary table" /><figcaption><span class="figcaption-prefix">Figure 36:</span> fgsea Summary table</figcaption></figure>

<p>An enrichment plot of the each of the top pathways can also be produced, one is shown below. The barcode pattern shows where the genes in the set are found in the list of ranked genes. Most of the bars to the left indicate enrichment of the set at the top of the ranked list of genes (upregulated) and most bars towards the right indicate enrichment at the bottom of the list (downregulated). The enrichment score reflects the degree to which the genes are enriched at the top or bottom of the list.</p>

<figure id="figure-37"><img src="../../images/limma-voom/fgsea_enrichplot.png" alt="fgsea Enrichment" title="fgsea Enrichment plot" /><figcaption><span class="figcaption-prefix">Figure 37:</span> fgsea Enrichment plot</figcaption></figure>

<h2 id="ensemble-gene-set-enrichment-analyses-with-egsea">Ensemble gene set enrichment analyses with <strong>EGSEA</strong></h2>

<p>The ensemble of genes set enrichment analyses (EGSEA) <a href="https://www.ncbi.nlm.nih.gov/pubmed/27694195">(Alhamdoosh et al, 2017)</a> is a method developed for RNA-sequencing data that combines results from multiple algorithms and calculates collective gene set scores, to try to improve the biological relevance of the highest ranked gene sets. EGSEA has built-in gene sets from MSigDB and KEGG for human and mouse. We’ll show here how it can be used with the MSigDB Hallmark collection and KEGG pathways. For input we need a count matrix and EGSEA will perform a limma-voom analysis before gene set testing. We can use the filtered counts output from limma, where the low count genes have been filtered out, we just need to remove the gene symbol and description columns. We also need a symbols mapping file containing just the Entrez ids and symbols, which we can generate from the filtered counts file. The third input we need is a factors information file, containing what groups the samples belong to, we can use the one we used with limma.</p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-perform-ensemble-gene-set-testing-with-egsea"><i class="fa fa-pencil" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Perform ensemble gene set testing with EGSEA</h3>

  <ol>
    <li>Rerun <strong>limma</strong> selecting <em>“Output Filtered Counts Table?”</em>: <code class="highlighter-rouge">Yes</code></li>
    <li><strong>Cut</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>: Run <strong>Cut columns from a table (cut)</strong> with the following settings:
      <ul>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“File to cut”</em>: <code class="highlighter-rouge">Filtered Counts</code> output from <strong>limma</strong></li>
        <li><em>“Operation”</em>: <code class="highlighter-rouge">Discard</code></li>
        <li><em>“List of fields”</em>: Select <code class="highlighter-rouge">Column:2</code>, <code class="highlighter-rouge">Column:3</code></li>
        <li>Rename to <code class="highlighter-rouge">EGSEA counts</code></li>
      </ul>
    </li>
    <li><strong>Cut</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span>: Run <strong>Cut columns from a table (cut)</strong> with the following settings:
      <ul>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“File to cut”</em>: <code class="highlighter-rouge">Filtered Counts</code> output from <strong>limma</strong></li>
        <li><em>“Operation”</em>: <code class="highlighter-rouge">Keep</code></li>
        <li><em>“List of fields”</em>: Select <code class="highlighter-rouge">Column:1</code>, <code class="highlighter-rouge">Column:2</code></li>
        <li>Rename to <code class="highlighter-rouge">EGSEA anno</code></li>
      </ul>
    </li>
    <li><strong>EGSEA</strong> <i class="fa fa-wrench" aria-hidden="true"></i><span class="visually-hidden">tool</span> with
      <ul>
        <li><em>“Count Files or Matrix?</em>”: <code class="highlighter-rouge">Single Count Matrix</code>
          <ul>
            <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“Count Matrix”</em>: Select <code class="highlighter-rouge">EGSEA counts</code></li>
          </ul>
        </li>
        <li><em>“Input factor information from file?”</em>: <code class="highlighter-rouge">Yes</code>
          <ul>
            <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“Factor File”</em>: Select <code class="highlighter-rouge">factordata</code></li>
          </ul>
        </li>
        <li><i class="fa fa-file-o" aria-hidden="true"></i><span class="visually-hidden">param-file</span> <em>“Symbols Mapping file”</em>: <code class="highlighter-rouge">EGSEA anno</code></li>
        <li><em>“Contrast of Interest”</em>: <code class="highlighter-rouge">basalpregnant-basallactate</code></li>
        <li><em>“Contrast of Interest”</em>: <code class="highlighter-rouge">luminallpregnant-luminalllactate</code></li>
        <li><em>“Species”</em>: <code class="highlighter-rouge">mouse</code></li>
        <li><em>“Gene Set Testing Methods”</em>: Tick <code class="highlighter-rouge">camera</code>, <code class="highlighter-rouge">safe</code>, <code class="highlighter-rouge">gage</code>, <code class="highlighter-rouge">zscore</code>, <code class="highlighter-rouge">gsva</code>, <code class="highlighter-rouge">globaltest</code>, <code class="highlighter-rouge">ora</code>, <code class="highlighter-rouge">ssgsea</code>, <code class="highlighter-rouge">padog</code>, <code class="highlighter-rouge">plage</code>, <code class="highlighter-rouge">fry</code></li>
        <li><em>“MSigDB Gene Set Collections”</em>: <code class="highlighter-rouge">H: hallmark gene sets</code></li>
        <li><em>“KEGG Pathways”</em>: <code class="highlighter-rouge">Metabolism</code> and <code class="highlighter-rouge">Signalling</code></li>
        <li><em>“I certify that I am not using this tool for commercial purposes”</em>: <code class="highlighter-rouge">Yes</code></li>
      </ul>
    </li>
  </ol>
</blockquote>

<p>This generates a report like below.</p>

<figure id="figure-38" width="50%"><img src="../../images/limma-voom/EGSEA_report.png" alt="EGSEA report" title="EGSEA report" /><figcaption><span class="figcaption-prefix">Figure 38:</span> EGSEA report</figcaption></figure>

<p>In addition to a table of results, plots are generated like the heatmaps of the top ranked pathways, shown below. Note that we see some similar pathways in the results here as with the fgsea analysis.</p>

<figure id="figure-39" width="50%"><img src="../../images/limma-voom/EGSEA_heatmaps.png" alt="EGSEA heatmaps" title="EGSEA heatmaps" /><figcaption><span class="figcaption-prefix">Figure 39:</span> EGSEA heatmaps</figcaption></figure>

<p>KEGG pathway diagrams are generated if KEGG pathways are selected, as shown below.  These show the expression values of the genes, genes upregulated in the contrast are shown in red, downregulated in blue.</p>

<figure id="figure-40"><img src="../../images/limma-voom/EGSEA_KEGG.png" alt="EGSEA KEGG" title="EGSEA KEGG pathways" /><figcaption><span class="figcaption-prefix">Figure 40:</span> EGSEA KEGG pathways</figcaption></figure>

<h1 class="no_toc" id="conclusion">Conclusion</h1>

<p>In this tutorial we have seen how counts files can be converted into differentially expressed genes with limma-voom, and some pathway analysis methods that can be applied. This follows on from the accompanying tutorial, RNA-seq reads to counts, that showed how to generate counts from the raw reads (FASTQ) for this dataset. In this part we have learnt ways to visualise the count data and QC checks that can be performed to help assess the quality and results. We have also reproduced results similar to what the authors found in the original paper with this dataset. For further reading on analysis of RNA-seq count data and the methods used here, see the articles; RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR <a href="https://f1000research.com/articles/5-1408/v2">(Law et al. 2016)</a> and From reads to genes to pathways: differential expression analysis of RNA-Seq experiments using Rsubread and the edgeR quasi-likelihood pipeline <a href="https://f1000research.com/articles/5-1438/v2">(Chen, Lun, Smyth 2016)</a>.</p>


        
        <blockquote class="key_points">
            <h3><i class="fa fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>

            <ul>
                
                <li>The limma-voom tool can be used to perform differential expression and output useful plots</li>
                
                <li>Multiple comparisons can be input and compared</li>
                
                <li>Results can be interactively explored with limma-voom via Glimma</li>
                
                <li>Multiple methods can be used to identify differentially expressed pathways</li>
                
            </ul>
        </blockquote>
        

        
        <h1>Useful literature</h1>
        <p>Further information, including links to documentation and original publications, regarding the tools, analysis techniques and the interpretation of results described in this tutorial can be found <a href="/training-material/topics/transcriptomics#references">here</a>.</p>
        

        

        <h3><i class="fa fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>
        <hr>
        <br>

        <iframe class="google-form" src="https://docs.google.com/forms/d/e/1FAIpQLSd4VZptFTQ03kHkMz0JyW9b6_S8geU5KjNE_tLM0dixT3ZQmA/viewform?embedded=true&entry.1235803833=RNA-seq counts to genes and pathways (Transcriptomics)">Loading...
        </iframe>

        <blockquote class="feedback">
            <h3><i class="fa fa-comments-o" aria-hidden="true"></i><span class="visually-hidden">feedback</span> Give us even more feedback on this content!</h3>
            <p>To give us more detailed feedback about these materials, please take a moment to fill in the extended <a href="https://docs.google.com/forms/d/e/1FAIpQLSc_GLZRpxZGAL9tN6DA1bE6WkNhmQdXT7B16TpOb-X4YwKUsQ/viewform?entry.1548889262=RNA-seq counts to genes and pathways (Transcriptomics)">Feedback Form</a>.</p>
        </blockquote>
    </section>
</div>


<footer>
    <div class="container">
        <p>
            This material is the result of a collaborative work. Thanks to the
            <a href="https://wiki.galaxyproject.org/Teach/GTN">Galaxy Training Network</a>
            and all the <a href="/training-material/hall-of-fame">contributors</a> (Anna Trigos, Belinda Phipson, Charity Law, Harriet Dashnow, Maria Doyle, Matt Ritchie)!
        </p>
        <p>
            Found a typo? Something is wrong in this tutorial? Edit it on
            <a href="https://github.com/galaxyproject/training-material/tree/master/topics/transcriptomics/tutorials/limma-voom/tutorial.md">GitHub</a>.
        </p>
    </div>
</footer>

    </body>
    <script type="text/javascript" src="/training-material/assets/js/jquery.slim.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/popper.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/bootstrap.min.js?v=3"></script>
    <script type="text/javascript" src="/training-material/assets/js/details-element-polyfill.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="/training-material/assets/js/main.js"></script>
</html>
