> ### {% icon hands_on %} Hands-on: Visualization of the reads in JBrowse
>
> 1. **JBrowse** {% icon tool %} genome browser, with the following parameters:
>    - *"Reference genome to display"*: Use a built-in genome
>       - *"Select a reference genome"*: `mm10`
>    - *"Produce Standalone Instance"*: Yes
>    - *"Insert Track Group"*:
>       - *"Insert Annotation Track"*:
>          - *"Track Type"*: BAM Pileups
>          - *"BAM Track Data"*: `aligned reads` (output of **Bowtie2** {% icon tool %})
>          - *"Autogenerate SNP Track"*: Yes
>          - *"Track Visibility"*: On for new users
> 2. Visualise the dataset {% icon galaxy-eye %}
> 3. Zoom on the `{{ include.region_to_zoom }}`
{: .hands_on}

> ### {% icon comment %} Slow
> This can take about a minute or two to run, depending on your training instance's resources. It takes time as the server builds up a small website for you, and pre-processes the reference genome into a more efficient format. If you wanted to share this with your colleagues, you could download this dataset and directly place it on your webserver
{: .comment}

The reads have a direction: they are mapped to the forward or reverse strand, respectively. When clicking on a read, extra information is displayed

> ### {% icon question %} Questions
>
> 1. What does the teardrop shape and line mean in the autogenerated SNP track?
>
> > ### {% icon solution %} Solution
> > 1. If enough reads have a different value, then it is marked with a teardrop icon. The coverage plot is marked in height with percentage of reads with a different call at that position
> {: .solution }
{: .question}
