<!DOCTYPE html>
<html>









  <head>
    <meta charset="utf-8">
    <title>Mapping</title>
    
    <script>
        (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
        (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
        m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
        })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
        ga('create', 'UA-45719423-18' , 'auto');
        ga('send', 'pageview');
    </script>
    
    <link rel="stylesheet" href="/training-material/assets/css/slides.css">
    <link rel="stylesheet" href="/training-material/assets/css/font-awesome.css" id="theme">
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="shortcut icon" href="/training-material/favicon.ico" type="image/x-icon" />

    
    
    
    
    <meta name="description" content="Analyses of sequences" />
    <meta property="og:title" content="Galaxy Training: Mapping" />
    <meta property="og:description" content="Analyses of sequences" />
    <meta property="og:image" content="/training-material/assets/images/GTNLogo1000.png" />
    <script type="application/ld+json">
      


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "citation": {
    "@type": "CreativeWork",
    "name": "Community-Driven Data Analysis Training for Biology",
    "url": "https://doi.org/10.1016/j.cels.2018.05.012"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "discussionUrl": "https://gitter.im/Galaxy-Training-Network/Lobby",
  "headline": "Mapping",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "/blob//LICENSE.md",
  "producer": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "provider": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "courseCode": "sequence-analysis / mapping / slides",
  "learningResourceType": "slides",
  "name": "Slides for 'Mapping' tutorial",
  "isPartOf": {
    "@type": "Course",
    "name": "Mapping",
    "description": "Mapping",
    "learningResourceType": "tutorial",
    "interactivityType": "expositive",
    "provider": {
      "@type": "Organization",
      "email": "galaxytrainingnetwork@gmail.com",
      "name": "Galaxy Training Network",
      "url": "https://galaxyproject.org/teach/gtn/"
    }
  },
  "url": "https://galaxyproject.github.io//training-material/topics/sequence-analysis/tutorials/mapping/slides.html",
  "timeRequired": "PT1H",
  "description": "Mapping\\nThe questions this  addresses are:\n - What is mapping (alignment)?\n - What is the BAM format?\n - How can we view aligned sequences?\n\n\\nThe objectives are:\n - Understand the basic concept of mapping\n - Learn about factors influencing alignment\n - See a genome browser used to better understand your aligned data\n\n",
  "coursePrerequisites": [
    {
      "@type": "CreativeWork",
      "url": "https://galaxyproject.github.io//training-material/topics/introduction/",
      "name": "Introduction to Galaxy Analyses",
      "description": "Introduction to Galaxy Analyses",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    }
  ],
  "hasPart": [

  ],
  "mentions": [
    {
      "@type": "Thing",
      "url": "https://doi.org/10.5281/zenodo.61771",
      "name": "Training data for Mapping tutorial"
    }
  ],
  "author": [
    {
      "@type": "Person",
      "name": "Joachim Wolff"
    },
    {
      "@type": "Person",
      "name": "Saskia Hiltemann"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Joachim Wolff"
    },
    {
      "@type": "Person",
      "name": "Saskia Hiltemann"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Sequence analysis",
      "description": "Analyses of sequences",
      "url": "https://galaxyproject.github.io//training-material/topics/sequence-analysis/"
    }
  ]
}
    </script>
  </head>
  <body>
    <textarea id="source">
name: inverse
layout: true
class: center, middle, inverse

<div class="my-header"><span>
<a href="/training-material/topics/sequence-analysis" title="Return to topic page" ><i class="fa fa-level-up" aria-hidden="true"></i></a>

<a class="nav-link" href="https://github.com/galaxyproject/training-material/edit/master/topics/sequence-analysis/tutorials/mapping/slides.html"><i class="fa fa-pencil" aria-hidden="true"></i></a>

</span></div>

<div class="my-footer"><span>

<img src="/training-material/assets/images/GTN-60px.png" alt="Galaxy Training Network" style="height: 40px;"/>

</span></div>

---

# Mapping



---

## Requirements

Before diving into this slide deck, we recommend you to have a look at:



    
        
        
            
        
- [Introduction to Galaxy Analyses](/training-material/topics/introduction)
    






.footnote[Tip: press `P` to view the presenter notes]

???
Presenter notes contain extra information which might be useful if you
intend to use these slides for teaching.

Press `P` again to switch presenter notes off

---





### &lt;i class=&quot;fa fa-question-circle&quot; aria-hidden=&quot;true&quot;&gt;&lt;/i&gt;&lt;span class=&quot;visually-hidden&quot;&gt;question&lt;/span&gt; Questions


- What is mapping (alignment)?

- What is the BAM format?

- How can we view aligned sequences?


---


### &lt;i class=&quot;fa fa-bullseye&quot; aria-hidden=&quot;true&quot;&gt;&lt;/i&gt;&lt;span class=&quot;visually-hidden&quot;&gt;objectives&lt;/span&gt; Objectives


- Understand the basic concept of mapping

- Learn about factors influencing alignment

- See a genome browser used to better understand your aligned data


---


# Example NGS pipeline

![High level view of a typical NGS workflow](../../images/mapping/variant_calling_workflow.png)

A high level view of a typical NGS bioinformatics workflow

???
- Mapping step occurs if a reference genome is available for the organism of interest
  - else: de-novo assembly
- Variant calling step is just an example, after mapping can do many steps
  - Structural Variants / Fusion genes
  - Differential Gene expression
  - Alternative Splicing
  - ..

---

# What is mapping?

.pull-left[
![Mapping vs assembly](../../images/mapping/mapping_assembly.png)
]

.pull-right[
- Short reads must be combined into longer fragments

- **Mapping:** use a reference genome as a guide

- **De-novo assembly:** without reference genome
]

???
- Mapping is also referred to as *alignment*
- Short reads produced by sequencer must be combined into larger contigs
  - e.g. reconstruct the chromosomes
  - mapping uses a reference genome as a guide
  - can subsequently find where our sample differs from reference (variants)


- This tutorial only deals with mapping/alignment
- There are other tutorials available for de-novo assembly



---
class: top

# Sequence alignment

- Determine position of short read on the genome
  ```
  Reference: . . . A A C G C C T T . . .

  Read:            A G G G G C C T T
  ```

???
Consider situation where we must map this read to this reference genome
- This area looks pretty similar, but not quite identical..

---
class: top

# Sequence alignment

- Determine position of short read on the genome
  ```
  Reference: . . . A A - C G C C T T . . .              | = match
  .                | : - : | | | | |                    : = mismatch
  Read:            A G G G G C C T T                    - = gap
  ```
???
But if we introduce gaps and allow for some mismatches in bases, this matches
up pretty well..

- `|` = match
- `:` = mismatch
- `-` = gap

---
class: top

# Sequence alignment

- Determine position of short read on the genome
  ```
  Reference: . . . A A - C G C C T T . . .              | = match
  .                | : - : | | | | |                    : = mismatch
  Read:            A G G G G C C T T                    - = gap
  ```


- Read could align to multiple places
  .center[.image-50[![Illustration of multimapped read](../../images/mapping/multimap.png)]]


- How to handle multi-mapped reads? Depends on tool:
  - Map to best region (but what about ties?)
  - Map to all regions
  - Map to one region randomly
  - Discard read


- How do we determine *best* region?
  - Assign ***alignment score*** to every mapping

???


---
class: top

# Alignment Scoring (basics)

- **Reward** for a match (e.g. +10), **penalty** for a mismatch (e.g. -5)
- **Penalty** for gaps
   - *Linear:* every gap same penalty (e.g. -5)
   - *Affine:* gap open vs gap extend (e.g. -5 and -1)


- Different tools use different scoring values (and give different results)


.center[
.image-25[![](../../images/mapping/scoring_example.png)]

**Example** (with affine gap penalty)
]

???
- Each locus get scored independently (first row of scores in example)
- Scores from all loci are added up (cumulative score row)
- Final score for entire alignment in this example is 29

- These reward and penalty values are just examples and will vary

---
class: top

# Alignment Scoring (advanced)

- **Base quality**
    - Mismatch of low-confidence base: lower penalty
    - Mismatch of high-confidence base: higher penalty


- **Transitions vs transversions**
  - Transitions about 2x as frequent as transversions


.center[
.image-50[ ![Transitions vs transversions](../../images/mapping/ti_tv.png) ]
.image-25[ ![Example scoring matrix](../../images/mapping/ti_tv_scoring.png) ]
]


- Knowledge about sequencing platform and biases
  - Optimize for read length, error rate, homopolymer accuracy, etc..


.footnote[More information about mapping algorithms: [10.1089/cmb.2012.0022](https://doi.org/10.1089/cmb.2012.0022)]
???

Many more complexities may be considered, different tools make different choices

Transitions are more likely to occur in real sequences, so may give lower penalty than transversions

**Transitions** are interchanges of two-ring purines (A G) or of one-ring pyrimidines (C T): they therefore involve bases of similar shape.

**Transversions** are interchanges of purine for pyrimidine bases, which therefore involve exchange of one-ring and two-ring structures.

![Transitions and transversions](../../images/mapping/transition_transversion.gif)


---

# Looks easy but..

---
class: top

# Sequence Alignment

```
Reference: AAA CAGTGA GAA
Observed:  AAA TCTCT  GAA
```

???

Suppose you have this reference sequence (top) and the bottom sequence is one of the reads in your sample.

How would you align these?

---
class: top

# Sequence Alignment

```
Reference: AAA CAGTGA GAA
Observed:  AAA TCTCT  GAA
```

&lt;table style=&quot;width:100%; table-layout: fixed; font-size:0.8em&quot;&gt;
&lt;th&gt;Alignment&lt;/th&gt;&lt;th&gt;&lt;/th&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt;Maybe like this?&lt;/td&gt;
&lt;/tr&gt;
&lt;/table&gt;

???
This is one possibility, is it the only one?

---
class: top

# Sequence Alignment

```
Reference: AAA CAGTGA GAA
Observed:  AAA TCTCT  GAA
```

&lt;table style=&quot;width:100%; table-layout:fixed; font-size:0.8em;&quot;&gt;
&lt;th&gt;Alignment&lt;/th&gt;&lt;th&gt;&lt;/th&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt;Maybe like this?&lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt; Or this? &lt;/td&gt;
&lt;/tr&gt;

&lt;/table&gt;

???
This is also a possible alignment. Not easy to say which is better.

---
class: top

# Sequence Alignment

```
Reference: AAA CAGTGA GAA
Observed:  AAA TCTCT  GAA
```

&lt;table style=&quot;width:100%; table-layout:fixed; font-size:0.8em&quot;&gt;
&lt;th&gt;Alignment&lt;/th&gt;&lt;th&gt;&lt;/th&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt;Maybe like this?&lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt; Or this? &lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTGAGAA
|||:-:|::|||
AAAT-CTCTGAA

&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt;Or..? &lt;/td&gt;
&lt;/tr&gt;

&lt;/table&gt;

???
And a third option

---
class: top

# Sequence Alignment

```
Reference: AAA CAGTGA GAA
Observed:  AAA TCTCT  GAA
```

&lt;table style=&quot;width:100%; table-layout:fixed; font-size:0.8em&quot;&gt;
&lt;th&gt;Alignment&lt;/th&gt;&lt;th&gt;&lt;/th&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt;Maybe like this?&lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt; Or this? &lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTGAGAA
|||:-:|::|||
AAAT-CTCTGAA

&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt;Or..? &lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTCA-----GAA
|||-----------|||
AAA------TCTCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt; What about this? &lt;/td&gt;
&lt;/tr&gt;
&lt;/table&gt;

???
There is no one right way to do alignment
  - Hard to say which of these is &quot;better&quot; or &quot;worse&quot;
  - Just different choices, but all valid

Mapping is a non-trivial problem!

---
class: top

# Sequence Alignment

```
Reference: AAA CAGTGA GAA
Observed:  AAA TCTCT  GAA
```

&lt;table style=&quot;width:100%; table-layout:fixed; font-size:0.8em&quot;&gt;
&lt;th&gt;Alignment&lt;/th&gt;&lt;th&gt;Tool&lt;/th&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt;Novoalign&lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt; Ssaha2 &lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTGAGAA
|||:-:|::|||
AAAT-CTCTGAA

&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt; BWA &lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTCA-----GAA
|||-----------|||
AAA------TCTCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt; Complete Genomics &lt;/td&gt;
&lt;/tr&gt;
&lt;/table&gt;


???
We didn't just make these up, these real aligners gave these different results

---
class: top

# Sequence Alignment

```
Reference: AAA CAGTGA GAA
Observed:  AAA TCTCT  GAA
```

&lt;table style=&quot;width:100%; table-layout:fixed; font-size:0.8em&quot;&gt;
&lt;th&gt;Alignment&lt;/th&gt;&lt;th&gt;Variant calls&lt;/th&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt;&lt;pre&gt;
ins T
del AG
sub GA -&gt; CT
&lt;/pre&gt;&lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt;&lt;pre&gt;
del C
sub AG -&gt; CT
sub GA -&gt; CT
&lt;/pre&gt;&lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTGAGAA
|||:-:|::|||
AAAT-CTCTGAA

&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt;&lt;pre&gt;
snp C -&gt; T
del A
snp G -&gt; C
sub GA -&gt; CT
&lt;/pre&gt;&lt;/td&gt;
&lt;/tr&gt;

&lt;tr&gt;&lt;td&gt;&lt;pre&gt;
AAACAGTCA-----GAA
|||-----------|||
AAA------TCTCTGAA
&lt;/pre&gt;&lt;/td&gt;
&lt;td&gt;&lt;pre&gt;
del CAGTGA
int TCTCT
&lt;/pre&gt;&lt;/td&gt;
&lt;/tr&gt;
&lt;/table&gt;


???
**Important:** Mapping can affect downstream analysis!

These different mappings led to different variants, and hard to tell the are equivalent.

---

# Try it yourself!

- Lego time! Who wants to volunteer?

- Or try this [online sequence alignment game](http://teacheng.illinois.edu/SequenceAlignment/):

.image-75[![Recording of alignment game](../../images/mapping/alignment.gif)]

.footnote[https://tinyurl.com/sequence-alignment]

???
Can have learners play around with this alignment game now

Or use lego bricks, each nucleotide a different colour


---
class: top

## Tools, tools, tools

60+ different mappers, many comparison papers


.left-column70[![Timeline of mapping tools](../../images/mapping/ngs_read_mappers_timeline.jpeg)]
]
.right-column30[
- Bowtie2
- HISAT
- MAQ
- BWA
- Eland
- SOAP
- STAR
- BISMARCK
- Novoalign
- TopHat2
- BLASR
- ..
]


.footnote[Feature comparison:  https://www.ebi.ac.uk/~nf/hts_mappers/]
---

## Choosing an Aligner

- Each tool makes **different choices** during alignment
  - Choice of aligner may **affect downstream results**
  - Default options may not be best for your data!


- Best tool for your data **depends on many factors**
  - Type of experiment (e.g. DNA, RNA, Bisulphite)
  - Sequencing platform
  - Compute resources vs sensitivity
  - Read characteristics (paired/single end, read length)


.center[
.image-40[![Mapping RNA](../../images/mapping/spliced_mapper.png)]
]
.footnote[**Figure:** mapping of RNA-seq reads is different than DNA-seq]


???
Choice of mapper depends on your experiment
 - Some mappers are good for DNA but not RNA
 - Some mappers do well in highly rearranged genomes (e.g. cancer), others less so
 - Some mappers do well on some platforms but worse on others
   - e.g. Oxford Nanopore with its long reads but high error rates
 - ..

Read characteristics such as
 - paired end vs single
 - length of reads
 -

Or other factors
 - STAR needs a LOT of RAM
 - Do you need results fast? or accurate? (e.g. medical setting)

&quot;Why not BLAST or BLAT?&quot;
- optimized for longer sequences
- not base quality aware
- too slow

---

# Know your data!


*“... there is no tool that outperforms all of the others in
all the tests. Therefore, the end user should clearly
specify his needs in order to choose the tool that
provides the best results.”* - Hatem et al BMC Bioinformatics 2013, 14:184


.footnote[ [DOI: 10.1186/1471-2105-14-184](https://doi.org/10.1186/1471-2105-14-184) ]

???

Know the data you are working with and pick the right mapper and parameters for the job!

Not an easy task..

---

# File Formats

---

# SAM/BAM file format

![Example of SAM file format](../../images/bam_file.png &quot;SAM/BAM file&quot;)

---

# SAM/BAM file format

![More detailed view of SAM format](../../images/bam_file_explained.png &quot;SAM/BAM file detail&quot;)

SAM: Sequence Alignment/Map format

BAM: Binary SAM

---

# Genome Browser

- Visualise aligned reads (BAM files)

![IGV Genome Browser](../../images/mapping/igv_animation.gif)

This is [IGV](https://software.broadinstitute.org/software/igv/)

???
- Can zoom in and out, drag left and right, explore your sample
- Zoom in for more information, mismatches, read information
- Many different genome browsers exist

---

# Genome Browers in Galaxy

- BAM datasets in Galaxy have links to display in genome browsers
  - [UCSC Genome Browser](https://genome-euro.ucsc.edu/cgi-bin/hgTracks), [Ensemble](https://www.ensembl.org), [IGV](https://software.broadinstitute.org/software/igv/), [IGB](https://bioviz.org/), [BAM.iobo](https://bam.iobio.io/home)


  ![Display application links in Galaxy](../../images/mapping/igv_link.png)


- 2 links for IGV (Integrative Genome Viewer)
  - **local:** start IGV on your machine first, then click link to load data
  - **[Reference genome name]**: downloads and starts IGV
     - **Human hg19** in screenshot
     - Requires [Java web start](https://java.com/en/download/faq/java_webstart.xml) be installed on your machine

???
In the mapping hands-on tutorial you will use IGV




---
### &lt;i class=&quot;fa fa-key&quot; aria-hidden=&quot;true&quot;&gt;&lt;/i&gt;&lt;span class=&quot;visually-hidden&quot;&gt;keypoints&lt;/span&gt; Key points


- Mapping is not trivial

- There are many mapping tools, best choice depends on your data

- Choice of mapper can affect downstream results

- Know your data!

- Genome browsers can be used to view aligned reads







---

## Thank you!

This material is the result of a collaborative work. Thanks to the [Galaxy Training Network](https://wiki.galaxyproject.org/Teach/GTN) and all the contributors!


<div class="contributors-line">


<a href="/training-material/hall-of-fame#joachimwolff" class="contributor-badge"><img src="https://avatars.githubusercontent.com/joachimwolff" alt="Joachim Wolff">Joachim Wolff</a>, <a href="/training-material/hall-of-fame#shiltemann" class="contributor-badge"><img src="https://avatars.githubusercontent.com/shiltemann" alt="Saskia Hiltemann">Saskia Hiltemann</a>


</div>



<img src="/training-material/assets/images/GTN.png" alt="Galaxy Training Network" style="height: 100px;"/>



    </textarea>
    <script src="/training-material/assets/js/remark-latest.min.js" type="text/javascript">
    </script>
    <script type="text/javascript">
      var slideshow = remark.create({navigation: {scroll: false,}});
      var hljs = remark.highlighter.engine;
    </script>
  </body>
</html>
