---
layout: tutorial_slides
logo: "GTN"

title: "Mapping"
zenodo_link: "https://doi.org/10.5281/zenodo.61771"
questions:
  - What two things are crucial for a correct mapping?
  - What is BAM?
objectives:
  - You will learn what mapping is
  - A genome browser is shown that helps you to understand your data
time_estimation: "1h"
key_points:
  - Know your data!
  - Mapping is not trivial
  - There are many mapping algorithms, it depends on your data which one to choose
contributors:
  - joachimwolff
  - shiltemann
---

# NGS pipelines

.left-column30[
![](../../images/mapping/variant_calling_workflow.png)
]

.right-column50[
A typical NGS workflow:

1. Raw reads from sequencer

2. **QC:** filtering and trimming

3. Mapping / Assembly

4. Data cleaning

5. Feature detection, e.g.
  - Variant Calling
  - Differential Gene Expression
  - .. many more
]

???

- Short reads by themselves don't tell us much
- Must be combined into larger contigs (chromosomes, plasmids, etc)

- Steps 1-4 are a part of many analyses
- Assemble reads can be used for analysis feature of interest
  - Variant analysis
  - Differential Gene Expression (RNA)
  - Taxonomy determination (Metagenomics)
  - .. many more

---

# Example NGS pipeline

![high level view of a typical NGS workflow](../../images/mapping/variant_calling_workflow.png)

A high level view of a typical NGS bioinformatics workflow

???
- Mapping step occurs if a reference genome is available for the organism of interest
  - else: de-novo assembly
- Variant calling step is just an example, after mapping can do many steps
  - Structural Variants / Fusion genes
  - Differential Gene expression
  - Alternative Splicing
  - ..

---

# What is mapping?

.pull-left[
![mapping vs assembly](../../images/mapping/mapping_assembly.png)
]

.pull-right[
- Short reads must be combined into longer contigs

- **Mapping:** use a reference genome as a guide

- **De-novo assembly:** without reference genome
]

???
- Mapping is also referred to as *alignment*
- This tutorial only deals with mapping/alighnment
- There are other tutorials available for de-novo assembly

---

# Sequence alignment

- Determine position of short read on the genome
  ```
  Reference: . . . A A C G C C T T . . .

  Read:            A G G G G C C T T
  ```
.hidden[
- Find best alignment by assigning a *score*
  - Reward for matches
  - Penalty for mismatches and gaps
    - gap open vs gap extend
- Further considerations
  - Take base quality into consideration?
  - transitions vs transversions?
    ![](../../images/mapping/transition_transversion.gif)
]

---

# Sequence alignment

- Determine position of short read on the genome
  ```
  Reference: . . . A A - C G C C T T . . .
  .                | : - : | | | | |
  Read:            A G G G G C C T T
  ```
.hidden[
- Find best alignment by assigning a *score*
  - Reward for matches
  - Penalty for mismatches and gaps
    - gap open vs gap extend
- Further considerations
  - Take base quality into consideration?
  - transitions vs transversions?
    ![](../../images/mapping/transition_transversion.gif)
]

???
| = match
: = mismatch
- = gap

---

# Sequence alignment

- Determine position of short read on the genome
  ```
  Reference: . . . A A - C G C C T T . . .
  .                | : - : | | | | |
  Read:            A G G G G C C T T
  ```
- Find best alignment by assigning a *score*
  - Reward for matches (`|`)
  - Penalty for mismatches (`:`) and gaps (`-`)
     - Gap open vs gap extend
.hidden[
- Further considerations
  - Take base quality into consideration?
  - Transitions vs transversions?
     - Transitions about 2x as frequent as transversions
       .image-05[ ![](../../images/mapping/transition_transversion.gif)]
]

---

# Sequence alignment

- Determine position of short read on the genome
  ```
  Reference: . . . A A - C G C C T T . . .
  .                | : - : | | | | |
  Read:            A G G G G C C T T
  ```
- Find best alignment by assigning a *score*
  - Reward for matches (`|`)
  - Penalty for mismatches (`:`) and gaps (`-`)
     - Gap open vs gap extend
- Further considerations:
  - Take base quality into consideration?
  - Transitions vs transversions?
     - Transitions about 2x as frequent as transversions
       .image-05[ ![](../../images/mapping/transition_transversion.gif) ]

???
**Transitions** are interchanges of two-ring purines (A G) or of one-ring pyrimidines (C T): they therefore involve bases of similar shape.

**Transversions** are interchanges of purine for pyrimidine bases, which therefore involve exchange of one-ring and two-ring structures.


---

# Looks easy but..

---

# Sequence Alignment

```
Sequence 1: AAA CAGTGA GAA
Sequence 2: AAA TCTCT  GAA
```

<table class="hidden" style="width:100%; font-size:0.8em">
<th>Alignment</th><th></th><th></th>

<tr><td><pre>
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
</pre></td>
<td>Maybe like this?</td>
</tr>

<tr><td><pre>
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
</pre></td>
<td> Or this? </td>
</tr>

<tr><td><pre>
AAACAGTGAGAA
|||:-:|::|||
AAAT-CTCTGAA
</pre></td>
<td>Or..? </td>
</tr>

<tr><td><pre>
AAACAGTCA-----GAA
|||-----------|||
AAA------TCTCTGAA
</pre></td>
<td> What about this? </td>
</tr>

</table>

---

# Sequence Alignment

```
Sequence 1: AAA CAGTGA GAA
Sequence 2: AAA TCTCT  GAA
```

<table style="width:100%; font-size:0.8em">
<th>Alignment</th><th></th><th></th>

<tr><td><pre>
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
</pre></td>
<td>Maybe like this?</td>
</tr>

<tr class="hidden"><td><pre>
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
</pre></td>
<td> Or this? </td>
</tr>

<tr class="hidden"><td><pre>
AAACAGTGAGAA
|||:-:|::|||
AAAT-CTCTGAA
</pre></td>
<td>Or..? </td>
</tr>

<tr class="hidden"><td><pre>
AAACAGTCA-----GAA
|||-----------|||
AAA------TCTCTGAA
</pre></td>
<td> What about this? </td>
</tr>

</table>


---

# Sequence Alignment

```
Sequence 1: AAA CAGTGA GAA
Sequence 2: AAA TCTCT  GAA
```

<table style="width:100%; font-size:0.8em;">
<th>Alignment</th><th></th><th></th>

<tr><td><pre>
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
</pre></td>
<td>Maybe like this?</td>
</tr>

<tr><td><pre>
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
</pre></td>
<td> Or this? </td>
</tr>

<tr class="hidden"><td><pre>
AAACAGTGAGAA
|||:-:|::|||
AAAT-CTCTGAA
</pre></td>
<td>Or..? </td>
</tr>

<tr class="hidden"><td><pre>
AAACAGTCA-----GAA
|||-----------|||
AAA------TCTCTGAA
</pre></td>
<td> What about this? </td>
</tr>

</table>

---

# Sequence Alignment

```
Sequence 1: AAA CAGTGA GAA
Sequence 2: AAA TCTCT  GAA
```

<table style="width:100%; font-size:0.8em">
<th>Alignment</th><th></th><th></th>

<tr><td><pre>
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
</pre></td>
<td>Maybe like this?</td>
</tr>

<tr><td><pre>
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
</pre></td>
<td> Or this? </td>
</tr>

<tr><td><pre>
AAACAGTGAGAA
|||:-:|::|||
AAAT-CTCTGAA
</pre></td>
<td>Or..? </td>
</tr>

<tr class="hidden"><td><pre>
AAACAGTCA-----GAA
|||-----------|||
AAA------TCTCTGAA
</pre></td>
<td> What about this? </td>
</tr>

</table>

---

# Sequence Alignment

```
Sequence 1: AAA CAGTGA GAA
Sequence 2: AAA TCTCT  GAA
```

<table style="width:100%; font-size:0.8em">
<th>Alignment</th><th></th><th></th>

<tr><td><pre>
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
</pre></td>
<td>Maybe like this?</td>
</tr>

<tr><td><pre>
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
</pre></td>
<td> Or this? </td>
</tr>

<tr><td><pre>
AAACAGTGAGAA
|||:-:|::|||
AAAT-CTCTGAA
</pre></td>
<td>Or..? </td>
</tr>

<tr><td><pre>
AAACAGTCA-----GAA
|||-----------|||
AAA------TCTCTGAA
</pre></td>
<td> What about this? </td>
</tr>
</table>

???
There is no one right way to do alignment
  - Hard to say which of these is "better" or "worse"
  - Just different choices, but all valid

Non-trivial problem

---

# Sequence Alignment

```
Sequence 1: AAA CAGTGA GAA
Sequence 2: AAA TCTCT  GAA
```

<table style="width:100%; font-size:0.8em">
<th>Alignment</th><th>Tool</th>

<tr><td><pre>
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
</pre></td>
<td>Novoalign</td>
</tr>

<tr><td><pre>
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
</pre></td>
<td> Ssaha2 </td>
</tr>

<tr><td><pre>
AAACAGTGAGAA
|||:-:|::|||
AAAT-CTCTGAA
</pre></td>
<td> BWA </td>
</tr>

<tr><td><pre>
AAACAGTCA-----GAA
|||-----------|||
AAA------TCTCTGAA
</pre></td>
<td> Complete Genomics </td>
</tr>
</table>


???
We didn't just make these up, different real aligners gave these different results

---

# Sequence Alignment

```
Sequence 1: AAA CAGTGA GAA
Sequence 2: AAA TCTCT  GAA
```

<table style="width:100%; font-size:0.8em">
<th>Alignment</th><th>Variants</th>

<tr><td><pre>
AAA-CAGTGAGAA
|||-|--|::|||
AAATC--TCTGAA
</pre></td>
<td><pre>
ins T
del AG
sub GA -> CT
</pre></td>
</tr>

<tr><td><pre>
AAACAGTGAGAA
|||-::|::|||
AAA-TCTCTGAA
</pre></td>
<td><pre>
del C
sub AG -> CT
sub GA -> CT
</pre></td>
</tr>

<tr><td><pre>
AAACAGTGAGAA
|||:-:|::|||
AAAT-CTCTGAA
</pre></td>
<td><pre>
snp C -> T
del A
snp G -> C
sub GA -> CT
</pre></td>
</tr>

<tr><td><pre>
AAACAGTCA-----GAA
|||-----------|||
AAA------TCTCTGAA
</pre></td>
<td><pre>
del CAGTGA
int TCTCT
</pre></td>
</tr>
</table>


???
**Important:** Mapping can affect downstream analysis, in this case
different alignments lead to very different variant calls


---

## Tools, tools, tools

- Many different mappers exist
  - Bowtie, HISAT, MAQ, BWA, Eland, SOAP, STAR, BISMARCK, Novoalign
  - Many more


- Each algorithm makes **different choices** during alignment


- Choice of aligner may **affect downstream results**
   - For example different variant calls (previous slide)


- Best tool for your data depends on many factors
  - Type of experiment
  - Sequencing platform
  - Compute resources vs sensitivity
  - Read characteristics

???
Choice of mapper depends on your experiment
 - Some mappers are good for DNA but not RNA
 - Some mappers do well in highly rearranged genomes, others less so (e.g. for cancer genomes)
 - Some mappers do well on some platforms but worse on others
 - ..

Or other factors
 - STAR needs a LOT of RAM
 - Do you need results fast? or accurate? (e.g. medical setting)

---

## Tools, tools, tools

- Over 60 different alignment tools
- Many comparison papers
- Feature comparison:  https://www.ebi.ac.uk/~nf/hts_mappers/

![timeline of mapping tools](../../images/mapping/ngs_read_mappers_timeline.jpeg)

---

# Try it yourself!

- Lego time! Who wants to volunteer?

- Or try this [online sequence alignment game](http://teacheng.illinois.edu/SequenceAlignment/):

.image-75[![recording of alignment game](../../images/mapping/alignment.gif)]

.footnote[https://tinyurl.com/sequence-alignment]


---

# SAM/BAM file format

![alt text](../../images/bam_file.png "SAM/BAM file")

---

# SAM/BAM file format

![alt text](../../images/bam_file_explained.png "SAM/BAM file detail")

SAM: Sequence Alignment/Map format

BAM: Binary SAM

---

# Genome Browser

- Visualise the alignments

![alt text](../../images/igv_mapping.png "IGV Browser")

---

# IGV Browser -- Hands-on

![alt text](../../images/igv_part2.png "igv detail")
