<!doctype html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="x-ua-compatible" content="ie=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        <title>Calling variants in non-diploid systems</title>
        
            <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
    ga('create', 'UA-45719423-18' , 'auto');
    ga('send', 'pageview');
</script>
        
        <link rel="stylesheet" href="/training-material/assets/css/bootstrap.min.css?v=3">
        <link rel="stylesheet" href="/training-material/assets/css/bootstrap-toc.min.css">
        <link rel="stylesheet" href="/training-material/assets/css/main.css?v=2">
        <script src="https://kit.fontawesome.com/67b3f98409.js" crossorigin="anonymous"></script>
        <link rel="stylesheet" href="/training-material/assets/css/academicons.css">
        <link rel="stylesheet" href="/training-material/assets/css/syntax_highlighting.css">
        <link rel="shortcut icon" href="/training-material/favicon.ico" type="image/x-icon" />

        
        
        
        
        
        <meta name="description" content="Genetic differences (variants) between healthy and diseas..." />
        <meta property="og:title" content="Galaxy Training: Calling variants in non-diploid systems" />
        <meta property="og:description" content="Genetic differences (variants) between healthy and diseas..." />
        <meta property="og:image" content="/training-material/assets/images/GTNLogo1000.png" />
    </head>
    <body data-spy="scroll" data-target="#toc">
        











<!-- Gitter -->


<script>
  ((window.gitter = {}).chat = {}).options = {
  room: 'Galaxy-Training-Network/Lobby'
  };
</script>
<script src="https://sidecar.gitter.im/dist/sidecar.v1.js" async defer></script>

<header>
    <nav class="navbar navbar-expand-lg navbar-dark">
        <div class="container">
            <a class="navbar-brand" href="/training-material/">
                <img src="/training-material/assets/images/GTN-60px.png" height="30" alt="Galaxy Training Network logo">
                Galaxy Training!
            </a>

            <button class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#top-navbar" aria-controls="top-navbar" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>
            <div class="collapse navbar-collapse" id="top-navbar">
                <ul class="navbar-nav">
                    <li class="nav-item">
                        <a class="nav-link" href="/training-material/topics/variant-analysis" title="Go back to list of tutorials">
                            <i class="far fa-folder" aria-hidden="true"></i><span class="visually-hidden">topic</span> Variant Analysis
                        </a>
                    </li>

                    <li class="nav-item dropdown">
                        <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Language Selector">
                            <i class="fas fa-language" aria-hidden="true"></i><span class="visually-hidden">language</span> Language
                        </a>
                        <div class="dropdown-menu">
                            
                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=fr&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Fvariant-analysis%2Ftutorials%2Fnon-dip%2Ftutorial.html&edit-text=&act=url" title="">
                                Français
                            </a>
                            
                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=ja&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Fvariant-analysis%2Ftutorials%2Fnon-dip%2Ftutorial.html&edit-text=&act=url" title="">
                                日本語
                            </a>
                            
                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=es&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Fvariant-analysis%2Ftutorials%2Fnon-dip%2Ftutorial.html&edit-text=&act=url" title="">
                                Español
                            </a>
                            
                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=pt&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Fvariant-analysis%2Ftutorials%2Fnon-dip%2Ftutorial.html&edit-text=&act=url" title="">
                                Português
                            </a>
                            
                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=ar&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Fvariant-analysis%2Ftutorials%2Fnon-dip%2Ftutorial.html&edit-text=&act=url" title="">
                                العربية
                            </a>
                            
                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Fvariant-analysis%2Ftutorials%2Fnon-dip%2Ftutorial.html&edit-text=&act=url" title="">
                                And more!
                            </a>
                        </div>
                    </li>

                    <li class="nav-item dropdown">
    <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Help">
        <i class="far fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">help</span> Help
    </a>
    <div class="dropdown-menu dropdown-menu-right">
        <form method="get" action="https://tess.elixir-europe.org/materials">
            <input type="text" id="search" name="q" value="" style="margin-left: 0.5em;/*! border-radius: 0px; */">
            <input type="hidden" value="Galaxy Training" name="content_provider">
            <input type="submit" value="Search on TeSS" style="width: 92%;border-radius: 0px;margin: 0.5em;background: #f47d20;border: 0px;padding: 0.25em;" class="">
        </form>

        <div class="dropdown-divider"></div>
        <a class="dropdown-item" href="/training-material/faq" title="Check our FAQ">
            FAQ
        </a>
        <a class="dropdown-item" href="https://help.galaxyproject.org/" title="Discuss on Galaxy Help">
            Discuss on Galaxy Help
        </a>

        <div class="dropdown-item">
            <div>
                Theme
            </div>

            <div id="theme-selector" data-toggle="buttons">
                <label data-value="default" class="btn btn-secondary">
                    <input type="radio" name="options" id="default" autocomplete="off"> Default
                </label>
                <label data-value="night" class="btn btn-secondary">
                    <input type="radio" name="options" id="night" autocomplete="off"> Night
                </label>
                <label data-value="midnight" class="btn btn-secondary">
                    <input type="radio" name="options" id="midnight" autocomplete="off"> Midnight
                </label>
                <label data-value="rainbow" class="btn btn-secondary">
                    <input type="radio" name="options" id="rainbow" autocomplete="off"> Rainbow
                </label>
            </div>

        </div>
    </div>
</li>


                    <li class="nav-item">
                        <a class="nav-link" href="https://github.com/galaxyproject/training-material/edit/master/topics/variant-analysis/tutorials/non-dip/tutorial.md">
                            <i class="fab fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> Edit
                        </a>
                    </li>
                </ul>
            </div>
        </div>
    </nav>
</header>

<div class="container main-content">
    <script type="application/ld+json">
        


{
  "@context": "http://schema.org",
  "@type": "Course",
  "accessMode": [
    "textual",
    "visual"
  ],
  "accessModeSufficient": [
    "textual",
    "visual"
  ],
  "accessibilityControl": [
    "fullKeyboardControl",
    "fullMouseControl"
  ],
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "accessibilitySummary": "Short descriptions are present but long descriptions will be needed for non-visual users",
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "citation": {
    "@type": "CreativeWork",
    "name": "Community-Driven Data Analysis Training for Biology",
    "url": "https://doi.org/10.1016/j.cels.2018.05.012"
  },
  "copyrightHolder": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "discussionUrl": "https://gitter.im/Galaxy-Training-Network/Lobby",
  "headline": "Calling variants in non-diploid systems",
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "https://github.com/galaxyproject/training-material/blob/master/LICENSE.md",
  "producer": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "provider": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "sourceOrganization": {
    "@type": "Organization",
    "email": "galaxytrainingnetwork@gmail.com",
    "name": "Galaxy Training Network",
    "url": "https://galaxyproject.org/teach/gtn/"
  },
  "isPartOf": {
    "@type": "CreativeWork",
    "name": "Variant Analysis",
    "description": "Genetic differences (variants) between healthy and diseased tissue,\nbetween individuals of a population, or between strains of an organism\ncan provide mechanistic insight into disease processes and the natural\nfunction of affected genes.\n\nThe tutorials in this section show how to detect evidence for\ngenetic variants in next-generation sequencing data, a process termed\nvariant calling.\n\nOf equal importance, they also demonstrate how you can interpret, for\na range of different organisms, the resulting sets of variants by\npredicting their molecular effects on genes and proteins, by\nannotating previously observed variants with published knowledge, and\nby trying to link phenotypes of the sequenced samples to their variant\ngenotypes.\n",
    "url": "https://training.galaxyproject.org//training-material/topics/variant-analysis/"
  },
  "courseCode": "variant-analysis / non-dip / hands-on",
  "learningResourceType": "hands-on tutorial",
  "name": "Hands-on for 'Calling variants in non-diploid systems' tutorial",
  "url": "https://training.galaxyproject.org//training-material/topics/variant-analysis/tutorials/non-dip/tutorial.html",
  "timeRequired": "PT1H30M",
  "keywords": "prokaryote",
  "description": "The questions this  addresses are:\n - How does frequency of mitochondrial polymorphisms change from mother to child?\n\n\\nThe objectives are:\n - Using Galaxy's main site we will see how to call variants in bacteria, viruses, and organelles.\n\n",
  "coursePrerequisites": [
    {
      "@type": "CreativeWork",
      "url": "https://training.galaxyproject.org//training-material/topics/introduction/",
      "name": "Introduction to Galaxy Analyses",
      "description": "Introduction to Galaxy Analyses",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    },
    {
      "@type": "Course",
      "url": "https://training.galaxyproject.org//training-material/topics/sequence-analysis/tutorials/quality-control/slides.html",
      "name": "Quality Control",
      "description": "Slides for 'Quality Control' tutorial",
      "learningResourceType": "slides",
      "interactivityType": "expositive",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    },
    {
      "@type": "Course",
      "url": "https://training.galaxyproject.org//training-material/topics/sequence-analysis/tutorials/quality-control/tutorial.html",
      "name": "Quality Control",
      "description": "Hands-on for 'Quality Control' tutorial",
      "learningResourceType": "hands-on tutorial",
      "interactivityType": "expositive",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    },
    {
      "@type": "Course",
      "url": "https://training.galaxyproject.org//training-material/topics/sequence-analysis/tutorials/mapping/slides.html",
      "name": "Mapping",
      "description": "Slides for 'Mapping' tutorial",
      "learningResourceType": "slides",
      "interactivityType": "expositive",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    },
    {
      "@type": "Course",
      "url": "https://training.galaxyproject.org//training-material/topics/sequence-analysis/tutorials/mapping/tutorial.html",
      "name": "Mapping",
      "description": "Hands-on for 'Mapping' tutorial",
      "learningResourceType": "hands-on tutorial",
      "interactivityType": "expositive",
      "provider": {
        "@type": "Organization",
        "email": "galaxytrainingnetwork@gmail.com",
        "name": "Galaxy Training Network",
        "url": "https://galaxyproject.org/teach/gtn/"
      }
    }
  ],
  "hasPart": [

  ],
  "author": [
    {
      "@type": "Person",
      "name": "Anton Nekrutenko"
    },
    {
      "@type": "Person",
      "name": "Alex Ostrovsky"
    }
  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "Anton Nekrutenko"
    },
    {
      "@type": "Person",
      "name": "Alex Ostrovsky"
    }
  ],
  "about": [
    {
      "@type": "CreativeWork",
      "name": "Variant Analysis",
      "description": "Genetic differences (variants) between healthy and diseased tissue,\nbetween individuals of a population, or between strains of an organism\ncan provide mechanistic insight into disease processes and the natural\nfunction of affected genes.\n\nThe tutorials in this section show how to detect evidence for\ngenetic variants in next-generation sequencing data, a process termed\nvariant calling.\n\nOf equal importance, they also demonstrate how you can interpret, for\na range of different organisms, the resulting sets of variants by\npredicting their molecular effects on genes and proteins, by\nannotating previously observed variants with published knowledge, and\nby trying to link phenotypes of the sequenced samples to their variant\ngenotypes.\n",
      "url": "https://training.galaxyproject.org//training-material/topics/variant-analysis/"
    },
    {
      "@type": "DefinedTerm",
      "@id": "http://edamontology.org/topic_0199",
      "inDefinedTermSet": "http://edamontology.org",
      "termCode": "topic_0199",
      "url": "https://bioportal.bioontology.org/ontologies/EDAM/?p=classes&conceptid=http%3A%2F%2Fedamontology.org%2Ftopic_0199"
    }
  ]
}
    </script>

    <section class="tutorial">
        <h1 data-toc-skip>Calling variants in non-diploid systems</h1>
        

        <div class="contributors-line">By: 

<a href="/training-material/hall-of-fame/nekrut/" class="contributor-badge"><img src="https://avatars.githubusercontent.com/nekrut" alt="Anton Nekrutenko">Anton Nekrutenko</a>, <a href="/training-material/hall-of-fame/astrovsky01/" class="contributor-badge"><img src="https://avatars.githubusercontent.com/astrovsky01" alt="Alex Ostrovsky">Alex Ostrovsky</a>

</div>

        <blockquote class="overview">
            <h3>Overview</h3>
            <strong><i class="far fa-question-circle" aria-hidden="true"></i><span class="visually-hidden">question</span> Questions</strong>
            <ul>
            
            <li><p>How does frequency of mitochondrial polymorphisms change from mother to child?</p>
</li>
            
            </ul>

            <strong><i class="fas fa-bullseye" aria-hidden="true"></i><span class="visually-hidden">objectives</span> Objectives</strong>
            <ul>
            
            <li><p>Using Galaxy’s main site we will see how to call variants in bacteria, viruses, and organelles.</p>
</li>
            
            </ul>

            
            <strong><i class="fas fa-check-circle" aria-hidden="true"></i><span class="visually-hidden">requirements</span> Requirements</strong>
            <ul>
            
    <li>
    
        
        
        <a href="/training-material/topics/introduction">Introduction to Galaxy Analyses</a>
        
    
    </li>

    <li>
    
        
        
        <a href="/training-material/topics/sequence-analysis">Sequence analysis</a>
        
            <ul>
                
                    
                        
                    
                        
                            
                            <li> Quality Control:
                            
                                <a href="/training-material/topics/sequence-analysis/tutorials/quality-control/slides.html"><i class="fab fa-slideshare" aria-hidden="true"></i><span class="visually-hidden">slides</span> slides</a>
                                
                            
                            
                                
                                    - <a href="/training-material/topics/sequence-analysis/tutorials/quality-control/tutorial.html"><i class="fas fa-laptop" aria-hidden="true"></i><span class="visually-hidden">tutorial</span> hands-on</a>
                                
                            
                            </li>
                        
                    
                
                    
                        
                            
                            <li> Mapping:
                            
                                <a href="/training-material/topics/sequence-analysis/tutorials/mapping/slides.html"><i class="fab fa-slideshare" aria-hidden="true"></i><span class="visually-hidden">slides</span> slides</a>
                                
                            
                            
                                
                                    - <a href="/training-material/topics/sequence-analysis/tutorials/mapping/tutorial.html"><i class="fas fa-laptop" aria-hidden="true"></i><span class="visually-hidden">tutorial</span> hands-on</a>
                                
                            
                            </li>
                        
                    
                        
                    
                
            </ul>
        
    
    </li>

            
            </ul>
            

            
            <p><strong><i class="fas fa-hourglass-half" aria-hidden="true"></i><span class="visually-hidden">time</span> Time estimation:</strong> 1 hour 30 minutes</p>
            

            

            
            

            
            <p id="supporting-materials"><strong><i class="fa fa-external-link" aria-hidden="true"></i> Supporting Materials</strong></p>
            <ul>
                <div class="supporting_material">
                
                    <li class="btn btn-default supporting_material"><a href="/training-material/topics/variant-analysis/slides/introduction.html" title="Introduction slides">
                        <i class="fab fa-slideshare" aria-hidden="true"></i><span class="visually-hidden">slides</span> Introduction slides
                    </a></li>
                

                
                    <li class="btn btn-default supporting_material">
<a class="topic-icon" href="https://doi.org/10.5281/zenodo.1251112">
    <i class="far fa-copy" aria-hidden="true"></i><span class="visually-hidden">zenodo_link</span> Datasets
</a>

</li>
                

                
                    <li class="btn btn-default supporting_material">
    <a class="topic-icon" href="/training-material/topics/variant-analysis/tutorials/non-dip/workflows/" title="Workflows" alt="Calling variants in non-diploid systems workflows">
        <i class="fas fa-share-alt" aria-hidden="true"></i><span class="visually-hidden">workflow</span> Workflows
    </a>

</li>
                

                
                    <li class="btn btn-default supporting_material">
<a class="topic-icon" href="https://github.com/galaxyproject/training-material/tree/master/topics/variant-analysis/tutorials/non-dip/tours/">
    <i class="fas fa-magic" aria-hidden="true"></i><span class="visually-hidden">interactive_tour</span> Tours
</a>

</li>
                

                
                    <li class="btn btn-default supporting_material">

    <a href="#" class="btn btn-default dropdown-toggle topic-icon" data-toggle="dropdown" aria-expanded="false" title="Where to run the tutorial">
        <i class="fas fa-globe" aria-hidden="true"></i><span class="visually-hidden">instances</span> Available on these Galaxies
    </a>
    <ul class="dropdown-menu">
    
        <li>
            <a class="dropdown-item" href="https://github.com/galaxyproject/training-material/tree/master/topics/variant-analysis/docker" title="Docker image for this tutorial">
                <i class="fab fa-docker" aria-hidden="true"></i><span class="visually-hidden">docker_image</span> Docker image
            </a>
        </li>
    
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
        <a class="dropdown-item" href="https://test.galaxyproject.org/" title="Galaxy Test">
            Galaxy Test
        </a>
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
    
        
        <a class="dropdown-item" href="https://usegalaxy.be/" title="UseGalaxy.be">
            UseGalaxy.be
        </a>
        
    
        
        <a class="dropdown-item" href="https://usegalaxy.eu" title="UseGalaxy.eu">
            UseGalaxy.eu
        </a>
        
    
        
        <a class="dropdown-item" href="https://usegalaxy.fr/" title="UseGalaxy.fr">
            UseGalaxy.fr
        </a>
        
    
        
        <a class="dropdown-item" href="https://usegalaxy.org" title="UseGalaxy.org (Main)">
            UseGalaxy.org (Main)
        </a>
        
    
        
        <a class="dropdown-item" href="https://usegalaxy.org.au" title="UseGalaxy.org.au">
            UseGalaxy.org.au
        </a>
        
    
        
    
        
    
    </ul>


</li>
                
                </div>
            </ul>
            

            <p><strong> <i class="far fa-calendar" aria-hidden="true"></i><span class="visually-hidden">last_modification</span> Last modification:</strong> Jul 28, 2020 </p>
        </blockquote>

        <div class="container">
            <div class="row">
                <!-- sidebar, which will move to the top on a small screen -->
                <div class="col-sm-2">
                    <nav id="toc" data-toggle="toc" class="sticky-top"></nav>
                </div>
                <div class="col-sm-10">
                    <h1 id="introduction">Introduction</h1>

<p>The majority of life on Earth is non-diploid and represented by prokaryotes, viruses, and their derivatives, such as our own mitochondria or plant’s chloroplasts. In non-diploid systems, allele frequencies can range anywhere between 0 and 100% and there could be multiple (not just two) alleles per locus. The main challenge associated with non-diploid variant calling is the difficulty in distinguishing between the sequencing noise (abundant in all NGS platforms) and true low frequency variants. Some of the early attempts to do this well have been accomplished on human mitochondrial <span class="notranslate">DNA</span> although the same approaches will work equally good on viral and bacterial genomes:</p>

<ul>
  <li>2014 - <a href="http://www.pnas.org/content/111/43/15474.abstract">Mate<span class="notranslate">rna</span>l age effect and severe germ-line bottleneck in the inheritance of human mitochondrial <span class="notranslate">DNA</span></a></li>
  <li>2015 - <a href="http://www.pnas.org/content/112/8/2491.abstract">Extensive tissue-related and allele-related mt<span class="notranslate">DNA</span> heteroplasmy suggests positive selection for somatic mutations</a>.</li>
</ul>

<p>As an example of non-diploid systems, we will be using human mitochondrial genome. However, this approach will also work for most bacterial and viral genomes.</p>

<p>There are two ways one can call variants:</p>

<ol>
  <li>By comparing <span class="notranslate">reads</span> against an existing genome assembly</li>
  <li>By assembling a genome first and then <span class="notranslate">mapping</span> against that assembly</li>
</ol>

<table>
  <thead>
    <tr>
      <th> </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><img src="../../images/ref_vs_assembly.jpg" alt="2 approaches" /></td>
    </tr>
    <tr>
      <td><small>This figure from a manuscript by <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4493402/">Olson:2015</a> contrasts the two approaches.</small></td>
    </tr>
  </tbody>
</table>

<p>In this tutorials we will take the <em>first</em> path, in which we map <span class="notranslate">reads</span> against an existing assembly. Later in the course (after we learn about assembly approaches) we will try the second approach as well.</p>

<p>The goal of this example is to detect heteroplasmies (variants within mitochondrial <span class="notranslate">DNA</span>). Mitochondria are transmitted mate<span class="notranslate">rna</span>lly, and heteroplasmy frequencies may change dramatically and unpredictably during the transmission due to a germ-line bottleneck <a href="https://www.nature.com/ng/journal/v40/n2/abs/ng.2007.63.html">Cree:2008</a>. As we mentioned above, the procedure for finding variants in bacterial or viral genomes will be essentially the same.</p>

<p>Zenodo contains <a href="https://doi.org/10.5281/zenodo.1251112">datasets representing a child and a mother</a>. These datasets were obtained by paired-end Illumina sequencing of human genomic <span class="notranslate">DNA</span> enriched for mitochondria. The enrichment was performed using long-range PCR with two primer pairs that amplify the entire mitochondrial genome. Samples will therefore still contain a lot of <span class="notranslate">DNA</span> from the nuclear genome, which, in this case, is a contaminant.</p>

<h1 id="importing-example-datasets">Importing example datasets</h1>

<p>For this tutorial we have prepared a subset of data previously <a href="http://www.pnas.org/content/111/43/15474.abstract">published</a> by our group. Let’s import these data into <span class="notranslate">Galaxy</span>. They are available from <a href="https://usegalaxy.org/library/list#folders/Fe4842bd0c37b03a7">this <span class="notranslate">Galaxy</span> Library</a> or via <a href="https://zenodo.org/record/582600">Zenodo</a></p>

<blockquote class="notranslate hands_on">
  <h3 id="hands_on-hands-on-getting-the-data"><i class="fas fa-pencil-alt" aria-hidden="true"></i><span class="visually-hidden">hands_on</span> Hands-on: Getting the data</h3>

  <p>Option 1: Data upload from a <span class="notranslate">Galaxy</span> Library</p>

  <ul>
    <li>
      <p>Create and name a new history for this tutorial.</p>

      <blockquote class="tip">

        <h3 id="tip-tip-creating-a-new-history"><i class="far fa-lightbulb" aria-hidden="true"></i><span class="visually-hidden">tip</span> Tip: Creating a new history</h3>

        <p>Click the <i class="fas fa-plus" aria-hidden="true"></i><span class="visually-hidden">new-history</span> icon at the top of the history panel</p>

        <p>If the <i class="fas fa-plus" aria-hidden="true"></i><span class="visually-hidden">new-history</span> is missing:</p>
        <ol>
          <li>Click on the <i class="fas fa-cog" aria-hidden="true"></i><span class="visually-hidden"><span class="notranslate">galaxy</span>-gear</span> icon (<strong>History options</strong>) on the top of the history panel</li>
          <li>Select the option <strong>Create New</strong> from the menu</li>
        </ol>
      </blockquote>
    </li>
    <li>
      <p>Import from <a href="https://zenodo.org/record/1251112">Zenodo</a> or from the data library the files:</p>

      <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>https://zenodo.org/record/1251112/files/raw_child-ds-1.fq
https://zenodo.org/record/1251112/files/raw_child-ds-2.fq
https://zenodo.org/record/1251112/files/raw_mother-ds-1.fq
https://zenodo.org/record/1251112/files/raw_mother-ds-2.fq
</code></pre></div>      </div>
    </li>
  </ul>

  <p>Option 2: Data upload from a <span class="notranslate">Galaxy</span> Library</p>

  <p><img src="../../images/mt_lib.png" alt="Data upload from a Galaxy Library" /></p>

  <ul>
    <li>Go to this <a href="https://usegalaxy.org/library/list#folders/Fe4842bd0c37b03a7">this <span class="notranslate">Galaxy</span> library</a></li>
    <li>You will see screen like the one shown above</li>
    <li>Click <strong>to History</strong> button.</li>
    <li><span class="notranslate">Galaxy</span> will prompt you to ask whether you want to import these data into already existing or new history.</li>
    <li>It is better to create a new history, so type some descriptive name within <code class="language-plaintext highlighter-rouge">or create new</code> text field</li>
    <li>Click <strong>Import</strong></li>
    <li>A green message will appear once the import is done. Click on it and will see the history you have just created. It will be populated with the four datasets as shown below:</li>
  </ul>

  <p><img src="../../images/mt_imported_data.png" alt="Imported data in history" /></p>

</blockquote>

<h1 id="qcing-the-data">QC’ing the data</h1>

<p>Before proceeding with the analysis, we need to find out how good the data actually is. For this will use <code class="language-plaintext highlighter-rouge">FastQC</code> tool that can be found in <strong>NGS: QC and manipulation → FastQC</strong> section of <span class="notranslate">Galaxy</span> tools:</p>

<blockquote class="notranslate hands_on">
  <h3 id="quality-control-of-the-data">Quality Control of the data</h3>

  <p><img src="../../images/mt_qc.png" alt="FastQC input and dependencies" /></p>

  <p>QC’ing <span class="notranslate">reads</span> using <a href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>. Note that we selected all four datasets at once by pressing the middle button <img src="../../images/mt_middle_button.png" alt="Middle button" /> adjacent to the <strong>Short read data from your current history</strong> widget. Once <code class="language-plaintext highlighter-rouge">FastQC</code> job runs, you will be able to look at the HTML reports generated by this tool.</p>

  <p>The data have generally high quality in this example:</p>

  <p><img src="../../images/mt_qc_plot.png" alt="FastQC output plot" /></p>

  <p>FastQC plot for one of the mitochondrial datasets shows that qualities are acceptable for 250 bp <span class="notranslate">reads</span> (mostly in the green, which is at or above <a href="https://en.wikipedia.org/wiki/Phred_quality_score">phred score</a> of 30).</p>
</blockquote>

<h1 id="mapping-the-reads"><span class="notranslate">Mapping</span> the <span class="notranslate">reads</span></h1>

<p>Our <span class="notranslate">reads</span> are long (250 bp) and as a result we will be using <a href="https://arxiv.org/pdf/1303.3997v2.pdf">bwa mem</a> to align them against the reference genome as it has good <span class="notranslate">mapping</span> performance for longer <span class="notranslate">reads</span> (100bp and up).</p>

<blockquote class="notranslate hands_on">
  <h3 id="mapping-with-bwa-mem"><span class="notranslate">Mapping</span> with <code class="language-plaintext highlighter-rouge">bwa mem</code></h3>

  <p><img src="../../images/mt_bwa_mem.png" alt="bwa mem input and parameters" /></p>

  <p>Running <code class="language-plaintext highlighter-rouge">bwa mem</code> on our datasets. Look <strong>carefully</strong> at parameter settings:</p>

  <ul>
    <li>We select <code class="language-plaintext highlighter-rouge">hg38</code> version of the human genome as the reference</li>
    <li>By using the middle button again <img src="../../images/mt_middle_button.png" alt="middle button" /> we select datasets 1 and 3 as <strong>Select the first set of <span class="notranslate">reads</span></strong> and datasets 2 and 4 as <strong>Select the second set of <span class="notranslate">reads</span></strong>. <span class="notranslate">Galaxy</span> will automatically launch two bwa-mem jobs using datasets 1,2 and 3,4 generating two resulting BAM files.</li>
    <li>By setting <strong>Set read groups information</strong> to <code class="language-plaintext highlighter-rouge">Set read groups (SAM/BAM specifications)</code> and clicking <strong>Auto-assign</strong> we will ensure that the <span class="notranslate">reads</span> in the resulting BAM dataset are properly set.</li>
  </ul>
</blockquote>

<h1 id="merging-bam-datasets">Merging BAM datasets</h1>

<p>Because we have set read groups, we can now merge the two BAM dataset into one. This is because read groups label each read as belonging to either <em>mother</em> or <em>child</em>.</p>

<p>We can BAM dataset using <strong>NGS: Picard</strong> → <strong>MergeSAMFiles</strong> tool:</p>

<blockquote class="notranslate hands_on">
  <h3 id="merging-multiple-datasets-into-one">Merging multiple datasets into one</h3>

  <p><img src="../../images/mt_bam_merging.png" alt="MergeSAMFiles input and parameters" /></p>

  <p>Merging two BAM datasets into one. Note that two inputs are highlighted.</p>
</blockquote>

<h2 id="removing-duplicates">Removing duplicates</h2>

<p>Preparation of sequencing libraries (at least at the time of writing) for technologies such as Illumina (used in this example) involves PCR amplification. It is required to generate sufficient number of sequencing templates so that a reliable detection can be performed by base callers. Yet PCR has its own biases, which are especially profound in cases of multi-template PCR used for construction of sequencing libraries (Kanagawa et al. <a href="https://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&amp;db=PubMed&amp;dopt=Abstract&amp;list_uids=16233530">2003</a>).</p>

<p>Duplicates can be identified based on their outer alignment coordinates or using sequence-based clustering. One of the common ways for identification of duplicate <span class="notranslate">reads</span> is the <code class="language-plaintext highlighter-rouge">MarkDuplicates</code> utility from <a href="https://broadinstitute.github.io/picard/command-line-overview.html">Picard</a> package. It is designed to identify both PCR and optical duplicates (the following is an excerpt from Picard documentation):</p>

<p><em>Duplicates are identified as read pairs having identical 5’ positions (coordinate and strand) for both <span class="notranslate">reads</span> in a mate pair (and optionally, matching unique molecular identifier <span class="notranslate">reads</span>; see BARCODE_TAG option). Optical, or more broadly Sequencing, duplicates are duplicates that appear clustered together spatially during sequencing and can arise from optical/imagine-processing artifacts or from bio-chemical processes during clonal amplification and sequencing; they are identified using the READ_NAME_REGEX and the OPTICAL_DUPLICATE_PIXEL_DISTANCE options. The tool’s main output is a new SAM or BAM file in which duplicates have been identified in the SAM flags field, or optionally removed (see REMOVE_DUPLICATE and REMOVE_SEQUENCING_DUPLICATES), and optionally marked with a duplicate type in the ‘DT’ optional attribute. In addition, it also outputs a metrics file containing the numbers of READ_PAIRS_EXAMINED, SECONDARY_OR_SUPPLEMENTARY_RDS, UNMAPPED_<span class="notranslate">READS</span>, UNPAIRED_<span class="notranslate">READS</span>, UNPAIRED_READ DUPLICATES, READ_PAIR_DUPLICATES, and READ_PAIR_OPTICAL_DUPLICATES. Usage example: java -jar picard.jar MarkDuplicates I=input.bam \ O=marked_duplicates.bam M=marked_dup_metrics.txt.</em></p>

<p>Let’s use <strong>NGS: Picard</strong> → <strong>MarkDuplicates</strong> tool:</p>

<blockquote class="notranslate hands_on">
  <h3 id="de-duplicating-mapped-data">De-duplicating mapped data</h3>

  <p><img src="../../images/mt_dedup.png" alt="MarkDuplicates tool parameters" /></p>

  <p>De-duplicating the merged BAM dataset</p>
</blockquote>

<p><strong>MarkDuplicates</strong> produces a BAM dataset with duplicates removed and also a metrics file. Let’s take a look at the metrics data:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>raw_child-ds-	55	27551	849	55	50	1658	1	0.061026	219750
raw_mother-ds-	95	54973	1951	95	89	4712	2	0.08645	302188
</code></pre></div></div>

<p>where columns are:</p>

<ul>
  <li>LIBRARY (read group in our case)</li>
  <li>UNPAIRED_<span class="notranslate">READS</span>_EXAMINED</li>
  <li>READ_PAIRS_EXAMINED-</li>
  <li>SECONDARY_OR_SUPPLEMENTARY_RDS</li>
  <li>UNMAPPED_<span class="notranslate">READS</span></li>
  <li>UNPAIRED_READ_DUPLICATES</li>
  <li>READ_PAIR_DUPLICATES</li>
  <li>READ_PAIR_OPTICAL_DUPLICATES</li>
  <li>PERCENT_DUPLICATION</li>
  <li>ESTIMATED_LIBRARY_SIZE</li>
</ul>

<p>In other words the two datasets had ~6% and ~9% duplicates, respectively.</p>

<h1 id="left-aligning-indels">Left-aligning indels</h1>

<p>Left aligning of indels (a variant of re-aligning) is extremely important for obtaining accurate variant calls. This concept, while not difficult, requires some explanation. For illustrating how left-aligning works we expanded on an example provided by <a href="https://academic.oup.com/bioinformatics/article/31/13/2202/196142/Unified-representation-of-genetic-variants">Tan:2015</a>. Suppose you have a reference sequence and a sequencing read:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Reference GGGCACACACAGGG
Read      GGGCACACAGGG
</code></pre></div></div>

<p>If you look carefully you will see that the read is simply missing a <code class="language-plaintext highlighter-rouge">CA</code> repeat. But it is not apparent to a mapper, so some of possible alignments and corresponding variant calls include:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Alignment                 Variant Call

GGGCACACACAGGG            Ref: CAC
GGGCAC--ACAGGG            Alt: C

GGGCACACACAGGG            Ref: ACA
GGGCA--CACAGGG            Alt: A

GGGCACACACAGGG            Ref: GCA
GGG--CACACAGGG            Alt: G
</code></pre></div></div>

<p>The last of these is <em>left-aligned</em>. In this case gaps (dashes) as moved as far left as possible (for a formal definition of left-alignment and variant normalization see <a href="https://bioinformatics.oxfordjournals.org/content/31/13/2202.abstract">Tan:2015</a>).</p>

<p>Let’s perform left alignment using <strong>NGS: Variant Analysis</strong> → <strong>BamLeftAlign</strong>:</p>

<blockquote class="notranslate hands_on">
  <h3 id="left-aligning-indels-1">Left-aligning indels</h3>

  <p><img src="../../images/mt_left_align.png" alt="BamLeftAlign parameters" /></p>

  <p>Left-aligning a de-duplicated BAM dataset</p>
</blockquote>

<h1 id="filtering-reads">Filtering <span class="notranslate">reads</span></h1>

<p>Remember that we are trying to call variants in mitochondrial genome. Let focus only on the <span class="notranslate">reads</span> derived from mitochondria genome by filtering everything else out. For this we will use <strong>NGS: BamTools</strong> → <strong>Filter</strong>:</p>

<blockquote class="notranslate hands_on">
  <h3 id="filtering-bam-data">Filtering BAM data</h3>

  <p><img src="../../images/mt_filtering.png" alt="Filtering BAM datasets tool parameters" /></p>

  <p>Filtering <span class="notranslate">reads</span>. There are several important point to note here:</p>

  <ul>
    <li><strong>mapQuality</strong> is set to ≥20. <span class="notranslate">Mapping</span> quality reflects the probability that the read is placed <em>incorrectly</em>. It uses <a href="https://en.wikipedia.org/wiki/Phred_quality_score">phred scale</a>. Thus 20 is 1/100 or 1% chance that the read is incorrectly mapped. By setting this parameter to ≥20, we will keep all <span class="notranslate">reads</span> that have 1% or less probability of being mapped incorrectly.</li>
    <li><em>isPaired</em> will eliminate singleton (unpaired) <span class="notranslate">reads</span> (make sure <strong>Yes</strong> is clicked on)</li>
    <li><em>isProperPair</em> will only keep <span class="notranslate">reads</span> that map to the same chromosome and are properly placed (again, make sure <strong>Yes</strong> is clicked)</li>
    <li><em>reference</em> is set to <em>chrM</em></li>
  </ul>
</blockquote>

<h1 id="calling-non-diploid-variants-with-freebayes">Calling non-diploid variants with FreeBayes</h1>

<p>FreeBayes is widely used for calling variants in diploid systems. However, it can also be used for calling variants in pooled samples where the number of samples is not known. This is the exact scenario we have here: in our sample we have multiple mitochondrial (or bacterial or viral) genomes, but we do not know exactly how many. Thus we will use the <code class="language-plaintext highlighter-rouge">--pooled-continuous</code> option of FreeBayes to generate <em>frequency-based</em> variant calls as well as some other options highlighted below (the tool is in <strong>NGS: Variant Analysis</strong> → <strong>FreeBayes</strong>):</p>

<blockquote class="notranslate hands_on">
  <h3 id="running-freebayes">Running <code class="language-plaintext highlighter-rouge">FreeBayes</code></h3>

  <p><img src="../../images/mt_freebayes_genome.png" alt="FreeBayes parameters" /></p>

  <p>Set genome to <code class="language-plaintext highlighter-rouge">hg38</code> (the latest version)</p>

  <p><img src="../../images/mt_freebayes_regions.png" alt="Genome parameter" /></p>

  <p>Set regions to <code class="language-plaintext highlighter-rouge">chrM</code> from <code class="language-plaintext highlighter-rouge">1</code> to <code class="language-plaintext highlighter-rouge">16000</code>. This will simply save us time since we are only interested in mitochondrial variants anyway</p>

  <p><img src="../../images/mt_freebayes_alloptions.png" alt="regions parameter" /></p>

  <p>Choose <code class="language-plaintext highlighter-rouge">Complete list of all samples</code> from <strong>Choose parameter selection level</strong> drop down.</p>

  <p><img src="../../images/mt_freebayes_popmodel.png" alt="Complete list of all samples" /></p>

  <p>This is one of the most important parameter choices one needs to make when calling variants in non-diploid systems. Here set <strong>Set population model</strong> to <code class="language-plaintext highlighter-rouge">Yes</code> and then:</p>

  <ul>
    <li>Set <strong>ploidy</strong> to <code class="language-plaintext highlighter-rouge">1</code></li>
    <li>Set <strong>Assume that samples result from pooled sequencing</strong> to <code class="language-plaintext highlighter-rouge">Yes</code></li>
    <li>Set <strong>Output all alleles which pass input filters, regardless of genotyping outcome or model</strong> to <code class="language-plaintext highlighter-rouge">Yes</code></li>
  </ul>

  <p><img src="../../images/mt_freebayes_allelic_scope.png" alt="Allelic scope parameter" /></p>

  <p>We will also set <strong>Allelic scope</strong> to <code class="language-plaintext highlighter-rouge">Yes</code> and restrict variant types to single nucleotide polymorphisms only by:</p>

  <ul>
    <li>Keeping <strong>Ignore SNP alleles</strong> and <strong>Ignore indels alleles</strong> set to <code class="language-plaintext highlighter-rouge">No</code></li>
    <li>Setting <strong>Ignore MNPs</strong> and <strong>Ignore complex events</strong> to <code class="language-plaintext highlighter-rouge">Yes</code></li>
  </ul>

  <p>Mitochondria has a number of low complexity regions (mononucleotide repeats). Setting these parameters as described above will decrease noise from these regions.</p>

  <p><img src="../../images/mt_freebayes_inputfilters.png" alt="Additional parameters" /></p>

  <p>Finally, let’s set <strong>Input filters</strong> to <code class="language-plaintext highlighter-rouge">Yes</code> and set:</p>

  <ul>
    <li><strong>Exclude alignments from analysis if they have a <span class="notranslate">mapping</span> quality less than</strong> to <code class="language-plaintext highlighter-rouge">20</code> (phred score of 20). This will make FreeBayes to only consider reliably aligned <span class="notranslate">reads</span>.</li>
    <li><strong>Exclude alleles from analysis if their supporting base quality less than</strong> to <code class="language-plaintext highlighter-rouge">30</code> (phred score of 30). This will make FreeBayes to only consider high quality bases.</li>
  </ul>
</blockquote>

<p>This will produce a <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF dataset</a> shows below (you may need to scroll sideways to see it in full). It lists 25 sites of interest (everything starting with a <code class="language-plaintext highlighter-rouge">#</code> is a comment):</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Chrom	Pos	ID	Ref	Alt	Qual	Filter	Info	Format	data
##fileformat=VCFv4.2
##fileDate=20200211
##source=freeBayes v1.1.0-46-g8d2b3a0-dirty
##reference=/<span class="notranslate">galaxy</span>/data/hg38/sam_index/hg38.fa
##contig=&lt;ID=chr1,length=248956422&gt;
##contig=&lt;ID=chr2,length=242193529&gt;
##contig=&lt;ID=chr3,length=198295559&gt;
##contig=&lt;ID=chr4,length=190214555&gt;
##contig=&lt;ID=chr5,length=181538259&gt;
##contig=&lt;ID=chr6,length=170805979&gt;
##contig=&lt;ID=chr7,length=159345973&gt;
##contig=&lt;ID=chr8,length=145138636&gt;
##contig=&lt;ID=chr9,length=138394717&gt;
##contig=&lt;ID=chr10,length=133797422&gt;
##contig=&lt;ID=chr11,length=135086622&gt;
##contig=&lt;ID=chr12,length=133275309&gt;
##contig=&lt;ID=chr13,length=114364328&gt;
##contig=&lt;ID=chr14,length=107043718&gt;
##contig=&lt;ID=chr15,length=101991189&gt;
##contig=&lt;ID=chr16,length=90338345&gt;
##contig=&lt;ID=chr17,length=83257441&gt;
##contig=&lt;ID=chr18,length=80373285&gt;
##contig=&lt;ID=chr19,length=58617616&gt;
##contig=&lt;ID=chr20,length=64444167&gt;
##contig=&lt;ID=chr21,length=46709983&gt;
##contig=&lt;ID=chr22,length=50818468&gt;
##contig=&lt;ID=chrX,length=156040895&gt;
##contig=&lt;ID=chrY,length=57227415&gt;
##contig=&lt;ID=chrM,length=16569&gt;
##phasing=none
##commandline="freebayes --region chrM:1..16000 --bam b_0.bam --fasta-reference /<span class="notranslate">galaxy</span>/data/hg38/sam_index/hg38.fa --vcf ./vcf_output/part_chrM:1..16000.vcf --theta 0.001 --ploidy 1 -J -K -X -u -n 0 --haplotype-length 3 --min-repeat-size 5 --min-repeat-entropy 0 -m 20 -q 30 -R 0 -Y 0 -e 1000 -F 0.2 -C 2 -G 1 --min-coverage 0 --min-alte<span class="notranslate">rna</span>te-qsum 0"
##INFO=&lt;ID=NS,Number=1,Type=Integer,Description="Number of samples with data"&gt;
##INFO=&lt;ID=DP,Number=1,Type=Integer,Description="Total read depth at the locus"&gt;
##INFO=&lt;ID=DPB,Number=1,Type=Float,Description="Total read depth per bp at the locus; bases in <span class="notranslate">reads</span> overlapping / bases in haplotype"&gt;
##INFO=&lt;ID=AC,Number=A,Type=Integer,Description="Total number of alte<span class="notranslate">rna</span>te alleles in called genotypes"&gt;
##INFO=&lt;ID=AN,Number=1,Type=Integer,Description="Total number of alleles in called genotypes"&gt;
##INFO=&lt;ID=AF,Number=A,Type=Float,Description="Estimated allele frequency in the range (0,1]"&gt;
##INFO=&lt;ID=RO,Number=1,Type=Integer,Description="Count of full observations of the reference haplotype."&gt;
##INFO=&lt;ID=AO,Number=A,Type=Integer,Description="Count of full observations of this alte<span class="notranslate">rna</span>te haplotype."&gt;
##INFO=&lt;ID=PRO,Number=1,Type=Float,Description="Reference allele observation count, with partial observations recorded fractionally"&gt;
##INFO=&lt;ID=PAO,Number=A,Type=Float,Description="Alte<span class="notranslate">rna</span>te allele observations, with partial observations recorded fractionally"&gt;
##INFO=&lt;ID=QR,Number=1,Type=Integer,Description="Reference allele quality sum in phred"&gt;
##INFO=&lt;ID=QA,Number=A,Type=Integer,Description="Alte<span class="notranslate">rna</span>te allele quality sum in phred"&gt;
##INFO=&lt;ID=PQR,Number=1,Type=Float,Description="Reference allele quality sum in phred for partial observations"&gt;
##INFO=&lt;ID=PQA,Number=A,Type=Float,Description="Alte<span class="notranslate">rna</span>te allele quality sum in phred for partial observations"&gt;
##INFO=&lt;ID=SRF,Number=1,Type=Integer,Description="Number of reference observations on the forward strand"&gt;
##INFO=&lt;ID=SRR,Number=1,Type=Integer,Description="Number of reference observations on the reverse strand"&gt;
##INFO=&lt;ID=SAF,Number=A,Type=Integer,Description="Number of alte<span class="notranslate">rna</span>te observations on the forward strand"&gt;
##INFO=&lt;ID=SAR,Number=A,Type=Integer,Description="Number of alte<span class="notranslate">rna</span>te observations on the reverse strand"&gt;
##INFO=&lt;ID=SRP,Number=1,Type=Float,Description="Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality"&gt;
##INFO=&lt;ID=SAP,Number=A,Type=Float,Description="Strand balance probability for the alte<span class="notranslate">rna</span>te allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality"&gt;
##INFO=&lt;ID=AB,Number=A,Type=Float,Description="Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of <span class="notranslate">reads</span> showing the reference allele to all <span class="notranslate">reads</span>, considering only <span class="notranslate">reads</span> from individuals called as heterozygous"&gt;
##INFO=&lt;ID=ABP,Number=A,Type=Float,Description="Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality"&gt;
##INFO=&lt;ID=RUN,Number=A,Type=Integer,Description="Run length: the number of consecutive repeats of the alte<span class="notranslate">rna</span>te allele in the reference genome"&gt;
##INFO=&lt;ID=RPP,Number=A,Type=Float,Description="Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality"&gt;
##INFO=&lt;ID=RPPR,Number=1,Type=Float,Description="Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality"&gt;
##INFO=&lt;ID=RPL,Number=A,Type=Float,Description="<span class="notranslate">Reads</span> Placed Left: number of <span class="notranslate">reads</span> supporting the alte<span class="notranslate">rna</span>te balanced to the left (5') of the alte<span class="notranslate">rna</span>te allele"&gt;
##INFO=&lt;ID=RPR,Number=A,Type=Float,Description="<span class="notranslate">Reads</span> Placed Right: number of <span class="notranslate">reads</span> supporting the alte<span class="notranslate">rna</span>te balanced to the right (3') of the alte<span class="notranslate">rna</span>te allele"&gt;
##INFO=&lt;ID=EPP,Number=A,Type=Float,Description="End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality"&gt;
##INFO=&lt;ID=EPPR,Number=1,Type=Float,Description="End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality"&gt;
##INFO=&lt;ID=DPRA,Number=A,Type=Float,Description="Alte<span class="notranslate">rna</span>te allele depth ratio. Ratio between depth in samples with each called alte<span class="notranslate">rna</span>te allele and those without."&gt;
##INFO=&lt;ID=ODDS,Number=1,Type=Float,Description="The log odds ratio of the best genotype combination to the second-best."&gt;
##INFO=&lt;ID=GTI,Number=1,Type=Integer,Description="Number of genotyping iterations required to reach convergence or bailout."&gt;
##INFO=&lt;ID=TYPE,Number=A,Type=String,Description="The	type	of	allele,	either	snp,	mnp,	ins,	del,	or	complex."&gt;
##INFO=&lt;ID=CIGAR,Number=A,Type=String,Description="The extended CIGAR representation of each alte<span class="notranslate">rna</span>te allele, with the exception that '=' is replaced by 'M' to ease VCF parsing. Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR."&gt;
##INFO=&lt;ID=NUMALT,Number=1,Type=Integer,Description="Number	of	unique	non-reference	alleles	in	called	genotypes	at	this	position."&gt;
##INFO=&lt;ID=MEANALT,Number=A,Type=Float,Description="Mean number of unique non-reference allele observations per sample with the corresponding alte<span class="notranslate">rna</span>te alleles."&gt;
##INFO=&lt;ID=LEN,Number=A,Type=Integer,Description="allele length"&gt;
##INFO=&lt;ID=MQM,Number=A,Type=Float,Description="Mean <span class="notranslate">mapping</span> quality of observed alte<span class="notranslate">rna</span>te alleles"&gt;
##INFO=&lt;ID=MQMR,Number=1,Type=Float,Description="Mean <span class="notranslate">mapping</span> quality of observed reference alleles"&gt;
##INFO=&lt;ID=PAIRED,Number=A,Type=Float,Description="Proportion of observed alte<span class="notranslate">rna</span>te alleles which are supported by properly paired read fragments"&gt;
##INFO=&lt;ID=PAIREDR,Number=1,Type=Float,Description="Proportion of observed reference alleles which are supported by properly paired read fragments"&gt;
##INFO=&lt;ID=MIN_DP,Number=1,Type=Integer,Description="Minimum depth in gVCF output block."&gt;
##INFO=&lt;ID=END,Number=1,Type=Integer,Description="Last position (inclusive) in gVCF output record."&gt;
##INFO=&lt;ID=technology.ILLUMINA,Number=A,Type=Float,Description="Fraction	of	observations	supporting	the	alte<span class="notranslate">rna</span>te	observed	in	<span class="notranslate">reads</span>	from	ILLUMINA"&gt;
##FORMAT=&lt;ID=GT,Number=1,Type=String,Description="Genotype"&gt;
##FORMAT=&lt;ID=GQ,Number=1,Type=Float,Description="Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype"&gt;
##FORMAT=&lt;ID=GL,Number=G,Type=Float,Description="Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alte<span class="notranslate">rna</span>te alleles given the sample ploidy"&gt;
##FORMAT=&lt;ID=DP,Number=1,Type=Integer,Description="Read Depth"&gt;
##FORMAT=&lt;ID=AD,Number=R,Type=Integer,Description="Number of observation for each allele"&gt;
##FORMAT=&lt;ID=RO,Number=1,Type=Integer,Description="Reference allele observation count"&gt;
##FORMAT=&lt;ID=QR,Number=1,Type=Integer,Description="Sum of quality of the reference observations"&gt;
##FORMAT=&lt;ID=AO,Number=A,Type=Integer,Description="Alte<span class="notranslate">rna</span>te allele observation count"&gt;
##FORMAT=&lt;ID=QA,Number=A,Type=Integer,Description="Sum of quality of the alte<span class="notranslate">rna</span>te observations"&gt;
##FORMAT=&lt;ID=MIN_DP,Number=1,Type=Integer,Description="Minimum depth in gVCF output block."&gt;
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	raw_child-ds-	raw_mother-ds-
chrM	73	.	A	G	30438.6	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=949;CIGAR=1X;DP=950;DPB=950;DPRA=0;EPP=144.879;EPPR=5.18177;GTI=0;LEN=1;MEANALT=1;MQM=55.8314;MQMR=60;NS=2;NUMALT=1;ODDS=2187.64;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=34948;QR=37;RO=1;RPL=322;RPP=215.867;RPPR=5.18177;RPR=627;RUN=1;SAF=421;SAP=29.2075;SAR=528;SRF=0;SRP=5.18177;SRR=1;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:226:0,226:0:0:226:8314:-739.462,0	1:724:1,723:1:37:723:26634:-2346.86,0
chrM	263	.	A	G	11603.3	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=364;CIGAR=1X;DP=364;DPB=364;DPRA=0;EPP=16.755;EPPR=0;GTI=0;LEN=1;MEANALT=1;MQM=60;MQMR=0;NS=2;NUMALT=1;ODDS=982.556;PAIRED=1;PAIREDR=0;PAO=0;PQA=0;PQR=0;PRO=0;QA=13201;QR=0;RO=0;RPL=276;RPP=213.858;RPPR=0;RPR=88;RUN=1;SAF=172;SAP=5.39653;SAR=192;SRF=0;SRP=0;SRR=0;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:108:0,108:0:0:108:3896:-350.812,0	1:256:0,256:0:0:256:9305:-837.344,0
chrM	310	.	TCC	CCC	4471	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=150;CIGAR=1X2M;DP=171;DPB=180.333;DPRA=0;EPP=7.70068;EPPR=4.45795;GTI=0;LEN=1;MEANALT=2.5;MQM=59.88;MQMR=60;NS=2;NUMALT=1;ODDS=407.373;PAIRED=1;PAIREDR=1;PAO=11;PQA=237;PQR=0;PRO=0;QA=5147;QR=194;RO=6;RPL=113;RPP=86.6265;RPPR=8.80089;RPR=37;RUN=1;SAF=35;SAP=95.6598;SAR=115;SRF=1;SRP=8.80089;SRR=5;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:54:0,49:0:0:49:1645:-150.867,0	1:117:6,101:6:194:101:3502:-315.963,0
chrM	513	.	GCACACACACAC	GCACACACACACAC	2095.84	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=109;CIGAR=1M2I11M;DP=150;DPB=172.167;DPRA=0;EPP=109.173;EPPR=6.05036;GTI=0;LEN=2;MEANALT=3.5;MQM=60;MQMR=60;NS=2;NUMALT=1;ODDS=225.991;PAIRED=1;PAIREDR=1;PAO=5;PQA=101.5;PQR=101.5;PRO=5;QA=3824;QR=1196;RO=35;RPL=19;RPP=103.436;RPPR=4.56135;RPR=90;RUN=1;SAF=80;SAP=54.8268;SAR=29;SRF=24;SRP=13.4954;SRR=11;TYPE=ins;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:34:5,28:5:163:28:996:-74.9521,0	1:116:30,81:30:1033:81:2828:-161.475,0
chrM	750	.	A	G	47254.6	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=1459;CIGAR=1X;DP=1459;DPB=1459;DPRA=0;EPP=48.5904;EPPR=0;GTI=0;LEN=1;MEANALT=1;MQM=59.8705;MQMR=0;NS=2;NUMALT=1;ODDS=3230.11;PAIRED=1;PAIREDR=0;PAO=0;PQA=0;PQR=0;PRO=0;QA=53335;QR=0;RO=0;RPL=594;RPP=112.315;RPPR=0;RPR=865;RUN=1;SAF=1007;SAP=461.453;SAR=452;SRF=0;SRP=0;SRR=0;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:331:0,331:0:0:331:11988:-1073.7,0	1:1128:0,1128:0:0:1128:41347:-3719.34,0
chrM	1438	.	A	G	75374.8	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=2237;CIGAR=1X;DP=2238;DPB=2238;DPRA=0;EPP=47.8812;EPPR=5.18177;GTI=0;LEN=1;MEANALT=1;MQM=59.8605;MQMR=60;NS=2;NUMALT=1;ODDS=5003.07;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=84494;QR=37;RO=1;RPL=1397;RPP=304.171;RPPR=5.18177;RPR=840;RUN=1;SAF=925;SAP=148.392;SAR=1312;SRF=0;SRP=5.18177;SRR=1;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:488:0,488:0:0:488:18416:-1656.39,0	1:1750:1,1749:1:37:1749:66078:-5935.03,0
chrM	2706	.	A	G	39098.5	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=1222;CIGAR=1X;DP=1235;DPB=1235;DPRA=0;EPP=3.188;EPPR=3.17734;GTI=0;LEN=1;MEANALT=1;MQM=59.9926;MQMR=60;NS=2;NUMALT=1;ODDS=2496.46;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=45333;QR=474;RO=13;RPL=507;RPP=79.8897;RPPR=3.17734;RPR=715;RUN=1;SAF=225;SAP=1062.06;SAR=997;SRF=5;SRP=4.51363;SRR=8;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:246:1,245:1:32:245:9136:-818.843,0	1:989:12,977:12:442:977:36197:-3215.79,0
chrM	3197	.	T	C	130583	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=3879;CIGAR=1X;DP=3892;DPB=3892;DPRA=0;EPP=59.2643;EPPR=3.17734;GTI=0;LEN=1;MEANALT=1;MQM=59.9938;MQMR=60;NS=2;NUMALT=1;ODDS=12192.8;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=146498;QR=463;RO=13;RPL=2190;RPP=143.521;RPPR=7.18621;RPR=1689;RUN=1;SAF=1459;SAP=519.999;SAR=2420;SRF=7;SRP=3.17734;SRR=6;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:1358:4,1354:4:150:1354:50893:-4563.63,0	1:2534:9,2525:9:313:2525:95605:-8569.92,0
chrM	3243	.	A	G	10397.4	.	AB=0;ABP=0;AC=1;AF=0.5;AN=2;AO=1365;CIGAR=1X;DP=3092;DPB=3092;DPRA=0;EPP=116.418;EPPR=46.9792;GTI=0;LEN=1;MEANALT=1;MQM=59.956;MQMR=59.8917;NS=2;NUMALT=1;ODDS=2394.09;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=49004;QR=63273;RO=1727;RPL=849;RPP=179.415;RPPR=105.14;RPR=516;RUN=1;SAF=443;SAP=368.01;SAR=922;SRF=637;SRP=261.033;SRR=1090;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:1035:341,694:341:12439:694:24871:-1118.32,0	0:2057:1386,671:1386:50834:671:24133:0,-2399.65
chrM	4769	.	A	G	48890.4	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=1545;CIGAR=1X;DP=1545;DPB=1545;DPRA=0;EPP=162.63;EPPR=0;GTI=0;LEN=1;MEANALT=1;MQM=51.4796;MQMR=0;NS=2;NUMALT=1;ODDS=4468.76;PAIRED=1;PAIREDR=0;PAO=0;PQA=0;PQR=0;PRO=0;QA=57939;QR=0;RO=0;RPL=463;RPP=541.537;RPPR=0;RPR=1082;RUN=1;SAF=890;SAP=80.6281;SAR=655;SRF=0;SRP=0;SRR=0;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:517:0,517:0:0:517:19209:-1665.33,0	1:1028:0,1028:0:0:1028:38730:-3308.01,0
chrM	5539	.	A	G	3281.62	.	AB=0;ABP=0;AC=1;AF=0.5;AN=2;AO=379;CIGAR=1X;DP=787;DPB=787;DPRA=0;EPP=216.428;EPPR=224.5;GTI=0;LEN=1;MEANALT=1;MQM=54.1504;MQMR=53.777;NS=2;NUMALT=1;ODDS=755.62;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=13803;QR=15250;RO=408;RPL=74;RPP=308.741;RPPR=351.808;RPR=305;RUN=1;SAF=286;SAP=216.428;SAR=93;SRF=318;SRP=279.681;SRR=90;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	0:299:221,78:221:8252:78:2824:0,-485.016	1:488:187,301:187:6998:301:10979:-358.012,0
chrM	7028	.	C	T	70453.4	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=2109;CIGAR=1X;DP=2114;DPB=2114;DPRA=0;EPP=63.8084;EPPR=3.44459;GTI=0;LEN=1;MEANALT=1;MQM=55.9113;MQMR=59.2;NS=2;NUMALT=1;ODDS=6416.42;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=78988;QR=192;RO=5;RPL=1123;RPP=22.3353;RPPR=3.44459;RPR=986;RUN=1;SAF=969;SAP=33.1175;SAR=1140;SRF=3;SRP=3.44459;SRR=2;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:708:1,707:1:39:707:26465:-2367.64,0	1:1406:4,1402:4:153:1402:52523:-4692.5,0
chrM	7269	.	G	A	59443.2	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=1773;CIGAR=1X;DP=1773;DPB=1773;DPRA=0;EPP=129.209;EPPR=0;GTI=0;LEN=1;MEANALT=1;MQM=58.3328;MQMR=0;NS=2;NUMALT=1;ODDS=5563.82;PAIRED=1;PAIREDR=0;PAO=0;PQA=0;PQR=0;PRO=0;QA=66393;QR=0;RO=0;RPL=929;RPP=11.8591;RPPR=0;RPR=844;RUN=1;SAF=848;SAP=10.2718;SAR=925;SRF=0;SRP=0;SRR=0;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:619:0,619:0:0:619:23040:-2069.26,0	1:1154:0,1154:0:0:1154:43353:-3894.52,0
chrM	8860	.	A	G	48674.5	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=1552;CIGAR=1X;DP=1556;DPB=1556;DPRA=0;EPP=36.1924;EPPR=5.18177;GTI=0;LEN=1;MEANALT=1;MQM=46.5528;MQMR=58.25;NS=2;NUMALT=1;ODDS=4986.6;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=56983;QR=146;RO=4;RPL=845;RPP=29.6556;RPPR=5.18177;RPR=707;RUN=1;SAF=844;SAP=28.8889;SAR=708;SRF=2;SRP=3.0103;SRR=2;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:598:0,598:0:0:598:21921:-1876.69,0	1:958:4,954:4:146:954:35062:-3002.8,0
chrM	9477	.	G	A	31596.7	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=944;CIGAR=1X;DP=946;DPB=946;DPRA=0;EPP=58.9901;EPPR=3.0103;GTI=0;LEN=1;MEANALT=1;MQM=59.3178;MQMR=60;NS=2;NUMALT=1;ODDS=2909.29;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=35363;QR=67;RO=2;RPL=521;RPP=25.1023;RPPR=7.35324;RPR=423;RUN=1;SAF=469;SAP=3.09311;SAR=475;SRF=1;SRP=3.0103;SRR=1;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:329:2,327:2:67:327:12108:-1080.66,0	1:617:0,617:0:0:617:23255:-2090.4,0
chrM	9548	.	G	A	23689.1	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=729;CIGAR=1X;DP=730;DPB=730;DPRA=0;EPP=65.6375;EPPR=5.18177;GTI=0;LEN=1;MEANALT=1;MQM=59.7133;MQMR=60;NS=2;NUMALT=1;ODDS=2252;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=26539;QR=38;RO=1;RPL=394;RPP=13.3792;RPPR=5.18177;RPR=335;RUN=1;SAF=339;SAP=10.7579;SAR=390;SRF=1;SRP=5.18177;SRR=0;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:261:1,260:1:38:260:9395:-838.704,0	1:469:0,469:0:0:469:17144:-1542,0
chrM	11467	.	A	G	157655	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=4755;CIGAR=1X;DP=4759;DPB=4759;DPRA=0;EPP=620.69;EPPR=3.0103;GTI=0;LEN=1;MEANALT=1;MQM=59.9394;MQMR=42.5;NS=2;NUMALT=1;ODDS=15848.9;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=179486;QR=151;RO=4;RPL=3558;RPP=2548.64;RPPR=3.0103;RPR=1197;RUN=1;SAF=1995;SAP=270.266;SAR=2760;SRF=2;SRP=3.0103;SRR=2;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:1820:0,1820:0:0:1820:68685:-6177.29,0	1:2939:4,2935:4:151:2935:110801:-9948.16,0
chrM	11719	.	G	A	86257.7	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=2676;CIGAR=1X;DP=2687;DPB=2687;DPRA=0;EPP=15.4873;EPPR=12.6832;GTI=0;LEN=1;MEANALT=1;MQM=59.6039;MQMR=60;NS=2;NUMALT=1;ODDS=6603.9;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=96531;QR=397;RO=11;RPL=1384;RPP=9.87851;RPPR=4.78696;RPR=1292;RUN=1;SAF=1292;SAP=9.87851;SAR=1384;SRF=2;SRP=12.6832;SRR=9;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:714:3,711:3:109:711:25643:-2282.19,0	1:1973:8,1965:8:288:1965:70888:-6347.71,0
chrM	12308	.	A	G	63434.1	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=1900;CIGAR=1X;DP=1905;DPB=1905;DPRA=0;EPP=73.302;EPPR=3.44459;GTI=0;LEN=1;MEANALT=1;MQM=59.9621;MQMR=60;NS=2;NUMALT=1;ODDS=5056.51;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=71094;QR=184;RO=5;RPL=1169;RPP=222.265;RPPR=3.44459;RPR=731;RUN=1;SAF=907;SAP=11.463;SAR=993;SRF=3;SRP=3.44459;SRR=2;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:547:5,542:5:184:542:20196:-1799.85,0	1:1358:0,1358:0:0:1358:50898:-4577.25,0
chrM	12372	.	G	A	58077.1	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=1737;CIGAR=1X;DP=1741;DPB=1741;DPRA=0;EPP=74.4189;EPPR=5.18177;GTI=0;LEN=1;MEANALT=1;MQM=60;MQMR=60;NS=2;NUMALT=1;ODDS=4910.27;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=64920;QR=149;RO=4;RPL=775;RPP=46.726;RPPR=3.0103;RPR=962;RUN=1;SAF=941;SAP=29.2942;SAR=796;SRF=3;SRP=5.18177;SRR=1;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:539:3,536:3:112:536:19922:-1781.72,0	1:1202:1,1201:1:37:1201:44998:-4043.68,0
chrM	13617	.	T	C	27663.7	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=846;CIGAR=1X;DP=848;DPB=848;DPRA=0;EPP=303.228;EPPR=3.0103;GTI=0;LEN=1;MEANALT=1;MQM=59.8972;MQMR=60;NS=2;NUMALT=1;ODDS=2549.91;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=31932;QR=78;RO=2;RPL=638;RPP=477.603;RPPR=7.35324;RPR=208;RUN=1;SAF=314;SAP=124.993;SAR=532;SRF=1;SRP=3.0103;SRR=1;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:284:0,284:0:0:284:10656:-956.335,0	1:564:2,562:2:78:562:21276:-1906.39,0
chrM	14766	.	C	T	54363.3	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=1633;CIGAR=1X;DP=1633;DPB=1633;DPRA=0;EPP=21.2132;EPPR=0;GTI=0;LEN=1;MEANALT=1;MQM=59.9853;MQMR=0;NS=2;NUMALT=1;ODDS=4408.47;PAIRED=1;PAIREDR=0;PAO=0;PQA=0;PQR=0;PRO=0;QA=61054;QR=0;RO=0;RPL=852;RPP=9.71354;RPPR=0;RPR=781;RUN=1;SAF=1113;SAP=470.614;SAR=520;SRF=0;SRP=0;SRR=0;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:473:0,473:0:0:473:17508:-1575.01,0	1:1160:0,1160:0:0:1160:43546:-3916.74,0
chrM	14793	.	A	G	52354.7	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=1605;CIGAR=1X;DP=1605;DPB=1605;DPRA=0;EPP=123.965;EPPR=0;GTI=0;LEN=1;MEANALT=1;MQM=59.9688;MQMR=0;NS=2;NUMALT=1;ODDS=4357.82;PAIRED=1;PAIREDR=0;PAO=0;PQA=0;PQR=0;PRO=0;QA=58638;QR=0;RO=0;RPL=950;RPP=120.75;RPPR=0;RPR=655;RUN=1;SAF=940;SAP=105.327;SAR=665;SRF=0;SRP=0;SRR=0;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:475:0,475:0:0:475:17313:-1557.59,0	1:1130:0,1130:0:0:1130:41325:-3717.26,0
chrM	15301	.	G	A	67202.5	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=2029;CIGAR=1X;DP=2036;DPB=2036;DPRA=0;EPP=73.6971;EPPR=4.45795;GTI=0;LEN=1;MEANALT=1.5;MQM=60;MQMR=60;NS=2;NUMALT=1;ODDS=5288.64;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=75132;QR=224;RO=6;RPL=962;RPP=14.8095;RPPR=3.0103;RPR=1067;RUN=1;SAF=1129;SAP=59.1336;SAR=900;SRF=1;SRP=8.80089;SRR=5;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:563:2,561:2:75:561:20714:-1856.3,0	1:1473:4,1468:4:149:1468:54418:-4880.92,0
chrM	15326	.	A	G	73749.6	.	AB=0;ABP=0;AC=2;AF=1;AN=2;AO=2200;CIGAR=1X;DP=2201;DPB=2201;DPRA=0;EPP=68.7112;EPPR=5.18177;GTI=0;LEN=1;MEANALT=1;MQM=60;MQMR=60;NS=2;NUMALT=1;ODDS=5718.37;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=82190;QR=38;RO=1;RPL=1013;RPP=32.8937;RPPR=5.18177;RPR=1187;RUN=1;SAF=1088;SAP=3.57883;SAR=1112;SRF=0;SRP=5.18177;SRR=1;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:597:1,596:1:38:596:22276:-2000.05,0	1:1604:0,1604:0:0:1604:59914:-5388.94,0

</code></pre></div></div>

<h1 id="filtering-vcf-data">Filtering VCF data</h1>
<p>After filtering the data with stringent input parameters (restricting base quality to a minimum of 30 and <span class="notranslate">mapping</span> quality to a minimum of 20) a considerable amount variants due to read-alignment bias exits. <a href="https://github.com/ekg">Erik Garrison</a> has a beautiful illustration of various biases potentially affecting called variants (and making a locus sequence-able):</p>

<p><img src="../../images/mt_biases.png" alt="Various biases potentially affecting called variants" />
<small>Here you can see that in an ideal case (indicated with a green star) a variant is evenly represent by different areas of sequencing <span class="notranslate">reads</span> (cycle and placement biases) and is balanced across the two strands (strand bias). Allele imbalance is not applicable in our case as it reflects significant deviation from the diploid (50/50) expectation (see <a href="../../images/freebayes.pdf">here</a> for more details).</small></p>

<p>A robust tool set for processing VCF data is provided by <a href="https://github.com/vcflib/vcflib">vcflib</a> developed by Erik Garrison, the author of FreeBayes. One way to filter VCF is using <code class="language-plaintext highlighter-rouge">INFO</code> fields of the VCF dataset. If you look at the VCF dataset shown above you will see all comment lines beginning with <code class="language-plaintext highlighter-rouge">##INFO</code>.  These are <code class="language-plaintext highlighter-rouge">INFO</code> fields. Each VCF record contains a list of <code class="language-plaintext highlighter-rouge">INFO</code> tags describing a wide range of properties for each VCF record. You will see that FreeBayes and NVC differ significantly in the number and types of <code class="language-plaintext highlighter-rouge">INFO</code> fields each of these caller generates. This why the two require different filtering strategies.</p>

<p>Among numerous types of data generated by FreeBayes let’s consider the following variant properties:</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">##INFO=&lt;ID=DP,Number=1,Type=Integer,Description="Total read depth at the locus"&gt;</code> This is simply the number of <span class="notranslate">reads</span> covering a given site.</li>
  <li><code class="language-plaintext highlighter-rouge">##INFO=&lt;ID=SRP,Number=1,Type=Float,Description="Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality"&gt;</code> The higher this quantity the better the site as it diminishes the chances of the sites having significant strand bias.</li>
  <li><code class="language-plaintext highlighter-rouge">##INFO=&lt;ID=SAP,Number=A,Type=Float,Description="Strand balance probability for the alte<span class="notranslate">rna</span>te allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality"&gt;</code> The higher this quantity the better the site as it diminishes the chances of the sites having significant strand bias  (also see <a href="https://groups.google.com/forum/#!topic/freebayes/fX4TOAqXJrA">here</a>).</li>
  <li><code class="language-plaintext highlighter-rouge">##INFO=&lt;ID=EPP,Number=A,Type=Float,Description="End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality"&gt;</code> The higher this number the lower the chance of having significant placement bias.</li>
  <li><code class="language-plaintext highlighter-rouge">QUAL</code> - phred scaled variant quality.</li>
</ul>

<p>To perform filtering we will use <strong>NGS: VCF Manipulation</strong> → <strong>VCFfilter</strong>):</p>
<blockquote class="notranslate hands_on">
  <h3 id="filtering-vcf-data-1">Filtering VCF data</h3>

  <p><img src="../../images/mt_vcffilter.png" alt="VCFfilter parameters" /></p>

  <p>Filtering FreeBayes VCF for strand bias (<code class="language-plaintext highlighter-rouge">SPR</code> and <code class="language-plaintext highlighter-rouge">SAP</code>), placement bias (<code class="language-plaintext highlighter-rouge">EPP</code>), variant quality (<code class="language-plaintext highlighter-rouge">QUAL</code>), and depth of coverage (<code class="language-plaintext highlighter-rouge">DP</code>).</p>
</blockquote>

<p>The resulting VCF only contains two variants (most comments fields are omitted here):</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	raw_child-ds-	raw_mother-ds-
chrM	3243	.	A	G	10397.4	.	AB=0;ABP=0;AC=1;AF=0.5;AN=2;AO=1365;CIGAR=1X;DP=3092;DPB=3092;DPRA=0;EPP=116.418;EPPR=46.9792;GTI=0;LEN=1;MEANALT=1;MQM=59.956;MQMR=59.8917;NS=2;NUMALT=1;ODDS=2394.09;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=49004;QR=63273;RO=1727;RPL=849;RPP=179.415;RPPR=105.14;RPR=516;RUN=1;SAF=443;SAP=368.01;SAR=922;SRF=637;SRP=261.033;SRR=1090;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	1:1035:341,694:341:12439:694:24871:-1118.32,0	0:2057:1386,671:1386:50834:671:24133:0,-2399.65
chrM	5539	.	A	G	3281.62	.	AB=0;ABP=0;AC=1;AF=0.5;AN=2;AO=379;CIGAR=1X;DP=787;DPB=787;DPRA=0;EPP=216.428;EPPR=224.5;GTI=0;LEN=1;MEANALT=1;MQM=54.1504;MQMR=53.777;NS=2;NUMALT=1;ODDS=755.62;PAIRED=1;PAIREDR=1;PAO=0;PQA=0;PQR=0;PRO=0;QA=13803;QR=15250;RO=408;RPL=74;RPP=308.741;RPPR=351.808;RPR=305;RUN=1;SAF=286;SAP=216.428;SAR=93;SRF=318;SRP=279.681;SRR=90;TYPE=snp;technology.ILLUMINA=1	GT:DP:AD:RO:QR:AO:QA:GL	0:299:221,78:221:8252:78:2824:0,-485.016	1:488:187,301:187:6998:301:10979:-358.012,0
</code></pre></div></div>
<h1 id="looking-at-the-data">Looking at the data</h1>

<p>For visualizing VCFs <span class="notranslate">Galaxy</span> relies on the two exte<span class="notranslate">rna</span>l tools.  The first is called <a href="http://vcf.iobio.io/">VCF.IOBIO</a> and is developed by <a href="http://marthlab.org/">Gabor Marth’s group</a> at the University of Utah. The second is called <a href="http://software.broadinstitute.org/software/igv/">IGV</a> developed by Broad Institute.</p>

<h2 id="vcfiobio">VCF.IOBIO</h2>

<p>VCF.IOBIO can be invoked by expanding a VCF dataset in <span class="notranslate">Galaxy</span>’s history by clicking on it:</p>

<blockquote class="notranslate hands_on">
  <h3 id="displaying-data-in-vcfiobio">Displaying data in VCF.IOBIO</h3>

  <p><img src="../../images/mt_vcf_dataset_collapsed.png" alt="VCFfilter output in history" /></p>

  <p>Clicking on the dataset above will expand it as shown below:</p>

  <p><img src="../../images/mt_vcf_dataset_expanded.png" alt="Expanded output" /></p>

  <p>At the bottom there is a link “display at vcf.iobio”
Clicking on this link will start indexing of VCF datasets, which is required to display them. After indexing VCF.IOBIO will open:</p>

  <p><img src="../../images/mt_vcfiobio.png" alt="VCF.IOBIO overview" /></p>

  <p>Of course there are not that many variants to look at in this example. Nevertheless there are helpful statistics such as Transition/Transversion (Ts/Tn) ratio.</p>
</blockquote>

<h2 id="igv">IGV</h2>

<p>Similarly to VCF.IOBIO expanding a history item representing a VCF dataset will reveal an IGV link:</p>

<blockquote class="notranslate hands_on">
  <h3 id="displaying-data-in-igv">Displaying data in IGV</h3>

  <p><img src="../../images/mt_vcf_dataset_expanded.png" alt="VCFfilter output expanded" /></p>

  <p>At the bottom there is a link “display at IGV: local Human hg38”
The difference between “local” and “Human hg38” links is explained in the following video:</p>

  <div class="embed-responsive embed-responsive-16by9"><iframe src="https://player.vimeo.com/video/123414437?portrait=0" webkitallowfullscreen="" mozallowfullscreen="" allowfullscreen=""></iframe></div>

  <p>Visualizing our FreeBayes dataset will produce this:</p>

  <p><img src="../../images/mt_igv.png" alt="IGV interface" /></p>

  <p>Here we focus on one particular variant at position 3,243 for reasons that will become apparent in the next section.</p>
</blockquote>

<h1 id="digging-into-the-data">Digging into the data</h1>

<p>Visualizing VCF dataset may be a good way to get an overall idea of the data, but it does not tell a lot of details. For example, above we have visualized site 3,243 using IGV. It is interesting but we need to find out more. One thing we can do is to convert VCF dataset into a tab-delimited representation and with it play a bit more.</p>

<p>Using <strong>NGS: VCF Manipulation</strong> → <strong>VCFtoTab-delimited</strong> on the filtered VCF dataset:</p>

<blockquote class="notranslate hands_on">
  <h3 id="from-vcf-to-tab-delimited-data">From VCF to Tab-delimited data</h3>

  <p><img src="../../images/mt_vcfToTab.png" alt="VCFtoTab-delimited parameters window" /></p>

  <p>Make sure <strong>Report data per sample</strong> is set to <code class="language-plaintext highlighter-rouge">Yes</code></p>

  <p>This will produce a dataset with <em>very</em> many columns:</p>

  <p><img src="../../images/mt_tab.png" alt="VCFtoTab-delimited output" /></p>

  <p>There are 62 columns in this dataset (not all are shown here).</p>
</blockquote>

<p>The columns in the above dataset represent INFO and Genotype fields on the original VCF dataset. Let’s restrict ourselves to just a few:</p>

<ul>
  <li>2 <code class="language-plaintext highlighter-rouge">POS</code> - position along mitochondrial genome</li>
  <li>4 <code class="language-plaintext highlighter-rouge">REF</code> - reference allele</li>
  <li>5 <code class="language-plaintext highlighter-rouge">ALT</code> - alte<span class="notranslate">rna</span>tive allele</li>
  <li>52 <code class="language-plaintext highlighter-rouge">SAMPLE</code> - name of the sample</li>
  <li>54 <code class="language-plaintext highlighter-rouge">AO</code> - number of alte<span class="notranslate">rna</span>tive observations (how many times do we see the alte<span class="notranslate">rna</span>tive allele at this position in this sample)</li>
  <li>55 <code class="language-plaintext highlighter-rouge">DP</code> - depth of coverage at this site for this sample</li>
</ul>

<p>To cut these columns out we will use <strong>Text Manipulation</strong> → <strong>Cut</strong></p>

<blockquote class="notranslate hands_on">
  <h3 id="cutting-columns-from-a-file">Cutting columns from a file</h3>

  <p><img src="../../images/mt_cut.png" alt="Cut tool options" /></p>

  <p>Note that column names are preceded with <code class="language-plaintext highlighter-rouge">c</code></p>

  <p>This will generate the following dataset:</p>

  <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>POS  REF ALT     SAMPLE       AO    DP
--------------------------------------
3243	A	G	raw_child-ds-	694	1035
3243	A	G	raw_mother-ds-	671	2057
5539	A	G	raw_child-ds-	78	299
5539	A	G	raw_mother-ds-	301	488
</code></pre></div>  </div>
</blockquote>

<p>Let’s look at site 3,243. At this site Mother has 691 <code class="language-plaintext highlighter-rouge">G</code>s (since <code class="language-plaintext highlighter-rouge">G</code> is an alte<span class="notranslate">rna</span>tive allele) and 1,035-2057=1386 <code class="language-plaintext highlighter-rouge">A</code>s. This child has 694 <code class="language-plaintext highlighter-rouge">G</code>s and 1,035-694=341 <code class="language-plaintext highlighter-rouge">A</code>s:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Allele     A     G
-------------------
Mother  1386   691
Child   341    694
</code></pre></div></div>

<p>Thus the <em>major</em> allele in mother (<code class="language-plaintext highlighter-rouge">A</code>) becomes the <em>minor</em> allele in child – a remarkable frequency change due to mitochondrial bottleneck!</p>

<h1 id="take-a-look-at-the-whole-thing">Take a look at the whole thing</h1>

<p>This entire analysis is available as a <a href="https://usegalaxy.org/u/aun1/h/non-diploid-freebayes"><span class="notranslate">Galaxy</span> history</a> that you can import into your <span class="notranslate">Galaxy</span> account and play with.</p>

<p>Now you know how to call variants in non-diploid system, so try it on bacteria, viruses, etc.</p>


                    
                    <blockquote class="key_points">
                        <h3><i class="fas fa-key" aria-hidden="true"></i><span class="visually-hidden">keypoints</span> Key points</h3>
                        <ul>
                            
                            <li><p>Variation in human mitochondria is a proxy for other haploid systems.</p>
</li>
                            
                            <li><p>Nucleotide variants in haploid systems may have any frequency between 0 and 1.</p>
</li>
                            
                            <li><p>Distinguishing true variants from noise is not easy and requires accounting <br /> for strand bias and read placement bias.</p>
</li>
                            
                            <li><p>FreeBayes can be effectively used to call variants in haploid systems.</p>
</li>
                            
                        </ul>
                    </blockquote>
                    

                    
                    <h1 data-toc-skip>Useful literature</h1>
                    <p>Further information, including links to documentation and original publications, regarding the tools, analysis techniques and the interpretation of results described in this tutorial can be found <a href="/training-material/topics/variant-analysis#references">here</a>.</p>
                    

                    

                    <h1>Feedback</h1>
                    <p class="text-muted">Did you use this material as an instructor? Feel free to give us feedback on <a href="https://github.com/galaxyproject/training-material/issues/1452" target="_blank">how it went</a>.</p>

                    <div id="feedback-button">
                        <img src="/training-material/shared/images/feedback.png" title="Click to activate" alt="Click here to load Google feedback frame" />
                    </div>
                    <div id="feedback-form">
                    </div>
                    <script type="text/javascript">
                        (function (window, document) {
                            function onDocumentReady(fn) {
                                if (document.attachEvent ? document.readyState === "complete" : document.readyState !== "loading") {
                                    fn();
                                } else {
                                    document.addEventListener('DOMContentLoaded', fn);
                                }
                            }

                            onDocumentReady(function () {
                                $("#feedback-button").click(function(evt){
                                    var e = $(evt.target)
                                    e.hide();

                                    $("#feedback-form").html(`
                                        <iframe id="feedback-google" class="google-form" src="https://docs.google.com/forms/d/e/1FAIpQLSd4VZptFTQ03kHkMz0JyW9b6_S8geU5KjNE_tLM0dixT3ZQmA/viewform?embedded=true&entry.1235803833=Calling variants in non-diploid systems (Variant Analysis)">Loading...</iframe>
                                    `)
                                })
                            });
                        })(window, document);
                    </script>



                    <h1>Citing this Tutorial</h1>
                    <p>
                        <ol>
                            <li id="citation-text">Anton Nekrutenko, Alex Ostrovsky, 2020 <b>Calling variants in non-diploid systems (Galaxy Training Materials)</b>. <a href="/training-material/topics/variant-analysis/tutorials/non-dip/tutorial.html">/training-material/topics/variant-analysis/tutorials/non-dip/tutorial.html</a> Online; accessed TODAY
                            </li>
                            <li>
                            Batut et al., 2018 <b>Community-Driven Data Analysis Training for Biology</b> Cell Systems <a href="https://doi.org/10.1016%2Fj.cels.2018.05.012">10.1016/j.cels.2018.05.012</a>
                            </li>
                        </ol>
                    </p>


                    <blockquote class="details">
                      <h3><i class="fa fa-info-circle" aria-hidden="true"></i><span class="visually-hidden">details</span> BibTeX</h3>
                      <p style="display: none;">

                    <div class="highlighter-rouge"><div class="highlight"><pre class="highlight">
<code id="citation-code">@misc{variant-analysis-non-dip,
    author = "Anton Nekrutenko and Alex Ostrovsky",
    title = "Calling variants in non-diploid systems (Galaxy Training Materials)",
    year = "2020",
    month = "07",
    day = "28"
    url = "\url{/training-material/topics/variant-analysis/tutorials/non-dip/tutorial.html}",
    note = "[Online; accessed TODAY]"
}
@article{Batut_2018,
        doi = {10.1016/j.cels.2018.05.012},
        url = {https://doi.org/10.1016%2Fj.cels.2018.05.012},
        year = 2018,
        month = {jun},
        publisher = {Elsevier {BV}},
        volume = {6},
        number = {6},
        pages = {752--758.e1},
        author = {B{\'{e}}r{\'{e}}nice Batut and Saskia Hiltemann and Andrea Bagnacani and Dannon Baker and Vivek Bhardwaj and Clemens Blank and Anthony Bretaudeau and Loraine Brillet-Gu{\'{e}}guen and Martin {\v{C}}ech and John Chilton and Dave Clements and Olivia Doppelt-Azeroual and Anika Erxleben and Mallory Ann Freeberg and Simon Gladman and Youri Hoogstrate and Hans-Rudolf Hotz and Torsten Houwaart and Pratik Jagtap and Delphine Larivi{\`{e}}re and Gildas Le Corguill{\'{e}} and Thomas Manke and Fabien Mareuil and Fidel Ram{\'{\i}}rez and Devon Ryan and Florian Christoph Sigloch and Nicola Soranzo and Joachim Wolff and Pavankumar Videm and Markus Wolfien and Aisanjiang Wubuli and Dilmurat Yusuf and James Taylor and Rolf Backofen and Anton Nekrutenko and Björn Grüning},
        title = {Community-Driven Data Analysis Training for Biology},
        journal = {Cell Systems}
}</code>
                    </pre></div></div>
                    </p>
                    </blockquote>


<script type="text/javascript">
// update the date on load, or leave fallback of 'today'
d = new Date();
document.getElementById("citation-code").innerHTML = document.getElementById("citation-code").innerHTML.replace("TODAY", d.toDateString());
document.getElementById("citation-text").innerHTML = document.getElementById("citation-text").innerHTML.replace("TODAY", d.toDateString());
</script>

                </div>
            </div>
        </div>

        <h3><i class="far fa-thumbs-up" aria-hidden="true"></i><span class="visually-hidden">congratulations</span> Congratulations on successfully completing this tutorial!</h3>

        

        
    </section>
</div>


<footer>
    <div class="container">
        <p>
            This material is the result of a collaborative work. Thanks to the
            <a href="https://wiki.galaxyproject.org/Teach/GTN">Galaxy Training Network</a>
            and all the <a href="/training-material/hall-of-fame">contributors</a> (Anton Nekrutenko, Alex Ostrovsky)!
        </p>
        <p>
            Found a typo? Something is wrong in this tutorial? Edit it on
            <a href="https://github.com/galaxyproject/training-material/tree/master/topics/variant-analysis/tutorials/non-dip/tutorial.md">GitHub</a>.
        </p>
        <p>
    The content of the tutorials and website is licensed under the <a rel="license" href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.
</p>
    </div>
</footer>

    </body>
    <script type="text/javascript" src="/training-material/assets/js/jquery.slim.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/popper.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/bootstrap.min.js?v=3"></script>
    <script type="text/javascript" src="/training-material/assets/js/details-element-polyfill.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="/training-material/assets/js/bootstrap-toc.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/main.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/theme.js"></script>

    <script type="text/javascript" src="/training-material/assets/js/clipboard.min.js"></script>
    <script type="text/javascript">
    var snippets=document.querySelectorAll('div.highlight');
    [].forEach.call(snippets,function(snippet){
        snippet.firstChild.insertAdjacentHTML('beforebegin','<button class="btn btn-light" data-clipboard-snippet><i class="fa fa-copy"></i>&nbsp;Copy</button>');
    });

    var clipboardSnippets=new ClipboardJS('[data-clipboard-snippet]',{
        target:function(trigger){return trigger.nextElementSibling;
    }});
    </script>
    
</html>
