---
layout: tutorial_slides
logo: "GTN"
title: "Clustering 3K PBMCs with Scanpy"
zenodo_link: ''
questions:
  - "What are the 6 main steps of scRNA-seq?"
  - "What kind of variation can confound an analysis?"
objectives:
  - "Learn the main stages of an scRNA-seq analysis"
  - "Understand the methods and concepts underlying scRNA-seq"
time_estimation: 1H
key_points:
  - "Understanding the purpose of barcoding"
  - "Knowing the difference between hard and soft clustering"
  - "A KNN graph can be generated from a count matrix."
  - "Community clustering can be generated from a KNN graph."
  - "Interpreting scRNA-seq plots"
contributors:
  - mtekman

---

### Single Cell RNA Pre-processing

.image-100[![fig0](../../images/scrna_workflow_upper.png)]


???
- Barcoding and (de-)multiplexing
- Quantification

---

### Single Cell RNA Downstream Analysis

.image-100[![fig0](../../images/scrna_workflow_lower.png)]

???
1. Filtering
1. Normalising
   * Confounder Removal
1. Dimension Reduction
1. Clustering
   * Annotation

---

### Barcoding Cells

.image-100[![fig1](../../images/scrna_pbb_barcodes_add.svg)]

---


### Filtering: Cell and Gene

.image-50[![fig2](../../images/raceid_libsize.svg)]

---

### Normalisation: Technical Variation

.image-50[![fig3](../../images/raceid_technical_variation.svg)]

---

### Normalisation: Biological Variation


.image-50[![fig4](../../images/raceid_cellcycle.svg)]

---

### DimRed: Relatedness of Cells

.image-50[![fig5](../../images/scrna_knn.svg)]


???
Build a KNN graph from distance matrix:
  * If P and q share distance which is 

---

### DimRed: Projection

.image-50[![fig6](../../images/raceid_dimred.svg)]

- Can use tSNE, PCA, UMAP

---

### Community Clustering: Louvain

.image-50[![fig7](../../images/commgraph1.svg)]

Aim: Maximise internal links and minimise external links

---

### Community Clustering: Louvain

.image-50[![fig8](../../images/commgraph2.svg)]

* Randomly pick a cell and try to place it in a neighbour's cluster
  * Accept if *Internal:External* increases
  * Reject and pick another

---

### Cell Type: Identifying Cluster Types

.pull-left[![fig9](../../images/singlecellplot3.png)]
.pull-right[![fig10](../../images/singlecellplot4.png)]

---

### Clustering: Hard vs Soft

.pull-left[
![fig11](../../images/singlecellplot3.png)

*Hard*
  * Big spaces between clusters
  * Cell types are well defined and the clustering reflects that

]

--

.pull-right[
![fig12](../../images/10xdata.png)

*Soft*
  * Clusters bleed into one another
  * Cell types seem to intermingle with one another.

]

???
Why? Why would there be clusters so close to one another?

---

### Continuous Phenotypes:

.image-50[![fig13](../../images/raceid_contpheno.svg)]

* Cells aren't discrete, they transition
* Continuously changing over time from a less mature type to more mature type


---

### What is Differential Expression in scRNA-seq?



---

### Interactive Environments: live.usegalaxy.eu


<video style="width:100%;height:auto" controls>
    <source src="https://zenodo.org/record/3687100/files/live_cellxgene.mp4" />
</video>


---

### Test

CellxGene

* We can probe clusters to see how they are so differentially expressed

> `pip3 install cellxgene`  
> `cellxgene launch https://cellxgene-example-data.czi.technology/pbmc3k.h5ad`

* Launch locally: http://127.0.0.1:5005

