---
layout: tutorial_slides
logo: "GTN"

title: "Network Analysis with Heinz"
questions:
  - "What is metatranscriptomics"
  - "Which pathways are potentially contributing to dental caries?"
objectives:
  - "Understanding metatranscriptomics"
  - "Familiarize yourself with KEGG terminology"
  - "Understand home pathway analysis works"
key_points:
  - "Heinz is useful for .."
contributors:
  - cicozhang
  - sanneabeln
---

# Metagenomics / metaproteomics

- Sequence all DNA/RNA in a sample from a specific microbial environment
- Get an overview of the entire microbial community

![sample collection from environment to sequence data](../../images/heinz/sampleextraction.png)

???
There are several advantages in sequencing all DNA (metagenomics) or RNA (metatranscriptomics) in a sample to consider the microbial content.

Many bacterial species are not easily grown on a petri-dish, and may therefore not be observed with classic  analysis techniques.

Metatranscriptomics can also reveal the functional state of the microbiome, suggesting which functional pathways are switched on.

---

# The human microbiome

- Over 100 trillion microbes
- Outnumber our human cells 1.3 to 1.
- Include bacteria, fungi, protozoa, viruses, in and outside our body.
- The number of genes in all the microbes in one person's microbiome is ~10-200 times the number of genes in the human genome [*].
- The microbiome may weigh as much as 200 grams


.image-50[ ![](../../../metagenomics/images/human_microbiome.png)]

.footnote[[Sender et al 2016, PLoS ONE](https://doi.org/10.1371/journal.pbio.1002533)]

???
The most common objective is to ide
[*] this figure is difficult to estimate, as we need to consider when two orthologous genes can be considered different. A rough estimate would be 500~10000 species of bacteria with a median of 2000 genes.

---

# Environments related to health

.pull-left[
- Gut / bowel / intestine
- Oral Cavity
- Nasal Cavity
- Skin

![oral cavity](../../images/heinz/oralcavity.jpg)
]

.pull-right[
![Nature cover "the other genome"](../../images/heinz/othergenome.png)
]
---

# What would we find?

.pull-left[
- Oral Cavity
- DNA from:
  - bacteria
  - viruses
  - human
  - archaea
  - spider (?!)
]

.pull-right[
![](../../images/heinz/oralcavity.jpg)
]

---

# Some bacterial species are difficult to culture

![petri dish with bacterial cultures](../../images/heinz/petridish.png)

- Such bacteria may not be found in current reference databases
- The only way to get hold of uch species may be through direct squencing of environmental samples
- Purpose: find novel species, genomes, and genes


---

# Metagenomics vs 16S

- Full shotgun sequencing of sample:
  - sequence all available DNA/ RNA
  - may need to sequence very deep
  - can also give functional composition

- Sequence ribosomal regions (16S)
  - indicates taxonomic composition
  - Need amplicons
  - Need less sequencing

???
For the workflow considered here we assume RNAseq of the microbiome.

---

# 16S ribosomal DNA

![overview of the bacterial genome](../../images/heinz/16Sgene.png)

- Multiple copies of **16S,23S** andd **5S**
- 16S: 9 hypervariable & 9 constant regions

???
16S sequencing is still very popular to estimate the diversity of a microbiome. In this work we are also interested in the functional content.

---

# Gum disease (priodontitis)

![journal article "Metatranscriptomics of the human oral microbiome during health and disease"](../../images/heinz/abstract.png)

???
For reference only

---

# Microbiome

![oral microbiome](../../images/heinz/oralmicrobiome.png)

???

When we work with metagenomics data there are two main aims
1) profiling of the species content of the microbiome
2) a functional analysis of the microbiome.

---

# MetaModules

![metamodules publication](../../images/heinz/metamodules.png)

???
Here we consider the “metaModules” workflow designed by Ali May et a.


---

# Exploring the networks of microbial function that are involved in disease

![](../../images/heinz/network.png)

???
The key idea is to explore functional  networks of the microbiome that are involved in disease

---

# Oral disease

.pull-left[
- **Dental caries**
  The most common chronic disease of children (25%), and adolescents (59%). Bacteria feed on food debris. Bacteria produce acids leading to demineralization and destruction of the teeth.

- **Periodontal disease**
  Bacteria living in tissues surrounding and supporting the teeth; An over-aggressive immune response can subsequently lead to loosening and finally loss of teeth.


]

.pull-right[

.image-75[ ![dental caries](../../images/heinz/caries.jpg)]

.image-75[ ![periodontal disease](../../images/heinz/periodontaldisease.jpg)]

]

---

# Healthy vs Disease

![comparing microbiomes from healthy and disease samples](../../images/heinz/healthyvsdisease.png)

???
The first part of the workflow entail RNA sequencing of oral microbiome samples.

From this we get the total RNA present, from all species, including ribosomal RNA.

This can be used to estimate both a taxonomic profile, and a functional profile.

With the workflow presented here, we go one step further, and consider which subnetworks in the metabolic pathways are most deregulated in the microbiome samples showing disease (caries).

---

## Using knowledge to pinpoint the key players

![](../../images/heinz/workflow.png)

???

So, now to the experimental design and analysis steps. We used a previously published metatranscriptomic dataset of microbial RNA reads from healthy and caries disease samples.

We did some preprocessing to filter out non-coding RNA etc, and mapped the remaining reads to a KEGG database of 3 million genes.

We calculated the gene counts in healthy and diseased samples and aggregated these gene counts in KOs.

This was followed b differential expression analysis at the KO level, from which we derived fold changes and P-values for KOs.

We use the P-values to obtain scores for each KO, use the scores to annotate the network and finally run an algorithm in the network to identify the maximum-scoring regions.

---

# Some (KEGG database) terminology

![KEGG species view](../../images/heinz/kegg1.png)

???
But let’s talk briefly about how we define these molecular functions first and get familiar with the terminology. In the database we use, KEGG, we have species and corresponding molecular pathways, like in this case S. Mutans and glycolysis.

---

# Some (KEGG database) terminology

![KEGG species view](../../images/heinz/kegg1-2.png)


???
Of course, each species has its own specific metabolic network - even though we may not have the information on all specific species in a sample.

---

# Some (KEGG database) terminology

![KEGG KO groups](../../images/heinz/kegg2.png)

???
Now, as orthologous genes in the different species perform the same function, we can collapse the metabolic network of all known bacterial species in to a single network. Here is node represents a KO: a KEGG Orthologous Group, containing orthologous genes performing the same metabolic function in the metabolic network. We use this network of KOs to analyse the microbial sample.

---

# Functional signatures of microbiome-related disease

The most common objective is to identify differentially expressed pathways by performing comparative metagenomic/metatranscriptomic analyses.

![funtional signatures of healthy and disease samples](../../images/heinz/functionalsignatures.png)

???
To understand the functional differences between two conditions, for instance in microbiome-related disease and health, what is commonly done is to look at differentially expressed pathways between the two conditions.

The figure on left shows the pathways that are significantly over- or under-abundant in diseased and healthy patients.

On the right is a figure from another study where researchers listed some important pathways and the KOs in these pathways that are up- (blue) or downregulated (red) in disease.

---

![pathway analysis of a network image](../../images/heinz/globalapproach.png)

???
We can now visualise the RNAseq data on our collapsed microbial metabolic network of KOs. The red nodes represent KOs that are overexpressed in disease, and the blue node those that are underexpressed in disease as compared to healthy samples.

It is immediately clear that the data contains many overexpressed genes, and it is difficult to immediately pinpoint the most deregulated part of the network. For this purpose, we use the Heinz methods to extract de most deregulated subnetwork.

---

# Identifying signatures of disease

- Where are the most important regions in this large network?
- How can we find such regions?

![](../../images/heinz/identification.png)

---

#
