<!doctype html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <meta http-equiv="x-ua-compatible" content="ie=edge">
        <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
        <title>Galaxy Training!</title>
        
        <script>
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
            ga('create', 'UA-45719423-18' , 'auto');
            ga('send', 'pageview');
        </script>
        
        <link rel="stylesheet" href="/training-material/assets/css/bootstrap.min.css?v=3">
        <link rel="stylesheet" href="/training-material/assets/css/main.css?v=2">
        <link rel="stylesheet" href="/training-material/assets/css/font-awesome.css">
        <link rel="stylesheet" href="/training-material/assets/css/academicons.css">
        <link rel="stylesheet" href="/training-material/assets/css/syntax_highlighting.css">
        <link rel="shortcut icon" href="/training-material/favicon.ico" type="image/x-icon" />

        
        
        
        
        
        <meta name="description" content="ChIP-sequencing is a method used to analyze protein inter..." />
        <meta property="og:title" content="Galaxy Training: ChIP-Seq data analysis" />
        <meta property="og:description" content="ChIP-sequencing is a method used to analyze protein inter..." />
        <meta property="og:image" content="/training-material/assets/images/GTNLogo1000.png" />
    </head>
    <body>
        <header>
    <nav class="navbar navbar-expand-md navbar-dark">
        <div class="container">
            <a class="navbar-brand" href="/training-material/">
                <img src="/training-material/assets/images/GTN-60px.png" height="30" alt="Galaxy Training Network logo">
                Galaxy Training!
            </a>

            <button class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#top-navbar" aria-controls="top-navbar" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>
            <div class="collapse navbar-collapse" id="top-navbar">
                <ul class="navbar-nav">
                    <li class="nav-item">
                        <a class="nav-link" href="https://github.com/galaxyproject/training-material" title="View on GitHub">
                            <i class="fa fa-github" aria-hidden="true"></i><span class="visually-hidden">github</span> View on GitHub
                        </a>
                    </li>
                    <li class="nav-item dropdown">
    <a href="#" class="nav-link dropdown-toggle" data-toggle="dropdown" aria-expanded="false" title="Help">
        <i class="fa fa-life-ring" aria-hidden="true"></i><span class="visually-hidden">help</span> Help
    </a>
    <div class="dropdown-menu dropdown-menu-right">
        <form method="get" action="https://tess.elixir-europe.org/materials">
            <input type="text" id="search" name="q" value="" style="margin-left: 0.5em;/*! border-radius: 0px; */">
            <input type="hidden" value="Galaxy Training" name="content_provider">
            <input type="submit" value="Search on TeSS" style="width: 92%;border-radius: 0px;margin: 0.5em;background: #f47d20;border: 0px;padding: 0.25em;" class="">
        </form>

        <div class="dropdown-divider"></div>
        <a class="dropdown-item" href="/training-material/faq" title="Check our FAQ">
            FAQ
        </a>
        <a class="dropdown-item" href="https://gitter.im/Galaxy-Training-Network/Lobby" title="Chat on Gitter">
            Chat on Gitter
        </a>
        <a class="dropdown-item" href="https://biostar.usegalaxy.org/" title="Ask on Biostars">
            Ask questions on Biostars
        </a>
    </div>
</li>

                </ul>
            </div>
        </div>
    </nav>
</header>








<div class="container main-content">
    <section>
        <h1>ChIP-Seq data analysis</h1>
        <p>ChIP-sequencing is a method used to analyze protein interactions with DNA.</p>

        
        <h2 id="requirements">Requirements</h2>
        <p>Before diving into this topic, we recommend you to have a look at:</p>

        <ul>
            
                <li>
                    
                    <a href="/training-material/topics/introduction/">Galaxy introduction</a>
                    
                </li>
            
                <li>
                    
                    <a href="/training-material/topics/sequence-analysis/">Quality control</a>
                    
                </li>
            
                <li>
                    
                    <a href="/training-material/topics/sequence-analysis/">Mapping</a>
                    
                </li>
            
        </ul>
        

        <h2 id="material">Material</h2>
        <table class="table table-responsive table-striped">
            <tr>
                <th>Lesson</th>
                <th>Slides</th>
                <th>Hands-on</th>

                
                <th>Input dataset</th>
                <th>Workflows</th>
                <th>Galaxy tour</th>
                

                
                <th>Galaxy instances</th>
                
            </tr>

            
                
                        <script type="application/ld+json">
                            
{
  "@context": "http://schema.org",
  "@type": "CreativeWork",
  "name": "Introduction to ChIP-Seq data analysis",
  "accessModeSufficient": "visual",
  "accessibilityControl": "fullMouseControl",
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "https://github.com/galaxyproject/training-material/blob/master/LICENSE.md",
  "learningResourceType": "introduction",
  "url": "",
  "mentions": [

  ],
  "contributor": [
    {
      "@type": "Person",
      "name": "bebatut"
    }
  ]
}

                        </script>
                    <tr>
                        <td>
                            Introduction to ChIP-Seq data analysis
                            
                        </td>
                        
                            <td>
                                
                                <a class="topic-icon" href="/training-material/topics/chip-seq/slides/introduction.html">
                                  <i class="fa fa-slideshare" aria-hidden="true"></i><span class="visually-hidden">slides</span>
                                </a>
                                
                            </td>
                            <td></td>
                            
                                <td></td>
                                <td></td>
                            
                            
                                <td></td>
                            
                        
                    </tr>
                
            
                
                        <script type="application/ld+json">
                            
{
  "@context": "http://schema.org",
  "@type": "CreativeWork",
  "name": "Formation of the Super-Structures on the Inactive X",
  "accessModeSufficient": "visual",
  "accessibilityControl": "fullMouseControl",
  "accessibilityFeature": [
    "alternativeText",
    "tableOfContents"
  ],
  "audience": {
    "@type": "EducationalAudience",
    "educationalRole": "students"
  },
  "inLanguage": {
    "@type": "Language",
    "name": "English",
    "alternateName": "en"
  },
  "interactivityType": "mixed",
  "isAccessibleForFree": true,
  "license": "https://github.com/galaxyproject/training-material/blob/master/LICENSE.md",
  "learningResourceType": "tutorial",
  "url": "",
  "mentions": [
    {
      "@type": "Thing",
      "url": "https://doi.org/10.5281/zenodo.1321974",
      "name": "Training data for Formation of the Super-Structures on the Inactive X tutorial"
    },
    {
      "@type": "Thing",
      "url": "https://github.com/galaxyproject/training-material/tree/master/topics/chip-seq/tutorials/formation_of_super-structures_on_xi/workflows/",
      "name": "Workflow for Formation of the Super-Structures on the Inactive X tutorial"
    }
  ],
  "timeRequired": "PT3H",
  "contributor": [
    {
      "@type": "Person",
      "name": "friedue"
    },
    {
      "@type": "Person",
      "name": "erxleben"
    },
    {
      "@type": "Person",
      "name": "bebatut"
    },
    {
      "@type": "Person",
      "name": "vivekbhr"
    },
    {
      "@type": "Person",
      "name": "fidelram"
    },
    {
      "@type": "Person",
      "name": "LeilyR"
    }
  ],
  "description": "The questions this tutorial addresses are:\n - Histone modification?\n - How is a raw set of ChIP-seq data processed and analyzed?\n - Where are the enriched regions for H3K27me3, H3K4me3 and CTCF on the chrX?\n\n\\nThe objectives are:\n - Inspect the read quality\n - Trim low quality bases\n - Map reads on a reference genome\n - Assess the quality of a ChIP-seq experiment\n - Extract coverage files\n - Call enriched regions or peaks\n\n",
  "hasPart": [
    {
      "@type": "CreativeWork",
      "url": "https://github.com/galaxyproject/training-material/tree/master/topics/chip-seq/tutorials/formation_of_super-structures_on_xi/tours/",
      "name": "Slides for Formation of the Super-Structures on the Inactive X",
      "learningResourceType": "slides",
      "interactivityType": "expositive"
    }
  ]
}

                        </script>
                    <tr>
                        <td>
                            Formation of the Super-Structures on the Inactive X
                            
                        </td>
                        
                            <td>
                                
                                    <a class="topic-icon" href="/training-material/topics/chip-seq/tutorials/formation_of_super-structures_on_xi/slides.html">
                                    <i class="fa fa-slideshare" aria-hidden="true"></i><span class="visually-hidden">slides</span>
                                </a>
                                
                            </td>
                            <td>
                                
                                    <span class="btn-group">
                                        <a href="/training-material/topics/chip-seq/tutorials/formation_of_super-structures_on_xi/tutorial.html" class="btn btn-default" title="Where to run the tutorial">
                                            <i class="fa fa-laptop" aria-hidden="true"></i><span class="visually-hidden">tutorial</span>
                                        </a>
                                        <a href="#" class="btn btn-default dropdown-toggle dropdown-toggle-split" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                                            <span class="sr-only">Toggle Dropdown</span>
                                        </a>
                                        <ul class="dropdown-menu">
                                        
                                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=fr&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Fchip-seq%2Ftutorials%2Fformation_of_super-structures_on_xi%2Ftutorial.html&edit-text=&act=url" title="">
                                                FR
                                            </a>
                                        
                                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=ja&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Fchip-seq%2Ftutorials%2Fformation_of_super-structures_on_xi%2Ftutorial.html&edit-text=&act=url" title="">
                                                JA
                                            </a>
                                        
                                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=es&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Fchip-seq%2Ftutorials%2Fformation_of_super-structures_on_xi%2Ftutorial.html&edit-text=&act=url" title="">
                                                ES
                                            </a>
                                        
                                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=pt&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Fchip-seq%2Ftutorials%2Fformation_of_super-structures_on_xi%2Ftutorial.html&edit-text=&act=url" title="">
                                                PT
                                            </a>
                                        
                                            <a class="dropdown-item" href="https://translate.google.com/translate?hl=jp&sl=en&tl=ar&u=https%3A%2F%2Ftraining.galaxyproject.org%2Ftopics%2Fchip-seq%2Ftutorials%2Fformation_of_super-structures_on_xi%2Ftutorial.html&edit-text=&act=url" title="">
                                                AR
                                            </a>
                                        
                                    </span>
                                
                            </td>

                            
                                <td>
                                    
                                    <a class="topic-icon" h href="https://doi.org/10.5281/zenodo.1321974">
                                        <i class="fa fa-files-o" aria-hidden="true"></i><span class="visually-hidden">zenodo_link</span>
                                    </a>
                                    
                                </td>
                                <td>
                                    
                                        <a href="https://github.com/galaxyproject/training-material/tree/master/topics/chip-seq/tutorials/formation_of_super-structures_on_xi/workflows/" title="" alt="Formation of the Super-Structures on the Inactive X workflows">
                                            <i class="fa fa-share-alt" aria-hidden="true"></i><span class="visually-hidden">workflow</span>
                                        </a>
                                    
                                </td>
                                <td>
                                    
                                    <a href="https://github.com/galaxyproject/training-material/tree/master/topics/chip-seq/tutorials/formation_of_super-structures_on_xi/tours/">
                                        <i class="fa fa-magic" aria-hidden="true"></i><span class="visually-hidden">interactive_tour</span>
                                    </a>
                                    
                                </td>
                            

                            
                                <td>
                                
                                </td>
                            
                        
                    </tr>
                
            
                
            
                
            
        </table>

        
            <h2 id="docker_image">Galaxy instances</h2>
            <p>You can use a public Galaxy instance which has been tested for the availability of the used tools. They are listed along with the tutorials above.</p>
            
        

        <h2 id="maintainers">Maintainers</h2>
        <p>This material is maintained by:</p>
        <ul>
        
        
            <li>
                <a href="/training-material/hall-of-fame#malloryfreeberg">Mallory Freeberg</a>
                
            </li>
        
            <li>
                <a href="/training-material/hall-of-fame#moheydarian">Mo Heydarian</a>
                
            </li>
        
            <li>
                <a href="/training-material/hall-of-fame#friedue">Friederike Dündar</a>
                
            </li>
        
        </ul>

        <p>
            <em>For any question related to this topic and the content, you can contact them or visit our <a href="https://gitter.im/Galaxy-Training-Network/Lobby">Gitter channel</a>.</em>
        </p>

        <h2 id="contributors">Contributors</h2>
        <p>This material was contributed to by:</p>

        <!-- get list of unique contributors for all materials in the topic -->
        
        

        <ul>
        
        
            
            <li>
                <a href="/training-material/hall-of-fame#bebatut">Bérénice Batut</a>
            </li>
        
            
            <li>
                <a href="/training-material/hall-of-fame#friedue">Friederike Dündar</a>
            </li>
        
            
            <li>
                <a href="/training-material/hall-of-fame#erxleben">Anika Erxleben</a>
            </li>
        
            
            <li>
                <a href="/training-material/hall-of-fame#vivekbhr">Vivek Bhardwaj</a>
            </li>
        
            
            <li>
                <a href="/training-material/hall-of-fame#fidelram">Fidel Ramirez</a>
            </li>
        
            
            <li>
                <a href="/training-material/hall-of-fame#LeilyR">Leily Rabbani</a>
            </li>
        
            
            <li>
                <a href="/training-material/hall-of-fame#malloryfreeberg">Mallory Freeberg</a>
            </li>
        
            
            <li>
                <a href="/training-material/hall-of-fame#moheydarian">Mo Heydarian</a>
            </li>
        
            
            <li>
                <a href="/training-material/hall-of-fame#joachimwolff">Joachim Wolff</a>
            </li>
        
        </ul>

        
        <h2 id="references">References</h2>
        <ul>
        
            <li class="citation">
                <strong>Stephen G Landt et al:</strong> <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3431496/">ChIP-seq guidelines and practices of the ENCODE and modENCODE consortia</a>
                
                <br><cite>A very useful encyclopedic paper with many details about the tools the (mod)ENCODE consortia use and also contains a long section about antibody validation</cite>
                
            </li>
        
            <li class="citation">
                <strong>Gabriel E Zentner and Steven Henikoff:</strong> <a href="http://genomebiology.biomedcentral.com/articles/10.1186/gb-2012-13-10-250">Surveying the epigenomic landscape, one base at a time</a>
                
                <br><cite>Overview of popular sequencing techniques with very nice descriptions of DNase-seq, MNase-seq, FAIRE-seq</cite>
                
            </li>
        
            <li class="citation">
                <strong>Benjamin L Kidder et al:</strong> <a href="http://www.nature.com/ni/journal/v12/n10/abs/ni.2117.html">ChIP-Seq: technical considerations for obtaining high-quality data</a>
                
                <br><cite>Nice, readable introduction into all aspects of ChIP-seq experiments (from antibodies to cell numbers to replicates to data analysis)</cite>
                
            </li>
        
            <li class="citation">
                <strong>Marion Leleu et al:</strong> <a href="http://bfg.oxfordjournals.org/content/9/5-6/466">Processing and analyzing ChIP-seq data</a>
                
                <br><cite>Fairly detailed review of key concepts of ChIP-seq data processing (less detailed on analysis)</cite>
                
            </li>
        
            <li class="citation">
                <strong>Peter J. Park:</strong> <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3191340/">ChIP-seq: Advantages and challenges of a maturing technology</a>
                
            </li>
        
            <li class="citation">
                <strong>Peter V Kharchenko et al:</strong> <a href="https://www.ncbi.nlm.nih.gov/pubmed/19029915">Design and analysis of ChIP-seq experiments for DNA-binding proteins</a>
                
            </li>
        
            <li class="citation">
                <strong>Edison T Liu et al:</strong> <a href="http://bmcbiol.biomedcentral.com/articles/10.1186/1741-7007-8-56">Q&A: ChIP-seq technologies and the study of gene regulation</a>
                
                <br><cite>Short overview of several (typical) issues of ChIP-seq analysis</cite>
                
            </li>
        
            <li class="citation">
                <strong>Thomas S. Carroll et al:</strong> <a href="http://journal.frontiersin.org/article/10.3389/fgene.2014.00075/full">Impact of artifact removal on ChIP quality metrics in ChIP-seq and ChIP-exo data</a>
                
            </li>
        
            <li class="citation">
                <strong>Shirley Pepke et al:</strong> <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4121056/">Computation for ChIP-seq and RNA-seq studies</a>
                
                <br><cite>First comparison of peak callers, focuses on the explanation of basic principles of ChIP-seq data processing and general workflows of peak calling algorithms</cite>
                
            </li>
        
            <li class="citation">
                <strong>Elizabeth G. Wilbanks & Marc T. Facciotti:</strong> <a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0011471">Evaluation of Algorithm Performance in ChIP-Seq Peak Detection</a>
                
                <br><cite>Another comparison of peak callers - focuses more on the evaluation of the peak callers performances than Shirley Pepke et al.</cite>
                
            </li>
        
            <li class="citation">
                <strong>Mariann Micsinai et al:</strong> <a href="http://nar.oxfordjournals.org/content/40/9/e70.full">Picking ChIP-seq peak detectors for analyzing chromatin modification experiments</a>
                
                <br><cite>How to choose the best peak caller for your data set - their finding: default parameters, surprisingly, yield the most reproducible results regardless of the data set type</cite>
                
            </li>
        
            <li class="citation">
                <strong>Jianxing Fen et al:</strong> <a href="http://www.nature.com/nprot/journal/v7/n9/abs/nprot.2012.101.html">Identifying ChIP-seq enrichment using MACS</a>
                
                <br><cite>How to use MACS</cite>
                
            </li>
        
            <li class="citation">
                <strong>Yong Zhang et al:</strong> <a href="http://genomebiology.biomedcentral.com/articles/10.1186/gb-2008-9-9-r137">Model-based Analysis of ChIP-Seq (MACS)</a>
                
                <br><cite>Original publication of MACS</cite>
                
            </li>
        
            <li class="citation">
                <strong>Modan K Das & Ho-Kwok Dai:</strong> <a href="http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-8-S7-S21">A survey of DNA motif finding algorithms</a>
                
                <br><cite>Review of motif analysis tools</cite>
                
            </li>
        
            <li class="citation">
                <strong>Philip Machanick and Timothy L. Bailey:</strong> <a href="http://bioinformatics.oxfordjournals.org/content/27/12/1696.short">MEME-ChIP: motif analysis of large DNA datasets</a>
                
                <br><cite>MEME-ChIP-paper</cite>
                
            </li>
        
            <li class="citation">
                <strong>Timothy L. Bailey and Philip Machanick:</strong> <a href="http://nar.oxfordjournals.org/content/40/17/e128">Inferring direct DNA binding from ChIP-seq</a>
                
                <br><cite>Centrimo: position-specific motif analysis, especially useful for ChIP-seq data</cite>
                
            </li>
        
            <li class="citation">
                <strong>Morgane Thomas-Chollier et al:</strong> <a href="http://www.nature.com/nprot/journal/v6/n12/abs/nprot.2011.409.html">Transcription factor binding predictions using TRAP for the analysis of ChIP-seq data and regulatory SNPs</a>
                
                <br><cite>How to use TRAP</cite>
                
            </li>
        
            <li class="citation">
                <strong>Helge G. Roider et al:</strong> <a href="http://bioinformatics.oxfordjournals.org/content/23/2/134.short">Predicting transcription factor affinities to DNA from a biophysical model</a>
                
                <br><cite>Theoretical background of TRAP</cite>
                
            </li>
        
        </ul>
        
    </section>
</div>

<footer>
    <div class="container">
        <p>
            The <a href="https://galaxyproject.org/teach/gtn/">Galaxy Training Network</a>
            provides researchers with online training materials, connects them with local trainers, and helps promoting open data analysis practices worldwide.
        </p>
    </div>
</footer>


    </body>
    <script type="text/javascript" src="/training-material/assets/js/jquery.slim.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/popper.min.js"></script>
    <script type="text/javascript" src="/training-material/assets/js/bootstrap.min.js?v=3"></script>
    <script type="text/javascript" src="/training-material/assets/js/details-element-polyfill.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="/training-material/assets/js/main.js"></script>
</html>
